<!DOCTYPE HTML PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html dir="ltr" xml:lang="en" xmlns="http://www.w3.org/1999/xhtml" lang="en"><head>


    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <meta name="KEYWORDS" content="Tutorial - MatrixREDUCE">
<meta name="robots" content="index,follow">

    <title>MatrixREDUCE</title>
   <link rel="stylesheet" type="text/css" href="ncicb.css">
    
    <!--[if lt IE 5.5000]><style type="text/css">@import "/workbench/skins/monobook/IE50Fixes.css";</style><![endif]-->
    <!--[if IE 5.5000]><style type="text/css">@import "/workbench/skins/monobook/IE55Fixes.css";</style><![endif]-->
    <!--[if gte IE 6]><style type="text/css">@import "/workbench/skins/monobook/IE60Fixes.css";</style><![endif]-->
    <!--[if IE]><script type="text/javascript" src="/workbench/skins/common/IEFixes.js"></script>
    <meta http-equiv="imagetoolbar" content="no" /><![endif]-->

<h3>MatrixREDUCE</h3>
<div class="editsection" style="float: right; margin-left: 5px;"></div><a name="Overview"></a><h3>Overview</h3>
<p>MatrixREDUCE is a tool for inferring cis-regulatory elements and
transcriptional module activities from microarray data. It attempts to
calculate a sequence-specific binding affinity for putative
transcription factors. The sequence specificity of the transcription
factors' DNA-binding domain is modeled using a position-specific
affinity matrix (PSAM), representing the change in the binding affinity
(Kd) whenever a specific position within a reference binding sequence
is mutated. The resulting PSAM can be displayed as an affinity logo or
in a simple tabular format.
</p><p>For further details see <a href="http://e5.onthehub.com/WebStore/OfferingDetails.aspx?ws=9a885376-53d9-db11-bd2e-0015c55eef34&amp;o=e87d6240-18e2-db11-b93f-0030485a6b08" class="external text" title="http://e5.onthehub.com/WebStore/OfferingDetails.aspx?ws=9a885376-53d9-db11-bd2e-0015c55eef34&amp;o=e87d6240-18e2-db11-b93f-0030485a6b08" rel="nofollow">STV webpage for MatrixREDUCE</a>.
</p>
<div class="editsection" style="float: right; margin-left: 5px;"></div><a name="Data_Files"></a><h3>Data Files</h3>
<p>MatrixREDUCE operates on two input files: a microarray data set and
a FASTA file containing the DNA sequences corresponding to the
regulatory region of genes probed in the microarray set. The gene/probe
identifiers used in the microarray dataset and the sequence identifiers
used in the FASTA file must match. However, case is not important; the
program will change the identifiers to lower case before attempting to
match the records. If an identifier appears more than once in a file,
the last instance is used.
</p>
<div class="editsection" style="float: right; margin-left: 5px;"></div><a name="Parameters"></a><h3>Parameters</h3>
<p>(default values are shown in brackets [])
</p>
<p><b>p_value</b>       - p-value threshold to stop looking for new motifs [0.001].
</p>
<p><b>max_motif</b>     - Maximum # of motifs to search [20].
</p>
<p><strong>Topological Pattern</strong> - Specify Pattern, Load from file - length of seed pattern, e.g. N8 = length of eight [N8]. </p>
<p><strong>Strand</strong> - Auto-detect, Leading, Reverse, Both - specifies which strand to search on [Auto-detect]. </p>
<p><strong>Save Run Log</strong> - save the output from the MatrixReduce executable [not checked]. </p>
<p>&nbsp;</p>
<p><br>
</p>
<div class="editsection" style="float: right; margin-left: 5px;"></div><a name="Graphical_Interface_for_results"></a><h3>Graphical Interface for results</h3>
<a name="PSAM_Detail_Tab"></a><h4>PSAM Detail Tab</h4>
<p>The PSAM detail tab displays the result PSAMs in a table format.
Users can modify the display so that each PSAM is represented either by
its sequence logo or its consensus sequence. The results can be
printed, exported or captured as an image node in the project panel
</p>
<a name="Sequence_Tab"></a><h4>Sequence Tab</h4>
<p>The sequence tab displays the DNA sequences of the regulatory region
associated with each gene that is probed. Users can visualize the
matching scores of PSAMs against each sequence. Sequence score is the
product of weights w across all positions in the sliding window. </p><p>When a PSAM is selected, its weighted score is displayed
graphically on the input (upstream) sequences (Affinity score graph).
The system computes for every position an aggregate affinity score for
all selected PSAMs and plots the scores along all sequence positions.
Each score is between 0 and 1. Only scores larger than the designated
cut-off threshold are drawn.
</p>
<a name="Example"></a><h3>Example</h3>
<p>In this example, we will use two files that are included in the data
directory of the geWorkbench distribution iteself. They are
Spellman1998AlphaTimeCourseMod and YeastUpstream.fasta. SpellmanReduced.txt contains  data from Spellman (1998). YeastUpstream.fasta contains the
corresponding upstream DNA sequences for these genes.
</p>
<p><br>
1. In the Project Folders component, either use an existing Project, or create a new one.
</p><p>2. Right-click on Project and select "Open File(s)".
</p><p>3. Browse to the file SpellmanReduced.txt and set the file type
to "Tab-Delimted". This file is found in the data directory of the
geWorkbench installation. Open the file.
</p><p>4. You will be asked for an annotations file.  This is not needed for this example, so you can hit Cancel.
</p><p>5. Go to the Analysis tab and select Matrix Reduce.
</p>
<p>&nbsp;</p>
<p><img src="MatrixReduce_files/T_MatrixReduce_Main.png" width="882" height="345" /></p>
<p>&nbsp;</p>
<p><br>
6. To load the sequence file, click the "Load..." button.  Browser to the sequence file "YeastUpstream.fasta" and open it.
</p>
<p><img src="MatrixReduce_files/T_MatrixReduce_Load.png" width="840" height="372" /></p>
<p>7. Click Analyze to run MatrixREDUCE. (If you are running
geWorkbench from a console window using ANT, you can follow the
progress of the calculations there).
</p>
<p>8. The result is placed as a node beneath the parent microarray
dataset in the Project Folders component. At the same time, the results
are displayed in the Visual Area of geWorkbench.
</p><p>There are two tabs in the viewer, <b>PSAM Detail</b> and <b>Sequence</b>.  Within PSAM Detail there are two options.  The first is the Image view, which depicts the PSAM graphically.
</p>
<p><img src="MatrixReduce_files/T_MatrixReduce_PSAM.png" width="698" height="652" /></p>
<p>&nbsp;</p>
<p>&nbsp;
</p>
<p><br>
The second viewing option is the Name view, which just shows the consensus sequence without the weighted components.
</p>
<p>&nbsp;</p>
<p><img src="MatrixReduce_files/T_MatrixReduce_PSAM_Name.png" width="700" height="111" /></p>
<p>&nbsp;
</p>
<p>Finally, the Sequence tab depicts scores along each sequence.
</p><p><br>
  <img src="MatrixReduce_files/T_MatrixReduce_Sequence.png" width="701" height="511" /></p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<h3>Services Tab </h3>
<p>MatrixReduce can also be run as a caGrid Service by choosing the Grid option on the services tab. </p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<div class="editsection" style="float: right; margin-left: 5px;"></div><a name="References"></a><h3>References</h3>
<p>Foat BC et al., (2005). Profiling condition-specific, genome-wide
regulation of mRNA stability in yeast. PNAS 102(49), 17675-17680. </p><p>Foat BC, Morozov AV, Bussemaker HJ. (2006). Statistical
mechanical modeling of genome-wide transcription factor occupancy data
by MatrixREDUCE. Bioinformatics 22(14):e141-e149.
</p><p>Spellman et al., (1998). Comprehensive Identification of Cell
Cycle-regulated Genes of the Yeast Saccharomyces cerevisiae by
Microarray Hybridization. Molecular Biology of the Cell 9, 3273-3297.
</p>
