<html>

<head>
<meta http-equiv=Content-Type content="text/html; charset=us-ascii">
<meta name=Generator content="Microsoft Word 11 (filtered)">
<title>Pattern Discovery Overview</title>

<style>
<!--
 /* Font Definitions */
 @font-face
	{font-family:PMingLiU;
	panose-1:2 2 3 0 0 0 0 0 0 0;}
@font-face
	{font-family:"Arial Unicode MS";
	panose-1:2 11 6 4 2 2 2 2 2 4;}
@font-face
	{font-family:Verdana;
	panose-1:2 11 6 4 3 5 4 4 2 4;}
@font-face
	{font-family:"\@Arial Unicode MS";
	panose-1:2 11 6 4 2 2 2 2 2 4;}
@font-face
	{font-family:"\@PMingLiU";
	panose-1:2 2 3 0 0 0 0 0 0 0;}
 /* Style Definitions */
 p.MsoNormal, li.MsoNormal, div.MsoNormal
	{margin:0cm;
	margin-bottom:.0001pt;
	font-size:12.0pt;
	font-family:"Times New Roman";}
h3
	{margin-right:0cm;
	margin-left:0cm;
	line-height:10.5pt;
	font-size:10.0pt;
	font-family:"Times New Roman";
	font-weight:bold;}
a:link, span.MsoHyperlink
	{color:#7373A5;
	text-decoration:underline;}
a:visited, span.MsoHyperlinkFollowed
	{color:#7373A5;
	text-decoration:underline;}
p
	{margin-right:0cm;
	margin-left:0cm;
	font-size:12.0pt;
	font-family:"Times New Roman";}
p.contentsides, li.contentsides, div.contentsides
	{margin-right:0cm;
	margin-left:0cm;
	font-size:10.0pt;
	font-family:Verdana;
	color:#312C2C;}
p.blueannounce, li.blueannounce, div.blueannounce
	{margin-right:0cm;
	margin-left:0cm;
	line-height:12.0pt;
	font-size:10.0pt;
	font-family:Verdana;
	color:#5A6B9F;}
p.content, li.content, div.content
	{margin-right:0cm;
	margin-left:0cm;
	line-height:10.5pt;
	font-size:10.0pt;
	font-family:Verdana;
	color:#312C2C;}
p.sidenavheader, li.sidenavheader, div.sidenavheader
	{margin-right:0cm;
	margin-left:0cm;
	font-size:10.0pt;
	font-family:Verdana;
	color:#76769A;
	font-weight:bold;}
p.style1, li.style1, div.style1
	{margin-right:0cm;
	margin-left:0cm;
	font-size:8.5pt;
	font-family:Verdana;
	color:#312C2C;}
p.Default, li.Default, div.Default
	{margin:0cm;
	margin-bottom:.0001pt;
	text-autospace:none;
	font-size:12.0pt;
	font-family:"Times New Roman";
	color:black;}
span.style11
	{font-family:Verdana;
	color:#312C2C;}
@page Section1
	{size:612.0pt 792.0pt;
	margin:72.0pt 90.0pt 72.0pt 90.0pt;}
div.Section1
	{page:Section1;}
 /* List Definitions */
 ol
	{margin-bottom:0cm;}
ul
	{margin-bottom:0cm;}
-->
</style>

</head>

<body lang=ZH-TW link="#7373A5" vlink="#7373A5">

<div class=Section1>

<h3><span lang=EN-US style='font-size:13.5pt;font-family:Verdana;color:#312C2C'>Pattern
Discovery</span></h3>

<p style='line-height:10.5pt'><span lang=EN-US style='font-size:8.5pt;
font-family:Verdana;color:#312C2C'>The Pattern Discovery is located in the <b>Selection
Area</b> in the lower left of the application.&nbsp; This component runs
SPLASH,&nbsp; an efficient application that identifies sparse amino acid or
nucleic acid patterns (called &quot;motifs&quot;) that match identically or
similarly in a set of protein or DNA sequences. For a full description of
SPLASH, go to <a
href="http://sharepoint.cu-genome.org/exchweb/bin/redir.asp%3fURL=http:/www.research.ibm.com/splash/index.htm"
target="_blank">http://www.research.ibm.com/splash/index.htm.</a> </span></p>

<h3><span lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C;
font-weight:normal'>SPLASH is hosted on a remote server. The caWorkbench
component communicates with the server via a SOAP interface. Search parameters
and protein/DNA sequences in FASTA file format are uploaded to the server via
SOAP. SPLASH is run and the matching sequences are retrieved and displayed in
the selection area. &nbsp;&nbsp;This section describes the<a
href="#_Basic_Usage"> basic usage</a> of the pattern discovery panel, <a
href="#_Modes_of_Pattern_Discovery">modes of pattern discovery</a> and <a
href="#_Pattern_Discovery_Results">Pattern Discovery Results</a>.</span></h3>

<p style='line-height:10.5pt'><a name="_Basic_Usage"></a><span lang=EN-US
style='font-size:8.5pt;font-family:Verdana;color:#312C2C'><img border=0
width=628 height=245 src="DiscoveryOverviewed_files/image001.jpg"></span></p>

<h3><span lang=EN-US style='font-family:Verdana;color:#312C2C'>Basic Usage</span></h3>

<p style='line-height:10.5pt'><span lang=EN-US style='font-size:8.5pt;
font-family:Verdana;color:#312C2C'>The steps of basic usage are indicated in
the screenshot above.</span></p>

<p style='margin-left:18.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>A.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>The modes
of Pattern Discovery are:<a href="#_Hierarchical_and_Exhaustive_Discove">
Normal Discovery</a>,<a href="#_Hierarchical_Discovery"> Hierarchical Discovery</a>
and<a href="#_Exhaustive_Discovery"> Exhaustive Discovery</a>. Across the top
of the Pattern Discovery panel are three radio buttons (Norm., Hierarch., and
Exhaust.) that are used to select the discovery mode. </span></p>

<p style='margin-left:18.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>B.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>The
default &quot;Basic&quot; parameters are shown. The Parameters help section
includes a full description of the Pattern Discovery parameters.</span></p>

<p style='margin-left:18.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>C.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>Specifies
use of a Globus (<a href="http://www.globus.org/toolkit/">http://www.globus.org/toolkit/</a>
) instance of SPLASH for Pattern discovery. This allows for discovery of SPLASH
instances running on the internet in accordance with the Globus
Alliance&#8217;s requirements (server discovery not fully implemented). If
unselected, a GSOAP (<a href="http://gsoap2.sourceforge.net/">http://gsoap2.sourceforge.net/</a>
) implementation of a SPLASH web service hosted at Columbia University is used.</span></p>

<p style='margin-left:18.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>D.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>Clicking
on the clockwise blue arrow will <b>Start Pattern Discovery</b>. This brings up
a <b>New Session</b> dialog box. The first time Pattern Discovery is run, the
dialog box must be filled out. Use the drop down menu to select the Server Type
and Server that hosts SPLASH. Then type-in the port and User Name. The image
below is a connection to a server running on the local host at port 80.</span></p>

<p style='margin-left:18.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Available
SOAP SPLASH server(s) are at </span><u><span lang=EN-US style='font-size:8.5pt;
font-family:Verdana;color:blue'>splash.cu-genome.org</span></u><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'> at port
80. Please select &#8216;splash&#8217; as the server type.</span></p>

<p style='margin-left:18.0pt'><span lang=EN-US style='color:#312C2C'><img
border=0 width=461 height=256 src="DiscoveryOverviewed_files/image002.jpg"></span></p>

<p style='margin-left:18.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>E.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>Clicking
on the stop sign icon will<b> Stop Pattern Discovery.</b></span></p>

<p style='margin-left:18.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>F.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>Clicking
on the folder icon will <b>Load a Pattern File</b> motif files (.pat).</span></p>

<p style='margin-left:18.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>G.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>The progress
bar used to provide a visual update of the percentage completion.</span></p>

<h3><span lang=EN-US style='font-size:12.0pt;font-family:Verdana;color:#312C2C'>Pattern
Discovery Results</span></h3>

<p style='line-height:10.5pt'><span lang=EN-US style='font-size:8.5pt;
font-family:Verdana;color:#312C2C'>Shown below are the results of a discovery
using default parameters.&nbsp; The Discovery Table is filled with information
on the discovered motifs. The steps of basic usage are indicated with letter in
the below screenshot.</span></p>

<p style='line-height:10.5pt'><span lang=EN-US style='font-size:10.0pt;
font-family:Verdana;color:#312C2C'><img border=0 width=701 height=433
src="DiscoveryOverviewed_files/image003.jpg"></span></p>

<p style='margin-left:18.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>A.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>The number
of patterns found is displayed below the discovery table.</span></p>

<p style='margin-left:18.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:10.0pt;font-family:Verdana;color:#312C2C'>B.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>Right
clicking on a pattern displays the option to:</span></p>

<p style='margin-left:36.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:10.0pt;font-family:"Arial Unicode MS";color:#312C2C'>&#10170;<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>Mask
pattern- suppresses the selected pattern from the set of patterns discovered.
So the sequence panel for instance will not show a masked pattern match in one
of its modes of operation and will exist in any ancillary dataset created (see
below).</span></p>

<p style='margin-left:36.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:10.0pt;font-family:"Arial Unicode MS";color:#312C2C'>&#10170;<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>Unmask all
patterns- causes all patterns discovered to be &#8220;active&#8221; and these
will be shown in any ancillary dataset created.<span style='background:lime'> </span></span></p>

<p style='margin-left:36.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:10.0pt;font-family:"Arial Unicode MS";color:#312C2C'>&#10170;<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>Save
patterns- caves the regular expression corresponding to the selected pattern to
file.</span></p>

<p style='margin-left:36.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:10.0pt;font-family:"Arial Unicode MS";color:#312C2C'>&#10170;<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>Save
selected patterns-</span><span lang=EN-US> s</span><span lang=EN-US
style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>ave the selected
patterns to file along with match coordinates on sequences on which the pattern
discovery was performed.</span></p>

<p style='margin-left:36.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:10.0pt;font-family:"Arial Unicode MS";color:#312C2C'>&#10170;<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>Save All
patterns- save all patterns to file along with match coordinates.</span></p>

<p style='margin-left:36.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:10.0pt;font-family:"Arial Unicode MS";color:#312C2C'>&#10170;<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>Add
pattern to project- creates an ancillary dataset in the Project Panel under the
sequence dataset used for pattern discovery.</span></p>

<p style='margin-left:18.0pt;text-indent:-18.0pt;line-height:10.5pt'><span
lang=EN-US style='font-size:10.0pt;font-family:Verdana;color:#312C2C'>C.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp; </span></span><span
lang=EN-US style='font-size:8.5pt;font-family:Verdana;color:#312C2C'>To display
selected motifs and their frequencies in the dataset, left click on a
pattern(s).&nbsp; The selected pattern(s) are then shaded gray in the discovery
table and are displayed graphically in the Sequence Panel, Promoter Panel and
Patterns.</span></p>

<h3><a name="_Modes_of_Pattern_Discovery"></a><span lang=EN-US
style='font-size:12.0pt;font-family:Verdana;color:#312C2C'>Modes of Pattern
Discovery</span></h3>

<h3><span lang=EN-US style='font-family:Verdana;color:#312C2C'>Normal Discovery</span></h3>

<p style='line-height:10.5pt'><span lang=EN-US style='font-size:8.5pt;
font-family:Verdana;color:#312C2C'>Normal discovery is based on the algorithm
SPLASH (Califano, A., 2000); it generates a list of all regular expression
patterns (motifs) that satisfy a user-defined minimum support and a minimum
density criteria.</span></p>

<b><span lang=EN-US style='font-size:10.0pt;font-family:"Times New Roman";
color:black'><br clear=all style='page-break-before:always'>
</span></b>

<h3><a name="_Hierarchical_Discovery"></a><span lang=EN-US style='font-family:
Verdana;color:#312C2C'>Hierarchical Discovery</span></h3>

<p style='line-height:10.5pt'><span lang=EN-US style='font-size:8.5pt;
font-family:Verdana;color:#312C2C'>Hierarchical discovery is used to analyze a
sequence database and produce a sequence of clusters based on discovered
patterns. For more information, visit <a
href="http://sharepoint.cu-genome.org/exchweb/bin/redir.asp%3fURL=http:/www.research.ibm.com/splash/Hyerarchical/HierarchicalDiscovery.htm"
target="_blank">http://www.research.ibm.com/splash/Hyerarchical/HierarchicalDiscovery.htm</a>.
</span></p>

<h3><a name="_Exhaustive_Discovery"></a><span lang=EN-US style='font-family:
Verdana;color:#312C2C'>Exhaustive Discovery</span></h3>

<p style='line-height:10.5pt'><span lang=EN-US style='font-size:8.5pt;
font-family:Verdana;color:#312C2C'>Exhaustive discovery can be used to exhaustively
analyze a sequence database for all non-overlapping motifs that are
statistically significant. This is useful for discovering in order of relative
sequence support, all regions of a protein family that have been preserved by
evolution and may therefore play a functional or structural role. For more
information, visit <a
href="http://sharepoint.cu-genome.org/exchweb/bin/redir.asp%3fURL=http:/www.research.ibm.com/splash/Exhaustive/ExhaustiveDiscovery.htm"
target="_blank">http://www.research.ibm.com/splash/Exhaustive/ExhaustiveDiscovery.htm</a>.
</span></p>

</div>

</body>

</html>
