<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html dir="ltr" lang="en"><head>
<title>Master Regulator Analysis - Workbench</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="generator" content="MediaWiki 1.16.0">


</head>
<body class="mediawiki ltr ns-0 ns-subject page-Master_Regulator_Analysis skin-monobook">
<div id="globalWrapper">
<div id="column-content"><div id="content">
	<a id="top"></a>
	
	<h1 id="firstHeading" class="firstHeading">Master Regulator Analysis</h1>
	
		<!-- start content -->
<p><br>
</p>
<a name="Overview"></a>
<h1> <span class="mw-headline" id="Overview">Overview</span></h1>
<p>Regulatory activity in the context of specific cellular phenotypes 
can be investigated using interaction networks. These are graphs where 
nodes represent genes and an edge between two nodes A and B means that 
genes <i>A</i> and <i>B</i> are participants in the same regulatory activity. E.g., <i>A</i> can be a transcription factor for <i>B</i>; or, <i>A</i> can be an miRNA that silences <i>B</i>. Analysis of such regulatory networks <a href="#Basso2005"> [Basso et al., 2005]</a>
 has convincingly demonstrated their scale-free nature which is 
dominated by a relatively small number of nodes with a large degree of 
connectivity. The genes corresponding to those nodes are known as 
"master regulators" and collectively orchestrate the regulatory program 
of the underlying cellular phenotype(s). 
</p><p>Master Regulator analysis <a href="#Lefebvre2010"> [Lefebvre et al., 2010]</a>
 is an algorithm used to identify transcription factors whose targets 
(e.g., as represented in an ARACNe-generated interactome) are enriched 
for a particular gene signature.  The enrichment is evaluated using a 
statistical test such as Fisher’s exact test.  The objective is to place
 the signature genes within a regulatory context and identify the master
 regulators responsible for coordinating their activity, thus 
highlighting the regulatory apparatus driving phenotypic 
differentiation. 
</p><p>Specifically, given an interaction network <i>I</i>, a (presumed) master regulator gene <i>A</i>, and a set of signature genes, MRA computes the intersection between two sets of genes:
</p>
<ol><li> The neighbors of <i>A</i> in the interaction network <i>I</i> (this gene set is called the <i><b>regulon</b></i> of <i>A</i>).
</li><li> The set of signature genes.  The signature may be supplied 
independently or calculated from e.g. a differential expression 
experiment.
</li></ol>
<p>Interaction networks are represented as "adjacency matrices".  An 
adjacency matrix lists the connections that each node takes part in, and
 includes a measure of the strength of that interaction (e.g. the mutual
 information in the case of matrices generated by ARACNe).
</p><p>Fisher’s exact test is used to quantify how likely it is to 
encounter an intersection of (at least) the observed size by chance 
alone. A small p-value is taken to imply that gene <i>A</i> may play a significant role in mediating the regulatory program that leads to the differential phenotypes.
</p><p>Generating a display of the effect of a master regulator on its 
regulon requires performing a t-test of differential expression on a 
dataset representing the phenotypes distinguished by the signature, 
ideally the original dataset from which the signature was determined.
</p>
<a name="Setting_up_an_MRA_run"> </a>
<h1> <span class="mw-headline" id="Setting_up_an_MRA_run">Setting up an MRA run</span></h1>
<h2> <span class="mw-headline" id="Prerequisites">Prerequisites</span></h2>
<ul><li> The Master Regulator Analysis (MRA) component must be loaded in the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Tutorial_-_Component_Configuration_Manager" title="Tutorial - Component Configuration Manager" class="mw-redirect"> Component Configuration Manager</a>.
<ul><li> The MRA component will be listed along with the other analysis routines within the geWorkbench Analysis Panel.
</li></ul>
</li><li> <b>Interaction Network</b> - An interaction network calculated
 with ARACNe from a dataset which includes the particular cellular 
phenotypes being investigated.  In calculating the network, all genes 
that will be tested as possible master regulators should be used as hubs
 in the ARACNe calculation.
</li><li> <b>Signature genes</b> - A list of signature genes which 
distinguish between two phenotypes.  This list may come from a t-test, 
clustering, or some combination of methods.  The user must define this 
set using methods relevant to the particular dataset and study goals.
</li><li> <b>Candidate master regulator list</b> -  A set of genes that 
will be tested as candidate master regulators.  This set may be 
comprised of e.g. transcription factor and signalling pathway genes.
</li><li> <b>Gene Expression dataset</b> - A gene expression dataset in 
which the phenotypic signature was identified or can be demonstrated.  A
 t-test of differential expression will be run to generate the graphic 
"bar code" display of the effect of the master regulator on its regulon.
 
</li></ul>
<h2> <span class="mw-headline" id="Parameters_and_Settings">Parameters and Settings</span></h2>
<p><img alt="MRA Parameters panel.png" src="MasterRegulatorAnalysis_files/MRA_Parameters_panel.png" height="444" width="1212">
</p>
<h3> <span class="mw-headline" id="Load_Network">Load Network</span></h3>
<p>There are 2 ways to designate the interaction network, represented by
 an adjacency matrix, that will be used for computing the regulons of 
the candidate master regulator genes:
</p>
<ul><li> <b>From File</b>: by choosing a file that describes a network. 
</li><li> <b>From Project</b>: by selecting an adjacency matrix node from the Project Folders component. Several analytical components in geWorkbench (e.g., <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Tutorial_-_ARACNE" title="Tutorial - ARACNE" class="mw-redirect"> ARACNE</a>) produce adjacency matrix results nodes that can be utilized for this purpose.
</li></ul>
<p>All edges in the network are assumed to be significant, and any strength value included is not used.
</p>
<h3> <span class="mw-headline" id="Master_Regulators">Master Regulators</span></h3>
<p>A set of candidate master regulator markers.  This set must be loaded
 into the Markers component before running MRA.  This can be read in 
either directly as markers, or as gene symbols.
</p>
<h3> <span class="mw-headline" id="Signature_Markers">Signature Markers</span></h3>
<p>A set of markers comprising the signature that distinguishes the 
chosen phenotype from others.  This set must be loaded into the Markers 
component before running MRA.  It can be read in either directly as 
markers, or as gene symbols (but be aware that a gene can be represented
 by more than one marker, and not all may correspond to the signature).
</p>
<h3> <span class="mw-headline" id="Fisher.27s_Exact_Test_threshold">Fisher's Exact Test threshold</span></h3>
<p>Enter a p-value for the significance at which to accept the overlap 
of the regulon of a candidate TF and the signature set of genes.
</p>
<h3> <span class="mw-headline" id="T-test_for_differential_expression">T-test for differential expression</span></h3>
<p>A "bar-code" graphic is generated using a t-test on a differential 
expression dataset.  However, all t-values are accepted (critical alpha =
 1) and used to order the bars representing the regulon markers.
</p><p>All  that is required is to define sets of arrays representing 
two phenotypes of interest (and distinguished by the signature). At 
least two sets of arrays must be activated, and at least one marked as 
"case", representing the target phenotype of the gene signature.  
"Control" is the default classification.  See also the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Differential_Expression" title="Differential Expression"> Differential Expression tutorial</a>).
</p><p><br>
  <img alt="Array set class assignment MRA.png" src="MasterRegulatorAnalysis_files/Array_set_class_assignment_MRA.png" height="400" width="228">
</p>
<a name="Viewing_MRA_analysis_results"></a>
<h1> <span class="mw-headline" id="Viewing_MRA_analysis_results">Viewing MRA analysis results</span></h1>
<p>Following the successful completion of the MRA computation, a result 
node appears in the Project Folder area of the geWorkbench interface, 
under the microarray experiment node used for the t-test:
</p><p><img alt="MRA results node.png" src="MasterRegulatorAnalysis_files/MRA_results_node.png" height="135" width="229"> 
</p>
<p>The results of the analysis can be visualized in the MRA Viewer component by selecting the result node.
</p>
<h2> <span class="mw-headline" id="MRA_Results_Viewer">MRA Results Viewer</span></h2>
<p>The MRA viewer is structured in 3 distinct areas.
</p><p><br>
  <img alt="MRA viewer GBM FOSL2.png" src="MasterRegulatorAnalysis_files/MRA_viewer_GBM_FOSL2.png" height="623" width="986">
</p>
<p><br>
</p>
<h3> <span class="mw-headline" id="Summary_Listing">Summary Listing</span></h3>
<p><img alt="MRA Summary listing.png" src="MasterRegulatorAnalysis_files/MRA_Summary_listing.png" height="179" width="481">
</p>
<p><br>
At upper left in the MRA viewer.  For each candidate master regulator 
found to have a significant effect using Fisher's Exact test, the 
following four columns are displayed:
</p>
<ul><li> <b>Master Regulator</b> - This is either the master regulator 
gene name or the marker/probeset name identifying the corresponding 
array feature (depending on the selection of the radio buttons “Symbol” 
and “Probe set”).
</li><li> <b>FET p-value</b>  -  the p-value from Fisher’s exact test. 
The test utilizes a 2x2 contingency table where rows classify markers as
 belonging to the signature set or not, while columns indicate if a 
marker belongs to the regulon of the master regulator or not. Counts are
 computed using all markers found in the input experiment data. 
(Fischer's exact test includes p-values for more-extreme tables).  
</li><li> <b>Genes in Regulon</b> - the number of markers (genes) found to be first neighbors of the master regulator in the loaded network - its regulon.
</li><li> <b>Genes in Intersection Set</b> - The number of markers found in the intersection of the signature and the regulon of the candidate MR.
</li></ul>
<p>The contents of the table can be ordered by any column, by clicking on the column name.
</p><p>Clicking on the radio button for any of the master regulators 
will display the list of intersection genes in a table to the right 
(Detailed Listing), and will draw the regulon bar graph below.
</p>
<h3> <span class="mw-headline" id="Detailed_Listing">Detailed Listing</span></h3>
<p>The detailed list shows the genes/markers contained in the intersection set of the  MR regulon and the signature.
</p><p><br>
  <img alt="MRA Detailed listing.png" src="MasterRegulatorAnalysis_files/MRA_Detailed_listing.png" height="186" width="471">
</p>
<p>The genes are displayed in a table with the following columns:
</p>
<ul><li> <b>Genes in intersection set</b>: the names of the genes in the
 intersection set. Either the gene name or the marker/probe set name is 
used (based on the choice of "Symbol" or "Probe Set" radio buttons).
</li><li> <b>T-test value</b>: The actual value of the t-test statistic 
for the gene. A positive value indicates that the expression of the gene
 is higher in cases than in controls. A negative value has the opposite 
meaning.
</li></ul>
<p><br>
</p>
<h3> <span class="mw-headline" id="Graph_View">Graph View</span></h3>
<p>For a given master regulator <i>A</i> and the intersection between 
its regulon and the set of differentially expressed genes, the graph 
view helps assess if the intersection genes are preferentially 
over-expressed in the cases versus the controls. The biological 
motivation comes from observing <a href="#Lim2009"> [Lim et al., 2009]</a>
 that regulators with multiple targets tend to affect the expression 
level of (most of) their targets in one particular direction: they 
either promote their expression or inhibit it; but they rarely do both 
equally. 
</p>
<ul><li> The red-blue gradient at the bottom of the graph represents the
 range between the lowest (blue)  and the highest (red) t-test statistic
 recorded among all differentially expressed genes. The white area in 
the middle represents zero.
</li><li> The vertical bars correspond to the genes displayed in the 
table under the “Detailed listing” portion of the interface, i.e., the 
intersection between the differentially expressed genes and the regulon 
of the master regulator <i>A</i> currently selected within the “Summary listing” table. 
</li><li> The relative location of a bar on the gradient represents the t-test statistic recorded for the corresponding gene. 
</li><li> Further, the color of each bar provides information about the 
correlation between the expression levels of the target gene and the 
putative master regulator A (correlations are computed as Pearson’s 
correlation, using data from all microarrays in the experiment): red 
means that the two genes are positively correlated (r &gt; 0) while blue
 means that correlation is negative (r &lt; 0).
</li></ul>
<p><br>
FOSL2:
</p><p><img alt="MRA graph GBM FOSL2.png" src="MasterRegulatorAnalysis_files/MRA_graph_GBM_FOSL2.png" height="98" width="960">
</p>
<p>The bar graph shown above, for FOSL2, indicates that more of its 
regulon genes have positive differential expression in the mesenchymal 
phenotype, and that expression of FOSL2 is positively correlated with 
the differential expression of its regulon in the mesenchymal phenotype.
 
</p><p><br>
ZNF238:
</p><p><img alt="MRA graph GBM ZNF238.png" src="MasterRegulatorAnalysis_files/MRA_graph_GBM_ZNF238.png" height="102" width="961">
</p>
<p>The bar graph shown above, for ZNF238, indicates that  expression
 of ZNF238 is negatively correlated with the differential expression of 
its regulon in the mesenchymal phenotype.
</p>
<h3> <span class="mw-headline" id="Control_buttons">Control buttons</span></h3>
<p>Additional functionality is made available through the following buttons: 
</p>
<ul><li> <b>Add to Set</b>: creates a set containing the markers that 
correspond to the genes currently displayed in the table. The new marker
 set appears in the Markers component and is named after the master 
regulator.
</li><li> <b>Export selected</b>: same as “Add to Set” but instead of being added to a marker set, the markers are stored into a file.
</li><li> <b>Export all</b>: stores into a file information for all 
master regulators (instead of only the one currently selected in the 
“Summary” view). Specifically, for each master regulator, the file lists
 the markers for all genes that are both differential expressed and also
 belong to the master regulator’s regulon.
</li></ul>
<a name="Dataset_History"></a>
<h1> <span class="mw-headline" id="Dataset_History">Dataset History</span></h1>
<p>Each results node stores the parameter settings used to setup the 
corresponding MRA run. The specific parameter values can be inspected 
within the Dataset History component, after clicking on the MRA results 
node in the Project Folders pane.
</p>
<a name="Example_of_running_MRA"></a>
<h1> <span class="mw-headline" id="Example_of_running_MRA">Example of running MRA</span></h1>
<p>This example uses a dataset comprised of 176 microarrays described in
 Phillips (2006).  The analysis follows that described in Carro et al. 
(2010) for master regulators of Glioblastoma.
</p><p><br>
</p>
<h2> <span class="mw-headline" id="Loading_and_preparing_the_example_data">Loading and preparing the example data</span></h2>
<h3> <span class="mw-headline" id="Microarray_dataset">Microarray dataset</span></h3>
<ol><li> Load a microarray dataset.  (See <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Tutorial_-_Local_Data_Files" title="Tutorial - Local Data Files" class="mw-redirect"> Local Data Files</a>).
</li><li> When prompted, load the annotation file.
</li></ol>
<h3> <span class="mw-headline" id="Marker_sets">Marker sets</span></h3>
<p>Load marker sets for:
</p>
<ol><li> the list of candidate master regulators
</li><li> the signature genes.
</li></ol>
<p><img alt="MRA GBM Marker sets.png" src="MasterRegulatorAnalysis_files/MRA_GBM_Marker_sets.png" height="402" width="246">
</p>
<h3> <span class="mw-headline" id="Array_sets">Array sets</span></h3>
<p>Array sets are shown defined for the three phenotypic classes of 
arrays in the dataset: Mesenchymal (MES), Proneural (PN), and 
Proliferative (Prolif).
</p>
<ul><li> MES and PN are "activated" for use in the t-test by checking the boxes next their names.
</li><li> The MES set is classifed as "Case".  Right click on the thumbtack adjacent to the set name.
</li></ul>
<p><img alt="Array set class assignment MRA.png" src="MasterRegulatorAnalysis_files/Array_set_class_assignment_MRA.png" height="400" width="228">
</p>
<h2> <span class="mw-headline" id="Setting_up_the_parameters_and_starting_MRA">Setting up the parameters and starting MRA</span></h2>
<p>In the Analysis Panel, select the "MRA Analysis" entry and set the parameters as follows:
</p>
<ul><li> <b>Load Network</b> - from the drop down choose the option "From Set", click on the "Load" button, and select the desired set.
</li><li> <b>Master regulators</b> - select the desired set from those loaded in the Markers component.
</li><li> <b>Signature markers</b> - select the desired set from those loaded in the Markers component.
</li><li> <b>P-value</b> - The p-value for the FET may be set as desired.
</li></ul>
<p><img alt="MRA GBM param setup.png" src="MasterRegulatorAnalysis_files/MRA_GBM_param_setup.png" height="306" width="1213">
</p>
<p><br>
</p>
<ul><li> Click on the <b>Analyze</b> button.
</li></ul>
<h2> <span class="mw-headline" id="Results">Results</span></h2>
<p>Upon completion of the analysis, an MRA results node is placed in the
 Project Folders tree. The analysis results can be browsed using the MRA
 viewer and are as shown above in the MRA Results Viewer section.
</p>
<p>&nbsp;</p>
<a name="References"></a>
<h1> <span class="mw-headline" id="References">References</span></h1>
<ul>
  <li> Basso K, Margolin AA, Stolovitzky G, Klein U, Dalla-Favera R, 
Califano A (2005). Reverse engineering of regulatory networks in human B
 cells. Nat Genet 37(4):382-390 (<a href="http://www.nature.com/ng/journal/v37/n4/abs/ng1532.html" class="external text" rel="nofollow">link to paper</a>).
</li>
 <li> Carro MS, Lim WK, Alvarez MJ, Bollo RJ, Zhao X, Snyder EY, 
Sulman EP, Anne SL, Doetsch F, Colman H, Lasorella A, Aldape K, Califano
 A, Iavarone A  (2010)  The transcriptional network for mesenchymal 
transformation of brain tumors.  Nature 463(7279):318-25.
</li>
<li> Lefebvre C, Rajbhandari P, Alvarez MJ, Bandaru P, Lim WK, Sato
 M, Wang K, Sumazin P, Kustagi M, Bisikirska BC, Basso K, Beltrao P, 
Krogan N, Gautier J, Dalla-Favera R, Califano A (2010)  A human B-cell 
interactome identifies MYB and FOXM1 as master regulators of 
proliferation in germinal centers.  Mol Syst Biol.  6:377. PMID: 
20531406 (<a href="http://www.ncbi.nlm.nih.gov/pubmed/20531406" class="external text" rel="nofollow">link to paper</a>).
</li>
<li> Lim WK, Lyashenko E, Califano A: Master regulators used as 
breast cancer metastasis classifier. Pac Symp Biocomput. 2009:504-15 (<a href="http://psb.stanford.edu/psb-online/proceedings/psb09/lim.pdf" class="external text" rel="nofollow">link to paper</a>).
</li>
<li> Phillips HS, Kharbanda S, Chen R, Forrest WF, Soriano RH, Wu 
TD, Misra A, Nigro JM, Colman H, Soroceanu L, Williams PM, Modrusan Z, 
Feuerstein BG, Aldape K (2006)  Molecular subclasses of high-grade 
glioma predict prognosis, delineate a pattern of disease progression, 
and resemble stages in neurogenesis.  Cancer Cell 9(3):157-73.
</li>

</body></html>