<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html dir="ltr" lang="en"><head>
<title>Promoter Analysis - Workbench</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="generator" content="MediaWiki 1.16.0">
<meta name="robots" content="noindex,follow">
<link rel="alternate" type="application/x-wiki" title="Edit" href="http://wiki.c2b2.columbia.edu/workbench/index.php?title=Promoter_Analysis&amp;action=edit">
<link rel="edit" title="Edit" href="http://wiki.c2b2.columbia.edu/workbench/index.php?title=Promoter_Analysis&amp;action=edit">
<link rel="shortcut icon" href="http://wiki.c2b2.columbia.edu/workbench/skins/common/images/favicon.ico">
<link rel="search" type="application/opensearchdescription+xml" href="http://wiki.c2b2.columbia.edu/workbench/opensearch_desc.php" title="Workbench (en)">
<link rel="alternate" type="application/atom+xml" title="Workbench Atom feed" href="http://wiki.c2b2.columbia.edu/workbench/index.php?title=Special:RecentChanges&amp;feed=atom">
<link rel="stylesheet" href="promoter_files/commonPrint.css" media="">
<link rel="stylesheet" href="promoter_files/index_003.css">
<link rel="stylesheet" href="promoter_files/index.css" media="">
<link rel="stylesheet" href="promoter_files/index_002.css">
<link rel="stylesheet" href="promoter_files/index_004.css">

</head>
<body class="mediawiki ltr ns-0 ns-subject page-Promoter_Analysis skin-monobook">
<div id="globalWrapper">
<div id="column-content"><div id="content">
	<a id="top"></a>
	
	<h1 id="firstHeading" class="firstHeading">Promoter Analysis</h1>
	<div id="bodyContent">
		<!-- start content -->
<p><br>
</p>
<a name="Overview"></a><h1> <span class="mw-headline" id="Overview">Overview</span></h1>
<p>The <b>Promoter</b> component scans one or more sequence profiles 
against nucleotide sequences that the user has loaded into geWorbench.  
Motifs from the JASPAR database of transcription factor binding sites 
are included with the component.  Additional motifs can be added by the 
user.
</p><p>The <b>Promoter</b> component will also display the results of hits found in the <b>Pattern Discovery</b> component.
</p>
<a name="JASPAR_CORE_database"></a><h1> <span class="mw-headline" id="JASPAR_CORE_database">JASPAR CORE database</span></h1>
<p>The promoter component of geWorkbench includes version 3.0 of the JASPAR CORE database (<a href="http://jaspar.genereg.net/" class="external free" rel="nofollow">http://jaspar.genereg.net/</a>).
  It contains 138 curated, non-redundant profiles.  These "profiles are 
derived from published collections of experimentally defined 
transcription factor binding sites for multi-cellular eukaryotes. The 
database represents a curated collection of target sequences"  (<a href="http://jaspar.genereg.net/html/TEMPLATES/help.html" class="external text" rel="nofollow">JASPAR Documentation</a>).
</p><p>The datafile used "MATRIX_DATA.txt" can be found at <a href="http://jaspar.genereg.net/html/DOWNLOAD/mySQL/JASPAR_CORE_2008/" class="external free" rel="nofollow">http://jaspar.genereg.net/html/DOWNLOAD/mySQL/JASPAR_CORE_2008/</a>.  
</p><p>The profiles represent counts of how many times each of the four 
nucleotide bases occurs at a particular position in the aligned promoter
 sequences.
</p>
<a name="Working_with_the_Promoter_graphical_interface"></a><h1> <span class="mw-headline" id="Working_with_the_Promoter_graphical_interface">Working with the Promoter graphical interface</span></h1>
<a name="Prerequisites"></a><h2> <span class="mw-headline" id="Prerequisites">Prerequisites</span></h2>
<ul><li> To use the Promoter component, first check that it has been loaded in the  Component Configuration Manager.
</li><li> The Promoter component appears when a data node of type "sequence" has been loaded in the Project Folders component.
</li><li> The Promoter component appears in the upper-right quadrant of geWorkbench, in the "Visual Area". 
</li></ul>
<a name="Layout"></a><h2> <span class="mw-headline" id="Layout">Layout</span></h2>
<p>The figure shows the display of profile "TBP:TATA-box:MA0108" from 
the list of those included in JASPAR.  The main features of the 
component include
</p>
<ul><li> The <b>TF Mapping</b> tab at left.  This area allows profiles to be searched for and selected for use in a sequence scan.
</li><li> The <b>Logo</b>, <b>Parameters</b> and <b>Sequence</b> tabs at right. These provide respectively visual display of the profile, the parameters, and the scan results.
</li><li> Control buttons at bottom left to manage scans and results.
</li></ul>
<p><img alt="Promoter Logo TBP.png" src="promoter_files/Promoter_Logo_TBP.png" height="479" width="818">
</p><p><br>
Further details of each part of the component are provided below.
</p>
<a name="TF_Mapping_tab"></a><h2> <span class="mw-headline" id="TF_Mapping_tab">TF Mapping tab</span></h2>
<a name="List_of_available_transcription_factors"></a><h3> <span class="mw-headline" id="List_of_available_transcription_factors">List of available transcription factors</span></h3>
<a name="TF_list"></a><h4> <span class="mw-headline" id="TF_list">TF list</span></h4>
<p>The TF list contains the names of available transcription factor 
binding site sequence profiles from JASPAR. Double-click on TF 
signatures in the TF list box (upper box) to add TF's to the Selected TF
 list (lower box). This lower list displays transcription factors which 
can be searched against the available genomic sequences by clicking on <b>Scan</b>. Double-clicking on a TF name clears it from the Selected TF list and returns it to the TF List. 
</p><p>The information displayed in the list contains the following fields from Jaspar, separated by colons:
</p><p>TF Name:class:Jaspar identifier
</p><p><br>
Multiple profiles can be moved to the selected list for scanning.
</p>
<a name="Search"></a><h4> <span class="mw-headline" id="Search">Search</span></h4>
<p>Searches for any portion of the text displayed in the list.  The 
search is progressive. As each character is entered, the list is updated
 to display only those entries containing a match to the current search 
string.
</p>
<a name="Taxa_pulldown_menu"></a><h4> <span class="mw-headline" id="Taxa_pulldown_menu">Taxa pulldown menu</span></h4>
<p>The list of displayed TFs can be restricted to a certain taxa.  Taxa assignments are obtained from the Jaspar motif files.
</p><p><br>
<img alt="Promoter Taxa Pulldown.png" src="promoter_files/Promoter_Taxa_Pulldown.png" height="201" width="228">
</p>
<a name="Selected_TF"></a><h3> <span class="mw-headline" id="Selected_TF">Selected TF</span></h3>
<p>This list contains the profiles that will be used in the next scan.  
Entries on the Selected TF list can be moved back up to the "available 
TFs" list by again double-clicking on their entry.
</p><p>Here the motif for TBP (class: Beta-sheet, family: TATA-binding) entry has been moved to the selected list.
</p><p><img alt="Promoter TBP selected.png" src="promoter_files/Promoter_TBP_selected.png" height="408" width="232">
</p>
<a name="Controls"></a><h3> <span class="mw-headline" id="Controls">Controls</span></h3>
<a name="Scan"></a><h4> <span class="mw-headline" id="Scan">Scan</span></h4>
<p>Scans the sequences in the Selected TF list against the available 
genomic sequences. If the Selected TF list is empty, the system displays
 an error message. 
</p>
<a name="Add_TF"></a><h4> <span class="mw-headline" id="Add_TF">Add TF</span></h4>
<p>Not currently available.  This button is disabled.
</p><p>The profile should consist of a tab delimited series of counts, 
one for each position in the profile.  It should could four lines, in 
the order A, C, G, T.  There are no header lines or row labels, just the
 numeric matrix.  For example, here is a profile showing the first six 
columns:
</p><p>0  12 0  0  1  0
</p><p>49 0  20 23 3  45
</p><p>0  37 29 2  45 4
</p><p>0  0  0  25 0  0
</p><p>Because the normalization step uses the count total (sequences 
aligned to generate the profile), loading a frequency matrix is not 
currently supported.
</p><p><br>
</p>
<a name="Save"></a><h4> <span class="mw-headline" id="Save">Save</span></h4>
<p>Saves to file a list of hits by a profile to a nucleotide sequence, 
including the sequence identifier, the transcription factor name and the
 start and stop points of the match along the sequence, as shown here:
</p><p>gi|65508003
</p><p>ATHB5:HOMEO-ZIP:MA0110 2104 2113 
</p><p>ATHB5:HOMEO-ZIP:MA0110 2115 2106 
</p><p>ATHB5:HOMEO-ZIP:MA0110 2882 2891 
</p><p>ATHB5:HOMEO-ZIP:MA0110 2893 2884 
</p><p><br>
</p>
<a name="Retrieve"></a><h4> <span class="mw-headline" id="Retrieve">Retrieve</span></h4>
<p>Not currently available. This button is disabled.
</p>
<a name="Stop"></a><h4> <span class="mw-headline" id="Stop">Stop<b></b></span></h4>
<p>Stop the current scan.
</p>
<a name="The_LOGO_tab"></a><h2> <span class="mw-headline" id="The_LOGO_tab">The LOGO tab</span></h2>
<a name="LOGO_display"></a><h3> <span class="mw-headline" id="LOGO_display">LOGO display</span></h3>
<p>The LOGO display implements the method of Schneider and Stephens 
(1990) to display the information at each position in a motif.  Briefly,
 the total height of the column of letters at a position shows the 
information available, on a scale of 0 to 2 bits (the information needed
 to represent the 4 possible nucleotide bases at each position).  The 
relative heights of each letter in a column show their individual 
contribution to the information at that position.
</p><p>The LOGO display in geWorkbench implements the "small sample 
correction" described by Schneider, the magnitude of which depends on 
the number of sequences aligned to generate the profile.  The correction
 is subtracted from the calculated information content at each position,
 with a minimum value (floor) of zero being displayed.
</p>
<a name="Table_display"></a><h3> <span class="mw-headline" id="Table_display">Table display</span></h3>
<p>A table is used to show the numeric data from which the LOGO diagram 
is generated.  The table depicts each position in the profile as a 
column, and has a row for each of the four nucleotide bases A, C, G and 
T.  The user can choose to display the data either as the original 
counts or as frequencies.
</p>
<a name="Display"></a><h4> <span class="mw-headline" id="Display">Display</span></h4>
<ul><li> Counts - Display the original counts of the four nucleotide bases at each position in the motif.
</li><li> Frequencies - Display the frequency of each base at each position in the motif.
</li></ul>
<a name="The_Parameters_tab"></a><h2> <span class="mw-headline" id="The_Parameters_tab">The Parameters tab</span></h2>
<p><img alt="T Promoter Parameters initial.png" src="promoter_files/T_Promoter_Parameters_initial.png" height="451" width="513">
</p>
<a name="Background_sequence_and_scoring_threshold_determination"></a><h3> <span class="mw-headline" id="Background_sequence_and_scoring_threshold_determination">Background sequence and scoring threshold determination</span></h3>
<p>A background sequence is used to estimate an appropriate scoring threshold.  This background can be generated in two ways.
</p>
<ol><li> determine base composition of input sequence and from this generate random sequence.
</li><li> (13K) - use a set of 13,000 promoter sequences as background.
</li></ol>
<ul><li> The length of background sequence scanned is given by 1000*Iterations/PValue.
</li></ul>
<ul><li> The threshold value is calculated by scanning the background 
sequence with the profile and finding the top 100 scores.  The 100th 
score is used as the threshold.
</li></ul>
<ul><li> Calculated p-values are Bonferroni corrected and also corrected for duplicates in the list of 100 top scores.
</li></ul>
<ul><li> Positive and negative strands are scanned and values above threshold are reported.
</li></ul>
<a name="Parameters"></a><h3> <span class="mw-headline" id="Parameters">Parameters</span></h3>
<ul><li> PValue / 1K - 
</li><li> Use Thr. - Use threshold - if checked, use a user-input threshold rather than a calculated threshold for scoring a match.
</li><li> 13K Set - If not checked (default), use the random background 
sequence described above.   If checked, use the 13K sequences as 
background. 
</li><li> Iterations - 
</li><li> Pseudocount -  a small-sample correction factor (default 1.0). See discussion below.
<ul><li> Sqrt(n) - if checked, set the pseudocount to be the square root of (n).  This option is not recommended, see discussion below.
</li></ul>
</li></ul>
<a name="Results"></a><h3> <span class="mw-headline" id="Results">Results</span></h3>
<a name="Total_hits_and_Sequences_with_hits"></a><h4> <span class="mw-headline" id="Total_hits_and_Sequences_with_hits">Total hits and Sequences with hits</span></h4>
<p>Total hits counts all hits regardless of how many times one sequence is hit. 
</p>
<ul><li> Expected - number of hits expected by chance.
</li><li> Actual - observed number of hits.
</li><li> Enrich. p-value - p-value for chance of getting this outcome by chance.
</li><li>&nbsp;% with hits
</li></ul>
<a name="5.27_hits_and_3.27hits"></a><h4> <span class="mw-headline" id="5.27_hits_and_3.27hits">5' hits and 3'hits</span></h4>
<ul><li> Expected - Expected number of hits
</li><li> Actual - Actual number of hits.
</li></ul>
<a name="The_Sequence_tab"></a><h2> <span class="mw-headline" id="The_Sequence_tab">The Sequence tab</span></h2>
<p><img alt="T Promoter Sequence tab with mouseover.png" src="promoter_files/T_Promoter_Sequence_tab_with_mouseover.png" height="304" width="708">
</p><p><br>
The Sequence tab can display either a line or a full character 
representation of the sequence which was searched against.  Clicking on a
 position along the line or character representation will cause that 
portion of the sequence to be displayed in the detail box at the bottom 
of the component.  This box also displays numbers representing position 
along the sequence, relative to the start of that particular sequence 
(not its genomic location).  Both the character and detail views will 
show the location and extent of any profile match to the sequence.
</p><p>If matches are found, the sequence will include blocks in various
 colors with solid arrows indicating the match orientation (forward or 
reverse complement). Individual hits can be identified by positioning 
the mouse pointer over them, which will display a tool tip. Clicking on 
an area with a match will show it in the Sequence Detail at the bottom 
with the hits shown as boxes around the characters.
</p><p>The tooltip format is as follows: numeric position, Transcription
 Factor name &lt;numeric position of the first character of the pattern,
 numeric position of last matching pattern character&gt;. 
</p>
<ul><li> <b>View</b> - <b>Line</b> or <b>Full Sequence</b>.  Line 
represents the sequence as a simple line, with any hits positioned along
 it.  Full shows the entire sequence as characters.
</li><li> <b>Show Patterns</b> - display hits from Pattern Discovery 
(this is a separate component, not part of the Promoter component).  
Implementation note - these hits are represented in the "Active 
Patterns" data-structure.
</li><li> <b>Show TFs</b> - show hits from a search in this component.  
Implementation note - these hits are represented in the "Active TFs" 
data-structure.
</li><li> <b>Clear All</b> - clear all hits from the sequence window 
(and from the associated data structures).  Note this will also clear 
the two adjacent check boxes.  The relevant box must be re-checked to 
see further results.
</li></ul>
<p><br>
The full character display shows any hits in white with a red 
background, and a small red arrow marks the start of the match and its 
direction.
</p><p><img alt="T Promoter CDH2 AP2 2000updn scan fullseq.png" src="promoter_files/T_Promoter_CDH2_AP2_2000updn_scan_fullseq.png" height="350" width="518">
</p><p><br>
</p>
<a name="Implementation_details"></a><h3> <span class="mw-headline" id="Implementation_details">Implementation details</span></h3>
<ol><li> Each time that a transcription factor (TF) matching operation 
is run, the "Active TFs" data structure is *AUGMENTED* with the results 
of the discovery operation (i.e., contents due to previous runs are 
maintained). The "Active TFs" data structure is not affected by pattern 
discovery.
</li><li> Each time that a Pattern Discovery analysis is run, the 
contents of the "active patterns" structure are *REPLACED* with the 
results of the discovery operation (i.e., contents due to previous runs 
are cleared).
</li></ol>
<a name="Scan_Implementation"></a><h1> <span class="mw-headline" id="Scan_Implementation">Scan Implementation</span></h1>
<a name="Normalization_and_the_Pseudocount"></a><h2> <span class="mw-headline" id="Normalization_and_the_Pseudocount">Normalization and the Pseudocount</span></h2>
<p>The count matrices are normalized to frequencies using an algorithm which includes a "<b>pseudocount</b>" (<a href="http://www.ncbi.nlm.nih.gov/pubmed/19106141" class="external text" rel="nofollow">see Nishida 2009</a>).
  The pseudocount is a way to compensate for the effects of small sample
 sizes in the original observations used to generate the profiles.  
Nishida et al. studied how to determine an appropriate value for the 
pseudocount.  They found that the optimal values were independent of the
 sample size and were correlated with the entropy of the original 
matrices.  They say that this implies that the less-conserved the 
binding site, the larger a value should be used for the pseudocount.  
They find that 0.8 is a good value "for practical uses".  They do not 
recommend use of the square root of the total count.
</p><p>geWorkbench allows a pseudocount factor to be directly entered, 
or it can be selected to be the square root of the total count of 
sequences used to generate the profile.  Prior to geWorkbench 1.8.0, 
setting the pseudocount to the square root of the total counts was 
directly coded and not changeable.  The current default is to set the 
pseudocount equal to 1.0.
</p><p>The normalization forumula used in calculating frequencies is 
then, where b is the pseudocount, and counts(i, j) is the observed count
 in a particular entry in the matrix,
</p><p>freq(i, j) = (counts(i, j) + b/4) / (totalCounts + b).
</p><p><br>
The resulting frequency matrix is used in the subsequent scan.
</p><p>Because the pseudocount is a settable parameter, the frequency matrix is recalculated for each scan from the original counts.
</p>
<a name="Scoring"></a><h2> <span class="mw-headline" id="Scoring">Scoring</span></h2>
<ul><li> Calculated p-values are Bonferroni corrected and also corrected
 for duplicates in the list of 100 top scores found during the 
background scan.
</li></ul>
<ul><li> Positive and negative strands are scanned and values above threshold are reported.
</li></ul>
<a name="Example:_Running_and_viewing_a_scan"></a><h1> <span class="mw-headline" id="Example:_Running_and_viewing_a_scan">Example: Running and viewing a scan</span></h1>
<a name="Prerequisites_2"></a><h2> <span class="mw-headline" id="Prerequisites_2">Prerequisites</span></h2>
<p>The Promoter component is only available when a sequence has been 
loaded, either from disk or for example using the Sequence Retriever 
component to obtain genomic sequence.
</p><p>For this example, we will obtain upstream genomic sequence for 
CDH2, the N-Cadherin gene.  However, using the Sequence Retriever 
component requires that a microarray dataset and its annotations be 
loaded.  Here, we will use the JB-ccmp_0120.txt file, which is an 
Affymetrix HG-U95Av2 MAS5 format text file and is part of the 
geWorkbench <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Download#Tutorial_data" class="external text" rel="nofollow">tutorial dataset</a>.  
</p><p>The Affymetrix HG-U95Av2 annotation file can be obtain from the Affymetrix website.  Please see instructions on the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/FAQ#Q._Where_can_I_obtain_the_latest_annotation_files_for_my_microarray_platform.3F" class="external text" rel="nofollow">geWorkbench FAQ</a>.
</p><p>1. In the Project Folders component, load the file JB-ccmp_0120.txt as type  MAS5/GCOS.
</p><p>2. When prompted, associate the HG-U95Av2 annotation file.  
</p><p>3. In the Markers component, search for the gene name "CDH2" using the <b>Find Next</b> button.  On this chip type, marker 2053_at represents the CDH2 gene.  
</p><p>4. You can double-click on the marker to add it to the default 
"Selection" set.  Or you can right-click on it and add it to a named 
set, such as "Cadherins".  This is depicted below.
</p><p>5. "Activate" (check the box next to) the set to which you added the marker.
</p><p><img alt="T SequenceRetriever Markers CDH2.png" src="promoter_files/T_SequenceRetriever_Markers_CDH2.png" height="414" width="229">
</p><p>6. Any activated markers will appear in the Sequence Retriever component, as shown below.
</p><p>7. Set the retrieval limits to 2000 base pairs up- and downstream from the transcription start site.
</p><p>8. Make sure the retrieval type is set to DNA, UCSC (the Santa Cruz genome sequence database).
</p><p>9. Click "<b>Get Sequence</b>".
</p><p><img alt="T SequenceRetriever CDH2 2000updn.png" src="promoter_files/T_SequenceRetriever_CDH2_2000updn.png" height="373" width="878">
</p><p>10. Once the sequence has been retrieved, check the box next to 
the sequence and then hit the button "Add to Project".  We now have the 
genomic sequence available for other components to use.
</p><p><br>
<img alt="T SequenceRetriever CDH2 seq in proj folder-2.png" src="promoter_files/T_SequenceRetriever_CDH2_seq_in_proj_folder-2.png" height="166" width="233">
</p>
<a name="Running_the_scan"></a><h2> <span class="mw-headline" id="Running_the_scan">Running the scan</span></h2>
<p>1. In the Promoter component, search for "ap2", which corresponds to transcription factor activator protein 2 alpha.
</p><p>2. Double-click on the TFAP2A entry to move it down to the search list.
</p><p><img alt="T Promoter CDH2 AP2 2000updn setup logo.png" src="promoter_files/T_Promoter_CDH2_AP2_2000updn_setup_logo.png" height="413" width="772">
</p><p>3. Hit the "<b>Scan</b>" button.  The result is displayed in the <b>Sequence</b> tab of the Promoter component as shown here.
</p><p><img alt="T Promoter CDH2 AP2 2000updn scan.png" src="promoter_files/T_Promoter_CDH2_AP2_2000updn_scan.png" height="408" width="772">
</p><p>4. Setting the View to Full Sequence shows the hits in white on 
the sequence. Red arrows indicate whether the hit is to the forward 
(right arrow) or reverse (complementary) (left-arrow) strand.
</p><p><img alt="T Promoter CDH2 AP2 2000updn scan fullseq.png" src="promoter_files/T_Promoter_CDH2_AP2_2000updn_scan_fullseq.png" height="350" width="518">
</p><p>5. The parameters tab displays the actual threshold values calculated during the run, and displays the enrichment results.
</p><p><br>
<img alt="T Promoter CDH2 AP2 2000updn scan finalparams.png" src="promoter_files/T_Promoter_CDH2_AP2_2000updn_scan_finalparams.png" height="349" width="517">
</p>
<a name="References"></a><h1> <span class="mw-headline" id="References">References</span></h1>
<ul><li> Lawrence and Reilly (1990) Searching putative regulatory 
sequences against a collection of known transcription factor DNA-binding
 signatures represented as a position weight matrices (PWMs) (citation 
unknown).  See perhaps: Lawrence, C. and Reilly, A. (1990). An 
expectation maximization (EM) algorithm for the identification and 
characterization of common sites in unaligned biopolymer sequences. 
Proteins, 7 (1), 41-51. <a href="http://www.ncbi.nlm.nih.gov/pubmed/2184437" class="external text" rel="nofollow">Link to Abstract</a>
</li><li> Nishida K, Frith MC, Nakai K. (2009) Pseudocounts for 
transcription factor binding sites.  Nucleic Acids Res. 
Feb;37(3):939-44.  <a href="http://www.ncbi.nlm.nih.gov/pubmed/19106141" class="external text" rel="nofollow">link to paper</a>
</li><li> Sandelin A, Alkema W, Engstrom P, Wasserman WW, Lenhard B.  
(2004) JASPAR: an open-access database for eukaryotic transcription 
factor binding profiles. Nucleic Acids Res.  Jan 1;32(Database 
issue):D91-4 (<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=14681366&amp;query_hl=6" class="external text" rel="nofollow">link to paper</a>).
</li><li> Schneider TD, Stephens RM. (1990) Sequence logos: a new way to
 display consensus sequences. Nucleic Acids Res. Oct 25;18(20):6097-100.
 (<a href="http://www.ncbi.nlm.nih.gov/pubmed/2172928" class="external text" rel="nofollow">link to paper</a>)
</li><li> Vlieghe D, Sandelin A, De Bleser PJ, Vleminckx K, Wasserman 
WW, van Roy F, Lenhard B. (2006)  A new generation of JASPAR, the 
open-access repository for transcription factor binding site profiles.  
Nucleic Acids Res.  Jan 1;34(Database issue):D95-7.
</li></ul>

<!-- 
NewPP limit report
Preprocessor node count: 146/1000000
Post-expand include size: 1889/2097152 bytes
Template argument size: 0/2097152 bytes
Expensive parser function count: 0/100
-->

<!-- Saved in parser cache with key workbench:pcache:idhash:1516-0!1!0!!en!2!printable=1 and timestamp 20110729024726 -->
<div class="printfooter">
Retrieved from "<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Promoter_Analysis">http://wiki.c2b2.columbia.edu/workbench/index.php/Promoter_Analysis</a>"</div>
		<div id="catlinks" class="catlinks catlinks-allhidden"></div>		<!-- end content -->
				<div class="visualClear"></div>
	</div>
</div></div>
<div class="visualClear"></div>
<div id="footer">

This page was last modified on 18 February 2011, at 15:21.

</div>
</div>

<!-- Served in 0.421 secs. -->
</body></html>