<html>



<head>

<meta http-equiv="Content-Language" content="en-us">

<meta http-equiv="Content-Type" content="text/html; charset=windows-1252">

<meta name="GENERATOR" content="Microsoft FrontPage 4.0">

<meta name="ProgId" content="FrontPage.Editor.Document">

<title>User Comments</title>

<link rel="stylesheet" type="text/css" href="ncicb.css">

</head>



<body>
    <h3>Promoter Panel</h3>  
    <p>The Promoter Panel is located in the View Area (upper right) of the application. The promoter panel can used to visualize transcription factor binding sites (TFBS) using Transcription Factor signature analysis: Searching putative regulatory sequences against a collection of known transcription factor DNA-binding signatures represented as a position weight matrices (PWMs) [Lawrence and Reilly, 1990]. This visualization panel uses the JASPAR Transcription Factor Binding Profile Database (<a href="http://jaspar.cgb.ki.se/cgi-bin/jaspar_db.pl">http://jaspar.cgb.ki.se/cgi-bin/jaspar_db.pl</a> ).</p> 
    <p>This help section describes the promoter panel <a href="#_Basic_Usage">basic usage</a>, <a href="#_View_Transcription_Factor_Analysis"> transcription factor analysis</a>  and <a href="#_View_Pattern_Discovery_Results">viewing results from the Pattern Discovery analysis component</a>.</p> 
    <h3><a name="_Sequence_Tab"></a> Sequence Tab</h3> 
    <h3> <a name="OLE_LINK2"></a><a name="OLE_LINK1"></a><a name="_Basic_Usage"></a><img border="0" width="918" height="435" src="promoterpanel_files/image001.gif"></h3> 
    <h3> Basic Usage </h3> 
    <p>The following describes the basic usage of the Promoter Panel:</p> 
    <ul> 
    <li><b>TF list:</b> The TF list contains the names of available transcription factors binding site sequence patterns from JASPAR or those that have been loaded from files using <b>Add TF</b>.  Double click on TF signatures in the TF list box (upper box) to add TF's to the Selected TF (lower box). This list box displays transcription factors which can be searched against the available genomic sequences by clicking on <b>Scan.</b> Double-clicking on a TF name clears it from the Selected TF list and returns it to the <b>TF List</b>. </li> 
    <li><b>Search: </b> Enter a portion of a TF name in the Search text box. The list scrolls to and highlights the TF containing the text string. Click <b>Find Next </b>to highlight the TF containing the text string. If nothing matches the text string, the entered text is changed to red.</li> 
    <li><b>Scan:</b> Scans the sequences in the <b>Selected TF</b> list against the available genomic sequences. If the Selected TF list is empty, the system displays an error message. </li> 
    <li><b>Save:</b> User can select where to stores a file in a comma or tab separated format with 3 columns with the Transcription Factor name, the name of the underlying sequence and the first matching position in the display area. </li> 
    <li><b>Stop:</b> Stops the scanning process. </li> 
    <li><b>Retrieve:</b> Not implemented. </li> 
    <li><b>Add TF:</b> Loads a file containing new transcription factors to add to the TF Listing. Users shall be able to specify an alternative JASPAR-formatted data file to load transcription factors from by selecting &quot;Add TF&quot;. Promoters added through the &#8220;Add TF&#8221; button persist across application invocations.</li> 
    <li><b>Sequence Details: </b>The <b><a href="#_Sequence_Tab">Sequence Tab </a></b> displays character strings of the sequence string selected.  Users single click on a sequence line or sequence string and Seqeunce Details located in the bottom pane is updated. The display of sequence details is proportional to the click position in the line. If you click at the beginning of the line, the bottom will show the first part of the sequence. If you click the center place of a line, the Sequence display will show the middle part of whole sequence.</li> 
    <li><b>Progress Bar:</b> The progress bar provides a visual update on the analysis status. </li> 
    <li><b>Clear All:</b> Clears the display area of active patterns and active TFs. This is useful mostly for resetting the &quot;active TFs&quot; data structure which is not automatically cleared before each <b>Scan </b> operation. </li> 
    <li><b>Show Patterns: </b> Checkbox that controls if the patterns currently in the &quot;active patterns&quot; structure are displayed (as hollow rectangle) over the sequences. When selected (checked), the patterns are displayed. Unselected, the patterns are not displayed. </li> 
    <li><b>Show TF:</b>  Controls the display of TFs currently in the &quot;Active TF&quot; (as solid rectangles in various colors) over the sequences. When selected (checked), the display area will include TF&#8217;s in the sequence display. Unselected, the TF&#8217;s are not displayed. </li> 
    <li><b>All Sequence:</b> Checkbox that limits the display to sequences in an activated Marker group.</li> 
    <h3>Tabs:</h3>  
    <p>o <b><a href="#_Logo">Logo Tab:</a> </b>Select a TF in either list and the Logo tab displays a Position Specific Scoring Matrix (PSSM), a sequence logo representation of the TF signature. </p> 
    <p>o <b><a href="#_Parameter_Tab">Parameters</a>: </b>Captures parameters and results for the Promoter Panel.</p> 
    <p>o <b><a href="#_Sequence_Tab">Sequence Tab:</a> </b>The lines in the display area represent each sequence. Double click on the sequences to see the textual representation of the DNA and the TF signature match, shown as a colored rectangle.</p> 
    <h3><a name="_View_Transcription_Factor_Analysis"></a><a name="_Logo"></a>Logo</h3> 
    <p><img border="0" width="611" height="304" src="promoterpanel_files/image002.jpg"></p> 
    <h3>View Transcription Factor Analysis</h3> 
    <p>1. Load a DNA Sequence FASTA File.</p> 
    <p>2. Select Promoter in the Visualization Area. The display area is updated to reflect the sequences for the FASTA file loaded.</p> 
    <p>3. Click on TF signatures in the TF list to the Selected TF list. </p> 
    <p>4. Click on <b>Scan</b>, the Selected TF's are searched against the available genomic sequences. The patterns will be mapped onto the sequences according to match criteria in the <a href="#_Parameter_Tab">Parameter Tab. </a> </p>
    <p> Note: Each time that a transcription factor (TF) matching operation is run, the &quot;active TFs&quot; data structure is *AUGMENTED* with the results of the discovery operation (i.e., contents due to previous runs are maintained). The &quot;active TFs&quot; is not affected by pattern discovery.</p> 
    <p>5. Click on <b>Show TF</b>. If matches are found, the sequence will include blocks in various colors with solid arrows indicating its orientation (negative or positive). Individual hits can be identified by positioning the mouse pointer over them, which will display the tool tip. Double click on a match and the Sequence Detail at the bottom of the pane displays character strings of the selected.</p> 
    <p><img border="0" width="515" height="238" src="promoterpanel_files/image003.gif"></p> 
    <p><i>The tooltip format is as follows: numeric position, Transcription Factor name (numeric position of the first character of the pattern, numeric position of last matching pattern character). </i></p> 
    <h3><a name="_Parameter_Tab"></a>Parameter Tab</h3> 
    <p><b><img border="0" width="672" height="458" src="promoterpanel_files/image004.png"></b></p> 
    <p><a name="OLE_LINK4"></a><a name="OLE_LINK3"></p> 
    <h3><a name="_View_Pattern_Discovery_Results"></a>View Pattern Discovery Results </h3> 
    <p>1. Load a FASTA File into the project pane.</p> 
    <p>2. Run Pattern Discovery analysis. (See Pattern Discovery online help for more information)</p> 
    <p>3. In the Pattern Discovery Panel, left click on the row(s) to display selected motifs and their frequencies in the dataset. The Promoter Panel, Sequence Panel, and Patterns are updated to display the patterns selected. </p> 
    <p>Note: Each time that a Pattern Discovery analysis is run, the contents of the &quot;active patterns&quot; structure are REPLACED with the results of the discovery operation (i.e., contents due to previous runs are cleared). </p> 
   

</body>



</html>
