<html>



<head>

<meta http-equiv="Content-Language" content="en-us">

<meta http-equiv="Content-Type" content="text/html; charset=windows-1252">

<meta name="GENERATOR" content="Microsoft FrontPage 4.0">

<meta name="ProgId" content="FrontPage.Editor.Document">

<title>User Comments</title>

<link rel="stylesheet" type="text/css" href="ncicb.css">

</head>



<body>
    <h3>Pattern Discovery</h3> 
    <p>The Pattern Discovery is located in the <b>Selection Area</b> in the lower left of the application. This component runs SPLASH, an efficient application that identifies sparse amino acid or nucleic acid patterns (called &quot;motifs&quot;) that match identically or similarly in a set of protein or DNA sequences. For a full description of SPLASH, go to <a href="http://sharepoint.cu-genome.org/exchweb/bin/redir.asp%3fURL=http:/www.research.ibm.com/splash/index.htm">http://www.research.ibm.com/splash/index.htm.</a> </p> 
    <h3>SPLASH is hosted on a remote server. The caWorkbench component communicates with the server via a SOAP interface. Search parameters and protein/DNA sequences in FASTA file format are uploaded to the server via SOAP. SPLASH is run and the matching sequences are retrieved and displayed in the selection area. This section describes the<a href="#_Basic_Usage"> basic usage</a> of the pattern discovery panel, <a href="#_Modes_of_Pattern_Discovery">modes of pattern discovery</a> and <a href="#_Pattern_Discovery_Results">Pattern Discovery Results</a>.</h3> 
    <p><a name="_Basic_Usage"></a><img border="0" width="628" height="245" src="DiscoveryOverviewed_files/image001.jpg"></p> 
    <p><b>Basic Usage</b></p>
    <p>The steps of basic usage are indicated in the screenshot above.</p> 
    <p>A. The modes of Pattern Discovery are:<a href="#_Hierarchical_and_Exhaustive_Discove"> Normal Discovery</a>,<a href="#_Hierarchical_Discovery"> Hierarchical Discovery</a> and<a href="#_Exhaustive_Discovery"> Exhaustive Discovery</a>. Across the top of the Pattern Discovery panel are three radio buttons (Norm., Hierarch., and Exhaust.) that are used to select the discovery mode. </p> 
    <p>B. The default &quot;Basic&quot; parameters are shown. The Parameters help section includes a full description of the Pattern Discovery parameters.</p> 
    <p>C. Specifies use of a Globus (<a href="http://www.globus.org/toolkit/">http://www.globus.org/toolkit/</a> ) instance of SPLASH for Pattern discovery. This allows for discovery of SPLASH instances running on the internet in accordance with the Globus Alliance&#8217;s requirements (server discovery not fully implemented). If unselected, a GSOAP (<a href="http://gsoap2.sourceforge.net/">http://gsoap2.sourceforge.net/</a> ) implementation of a SPLASH web service hosted at Columbia University is used.</p> 
    <p>D. Clicking on the clockwise blue arrow will <b>Start Pattern Discovery</b>. This brings up a <b>New Session</b> dialog box. The first time Pattern Discovery is run, the dialog box must be filled out. Use the drop down menu to select the Server Type and Server that hosts SPLASH. Then type-in the port and User Name. The image below is a connection to a server running on the local host at port 80.</p> 
    <p> Available SOAP SPLASH server(s) are at <u>splash.cu-genome.org</u> at port 80. Please select &#8216;splash&#8217; as the server type.</p> 
    <p><img border="0" width="461" height="256" src="DiscoveryOverviewed_files/image002.jpg"></p> 
    <p>E. Clicking on the stop sign icon will<b> Stop Pattern Discovery.</b></p> 
    <p>F. Clicking on the folder icon will <b>Load a Pattern File</b> motif files (.pat).</p> 
    <p>G. The progress bar used to provide a visual update of the percentage completion.</p> 
    <p><b>Pattern Discovery Results</b></p> 
    <p>Shown below are the results of a discovery using default parameters. The Discovery Table is filled with information on the discovered motifs. The steps of basic usage are indicated with letter in the below screenshot.</p> 
    <p><img border="0" width="701" height="433" src="DiscoveryOverviewed_files/image003.jpg"></p> 
    <p>A. The number of patterns found is displayed below the discovery table.</p> 
    <p>B. Right clicking on a pattern displays the option to:</p> 
    <p>&#10170; Mask pattern- suppresses the selected pattern from the set of patterns discovered. So the sequence panel for instance will not show a masked pattern match in one of its modes of operation and will exist in any ancillary dataset created (see below).</p> 
    <p>&#10170; Unmask all patterns- causes all patterns discovered to be &#8220;active&#8221; and these will be shown in any ancillary dataset created. </p> 
    <p>&#10170; Save patterns- caves the regular expression corresponding to the selected pattern to file.</p> 
    <p>&#10170; Save selected patterns- save the selected patterns to file along with match coordinates on sequences on which the pattern discovery was performed.</p> 
    <p>&#10170; Save All patterns- save all patterns to file along with match coordinates.</p> 
    <p>&#10170; Add pattern to project- creates an ancillary dataset in the Project Panel under the sequence dataset used for pattern discovery.</p> 
    <p>C. To display selected motifs and their frequencies in the dataset, left click on a pattern(s). The selected pattern(s) are then shaded gray in the discovery table and are displayed graphically in the Sequence Panel, Promoter Panel and Patterns.</p> 
    <p><b><a name="_Modes_of_Pattern_Discovery"></a>Modes of Pattern Discovery</b></p>
    <p><b>Normal Discovery</h3> 
    <p>Normal discovery is based on the algorithm SPLASH (Califano, A., 2000); it generates a list of all regular expression patterns (motifs) that satisfy a user-defined minimum support and a minimum density criteria.</p> <b> </b> 
    <p><b><a name="_Hierarchical_Discovery"></a>Hierarchical Discovery</b></p>
    <p>Hierarchical discovery is used to analyze a sequence database and produce a sequence of clusters based on discovered patterns. For more information, visit <a href="http://sharepoint.cu-genome.org/exchweb/bin/redir.asp%3fURL=http:/www.research.ibm.com/splash/Hyerarchical/HierarchicalDiscovery.htm">http://www.research.ibm.com/splash/Hyerarchical/HierarchicalDiscovery.htm</a>. </p> 
    <p><b><a name="_Exhaustive_Discovery"></a>Exhaustive Discovery</b></p> 
    <p>Exhaustive discovery can be used to exhaustively analyze a sequence database for all non-overlapping motifs that are statistically significant. This is useful for discovering in order of relative sequence support, all regions of a protein family that have been preserved by evolution and may therefore play a functional or structural role. For more information, visit <a href="http://sharepoint.cu-genome.org/exchweb/bin/redir.asp%3fURL=http:/www.research.ibm.com/splash/Exhaustive/ExhaustiveDiscovery.htm">http://www.research.ibm.com/splash/Exhaustive/ExhaustiveDiscovery.htm</a>. </p>
    <p>4 Select a <b>Database </b>from the dropdown list.</p>
    <p>5 Modify the values of <b>subsequence</b> if needed</p>
    <p>6 Select an Algorithm from the dropdown list. The includes:</p>
    <p>&#12290; Smith-Waterman Protein: compares an amino acid query sequence against a protein sequence database </p>
    <p>&#12290; Smith-Waterman DNA: compares a nucleotide query sequence against a nucleotide sequence database </p>
    <p>&#12290; Frame(from DNA sequence to protein DB): compares a nucleotide query sequence against a protein sequence database </p>
    <p>&#12290; Frame(from Protein sequence to DNA DB): compares an amino acid query sequence against a nucleotide sequence database</p>
    <p>7 Click on <b>Search</b>. </p>
    <p>&#12290; The result will be showed on the Alignment Results visualization component.</p>
    <p><img border="0" width="493" height="440" src="BlastOverview_files/image003.jpg"></p>
    <p><a name="_To_Add_Sequences_to_a_Project"></a> </p>
</body>
</html>

