<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html dir="ltr" lang="en"><head>
<title>Sequence Retriever - Workbench</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="generator" content="MediaWiki 1.16.0">
<meta name="robots" content="noindex,follow">

</head>
<body class="mediawiki ltr ns-0 ns-subject page-Sequence_Retriever skin-monobook">
<div id="globalWrapper">
<div id="column-content"><div id="content">
	<a id="top"></a>
	
	<h1 id="firstHeading" class="firstHeading">Sequence Retriever</h1>
	<div id="bodyContent">
		<!-- start content -->
<p><br>
</p>
<p><br>
</p>
<a name="Overview"></a><h2> <span class="mw-headline" id="Overview">Overview</span></h2>
<p>The Sequence Retriever component fetches the DNA or protein sequences
 for selected gene markers from remote databases.  Nucleotide sequences 
are retrieved from the GoldenPath database hosted at the UC Santa Cruz. 
 Amino-acid sequences are retrieved from the European Bioinformatics 
Institute (EBI).
</p><p>As of geWorkbench release 2.2.1, DNA sequences are retrieved 
using the transcript start site obtained for the desired gene from the 
refGenes table at UCSC. This table lists refSeq genes available for each
 supported species.  Previously, geWorkbench used the KnownGenes table, 
which supported only human, mouse and rat. 
</p><p>The Sequence Retriever component functions only in the context of
 microarray datasets. The component also depends on the microarray 
platform annotation file, which must be loaded when the microarray 
dataset is read into geWorkbench.  
</p><p>For DNA queries, the RefSeq gene(s) associated with a given 
probeset are obtained from the annotation file and used for the query.  
For protein queries, UniProt IDs from the annotation file are used.  If 
multiple IDs are present for a marker, they are all used.
</p><p>One the desired sequences have been downloaded, they can be 
viewed or saved to the Project Folders component for use by other 
components. Examples include running <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/BLAST" title="BLAST">BLAST</a> searches, looking for the presence of known promoter motifs (<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Promoter_Analysis" title="Promoter Analysis"> Promoter Analysis</a>), or looking for patterns in a set of sequences using <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Pattern_Discovery" title="Pattern Discovery">Pattern Discovery</a>.
</p>
<a name="Controls"></a><h2> <span class="mw-headline" id="Controls">Controls</span></h2>
<ul><li> <b>Type</b> - DNA or Protein
<ul><li> Source - The pulldown menu to the right of the "Type" pulldown 
shows the source to query for the data.  At present, only one choice for
 each type is supported. DNA sequences are retrieved from the UCSC Santa
 Cruz using RefSeq IDs.  Protein sequences are retrieve from the EBI 
using UniProt IDs.
</li></ul>
</li></ul>
<ul><li> <b>Marker</b> - This panel shows markers that are in any activated set in the Markers component.
</li><li> <b>Find a Marker</b> - This tab allows one to search in the 
component for a particular marker.  However, we recommend performing 
searches in the Markers component and adding the results to a set.
</li></ul>
<ul><li> <b>"- and +"</b> text fields - For DNA sequence retrieval, 
these two text fields specify the distance upstream (-) and downstream 
(+) from the transcript start site for the request.  They are disabled 
when a protein query is selected.  The default setting is 2000 bp 
upstream and 1000 bp downstream.
</li></ul>
<ul><li> <b>Stop</b> - Stop the current sequence download.  Only available while a download is in progress.
</li><li> <b>Clear</b> - clear the query results display.
</li></ul>
<ul><li> <b>Get Sequence</b> - Launches the query using the selected markers.
</li></ul>
<ul><li> <b>Add to Project</b> - Add selected sequences returned by the query to the Project Folders component.
</li></ul>
<ul><li> <b>View</b> - Line or Full Sequence. 
</li></ul>
<ul><li> <b>Show only unique transcript-start sites</b> - Applies to DNA
 queries only.  When checked, only a single sequence entry for a given 
transcript start site will be shown.  If the box is unchecked, all 
RefSeq sequences associated with the probeset in the annotation file, 
and for which sequences were retrieved, will be shown.
</li></ul>
<ul><li> <b>Include</b> - Check boxes for sequences to copy to the Project Folders component when "Add to Project" is pushed.
</li></ul>
<ul><li> <b>Name</b>  -  The name for a sequence is formed by 
concatenating the marker name with the chromosome number of physical 
location of the sequence on the chromosome, e.g. in 
35694_at_chr2_102314487, "35694_at" is the probeset id, and the sequence
 is located on chromosome 2 at position 102314487.
</li></ul>
<ul><li> <b>Sequence Detail</b> - graphically depicts the upstream 
(blue)and downstream (red) sequences areas, and marks the position of 
the transcript start site and the orientation of the transcript on the 
genome build.
<ul><li> right-arrowhead - transcript mapped on genome plus strand.
</li><li> left-arrowhead - transcript mapped on genome minus strand.
</li></ul>
</li></ul>
<ul><li> <b>Detailed sequence display box</b>  - At the bottom of the component, clicking on a sequence in the display above will show the letters with a position scale.
</li></ul>
<p><br>
<img alt="Sequence Retriever MAP4K4 DNA result.png" src="sequenceRetriever_files/Sequence_Retriever_MAP4K4_DNA_result.png" height="406" width="881">
</p>
<a name="Using_the_Sequence_Retriever"></a><h2> <span class="mw-headline" id="Using_the_Sequence_Retriever">Using the Sequence Retriever</span></h2>
<p>The Sequence Retriever displays a list of the contents of any marker 
set or sets that are "activated" in the Markers component.  "Activated" 
means that the check-box next to the set's name in the Markers component
 is checked. These are the markers that will be used in a query to a 
remote data source.
</p><p>The Markers component provides a dynamic search feature with which the user can locate desired markers and add them to a set.
</p><p>DNA or protein sequences can be requested. 
</p><p>For DNA sequences, the Sequence Retriever retrieves genomic 
sequence from the Golden Path database at the UC Santa Cruz.  The amount
 of sequence upstream and downstream to fetch can be specified in the 
range settings (-+).  
</p><p>Once sequences have been retrieved, some or all can be added to 
the Project Folders component as a new data node (Add to Project 
button).
</p>
<a name="Prerequisites"></a><h3> <span class="mw-headline" id="Prerequisites">Prerequisites</span></h3>
<ul><li> A microarray dataset must be loaded.
</li><li> An annotation file must be associated with the microarray 
dataset at the time it is loaded.  At present, only Affymetrix-format 
annotation files can be read in.  These files can be obtained for 
Affymetrix chip types from affymetrix.com.  For exact instructions, 
please see the geWorkbench FAQ page: <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/FAQ#Q._Where_can_I_obtain_the_latest_annotation_files_for_my_microarray_platform.3F" class="external text" rel="nofollow">FAQ</a>
</li><li> A set in the Markers component containing one or more markers of interest must be activated.
</li></ul>
<ul><li> There must be annotation for each marker being queried on.  If 
not, the user will be warned when a marker without annotation is 
encountered.
</li></ul>
<p>Here we show searching for MAP4K4 in the Markers component and adding
 it to the default "Selection" set.  When this set is activated, the 
marker will appear in the Sequence Retriever component.
</p><p><br>
<img alt="Markers MAP4K4.png" src="sequenceRetriever_files/Markers_MAP4K4.png" height="446" width="288">
</p>
<a name="Retrieving_Sequences"></a><h3> <span class="mw-headline" id="Retrieving_Sequences">Retrieving Sequences</span></h3>
<ul><li> Type - Select DNA or Protein sequences
</li><li> Range - If downloading genomic DNA, check/set the range 
upstream and downstream of the transcript start site for which to 
retrieve sequence.
<ul><li> The default range is from -2000 (upstream) +1000 (downstream) relative to the start of the transcript.  
</li></ul>
</li><li> Hit <b>Get Sequence</b>.
</li><li> <b>Select Genome version</b> - If the query is for DNA, the 
user is prompted to select for the species and version of the genome 
build to query.  You should choose the species corresponding to the 
microarray chip of the dataset you have loaded.  The list of available 
genome builds is obtained automatically from Santa Cruz.
</li></ul>
<p><br>
<img alt="Sequence Retriever Select Genome.png" src="sequenceRetriever_files/Sequence_Retriever_Select_Genome.png" height="375" width="377">
</p><p><br>
The retrieved sequences are depicted in the viewer area on the 
right-hand side of the component.  All sequences associated with a given
 gene symbol are retrieved.  
</p><p><br>
</p>
<a name="DNA_Sequences"></a><h4> <span class="mw-headline" id="DNA_Sequences">DNA Sequences</span></h4>
<p>For the MAP4K4 marker, genomic sequence for two unique transcript start sites was retrieved.  
</p><p><b>DNA Sequence Naming</b> - Each retrieved sequence is given a 
name formed by concatenating the probeset name for the marker, the 
refSeq identifier, the chromosome, and the location on the chromosome of
 the transcript start site obtained from the UCSC refGene table.
</p><p>Note that sequence is only retrieved for the particular RefSeq 
IDs associated with the probeset in the annotation file.  The results 
may differ from what would be retrieved from UCSC using the gene symbol 
directly.  For example, a direct query to UCSC in this example would 
return 5 sequences, not just the 2 shown here for this probeset.
</p><p><br>
<img alt="Sequence Retriever MAP4K4 DNA result.png" src="sequenceRetriever_files/Sequence_Retriever_MAP4K4_DNA_result.png" height="406" width="881">
</p><p><br>
Unchecking the "Show only unique transcript-start sites" adds one 
additional result.  The first two sequences are now identical, though 
obtained through different RefSeq IDs:
</p><p><br>
<img alt="Sequence Retriever MAP4K4 DNA result unrestricted.png" src="sequenceRetriever_files/Sequence_Retriever_MAP4K4_DNA_result_unrestricted.png" height="185" width="613">
</p>
<a name="Protein_Sequences"></a><h4> <span class="mw-headline" id="Protein_Sequences">Protein Sequences</span></h4>
<p>For the MAP4K4 marker, six sequences were returned.  Note the differences in length as shown graphically.
</p><p><br>
<b>Protein Sequence Naming</b> - For protein results, the sequence name is formed by concatenating the probeset id with the UniProt ID of the sequence.
</p><p><br>
<img alt="Sequence Retriever MAP4K4 Protein Result.png" src="sequenceRetriever_files/Sequence_Retriever_MAP4K4_Protein_Result.png" height="471" width="927">
</p>
<a name="Viewing_Sequences"></a><h3> <span class="mw-headline" id="Viewing_Sequences">Viewing Sequences</span></h3>
<p>The graphic sequence viewer pane of the Sequence Retriever component 
has two viewing modes - "Line" and "Full Sequence", controlled by the 
"View" pulldown menu.
</p>
<a name="Line_View"></a><h4> <span class="mw-headline" id="Line_View">Line View</span></h4>
<p>When the view mode is set to "Line", each sequence is represented by a line in the graphic.  
</p><p>Line details for DNA Sequences:
</p>
<ul><li> <b>Blue</b> - for sequence upstream of transcript start site, the line is blue.
</li><li> <b>Red</b> - for sequence downstream from transcript start site, the line is red.
</li><li> <b>Blue arrowhead</b> - marks the position of the transcript start site and the orientation of the transcript on the genome build.
<ul><li> right-arrowhead - transcript mapped on genome plus strand.
</li><li> left-arrowhead - transcript mapped on genome minus strand.
</li></ul>
</li></ul>
<p>The line mode has several actions:
</p>
<ul><li> <b>Click</b> - Clicking at any position along a sequence line 
view causes the local sequence at that position to be displayed in the 
field at the bottom of the viewer.
</li></ul>
<ul><li> <b>Hover</b> - Placing the mouse cursor over the sequence line 
view will cause a hover text to appear with the position number and the 
next 10 sequence letters starting at that position.
</li></ul>
<p><br>
<img alt="Sequence Retriever DNA Line View.png" src="sequenceRetriever_files/Sequence_Retriever_DNA_Line_View.png" height="268" width="824">
</p><p><br>
</p>
<ul><li> <b>Double-click to Full View</b> - Double-clicking with the 
mouse on a particular line will cause just that line's full sequence to 
be displayed in the viewing window.
</li></ul>
<p><br>
<img alt="Sequence Retriever DNA Full View.png" src="sequenceRetriever_files/Sequence_Retriever_DNA_Full_View.png" height="368" width="620">
</p><p>Double-clicking again in this view will return the display to the line view.
</p><p><br>
Line view for protein sequences:
</p><p>For protein sequences, the entire line is blue.  Hover and click actions are the same as described for DNA sequences.
</p><p><br>
<img alt="Sequence Retriever Protein Line View.png" src="sequenceRetriever_files/Sequence_Retriever_Protein_Line_View.png" height="372" width="622">
</p>
<a name="Full_Sequence_View"></a><h4> <span class="mw-headline" id="Full_Sequence_View">Full Sequence View</span></h4>
<p>When the view mode is set to "Full Sequence", the entire sequence for each marker is displayed.
</p><p><br>
<img alt="Sequence Retriever DNA Full Sequence View.png" src="sequenceRetriever_files/Sequence_Retriever_DNA_Full_Sequence_View.png" height="373" width="630">
</p>
<a name="Saving_Sequences_to_the_Project"></a><h3> <span class="mw-headline" id="Saving_Sequences_to_the_Project">Saving Sequences to the Project</span></h3>
<p>To the left of each retrieved sequence is a check-box, in the column 
titled "Include".  To save sequences to the Project Folders component, 
select those desired by checking the adjacent boxes, and then press 
"Save to Project". 
</p><p><br>
<img alt="Sequence Retriever MAP4K4 DNA Add to project.png" src="sequenceRetriever_files/Sequence_Retriever_MAP4K4_DNA_Add_to_project.png" height="224" width="625">
</p><p><br>
All sequences saved at one time will appear in a single new node in the 
Project Folders component.  You will be prompted to enter a name for the
 new data sequence node.  Here we have given the node the name 
"MAP4K4_seqs".
</p><p><br>
<img alt="Project Folders MAP4K4 2 seqs.png" src="sequenceRetriever_files/Project_Folders_MAP4K4_2_seqs.png" height="146" width="244">
</p><p><br>
Note that when this node is selected, the Viewing area of the 
geWorkbench GUI will now show components that support working with 
sequences.  However, the Sequence Retrieval component will no longer be 
visible!  You must select the Project or the sequence's parent object to
 see the Sequence Retrieval component again.
</p>
<a name="Saving_sequences_to_an_external_FASTA_file"></a><h2> <span class="mw-headline" id="Saving_sequences_to_an_external_FASTA_file">Saving sequences to an external FASTA file</span></h2>
<p>Once a group of sequences has been saved to the Project Folders component, they can also be saved to a file on disk.
</p>
<ol><li> Right-click on the sequence node in the Project Folders component.
</li><li> Select Save.
</li><li> Enter a suitable name and save the file.
</li></ol>

<!-- 
NewPP limit report
Preprocessor node count: 50/1000000
Post-expand include size: 1907/2097152 bytes
Template argument size: 0/2097152 bytes
Expensive parser function count: 0/100
-->

<!-- Saved in parser cache with key workbench:pcache:idhash:1514-0!1!0!!en!2!edit=0!printable=1 and timestamp 20120309192145 -->
<div class="printfooter">
Retrieved from "<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Sequence_Retriever">http://wiki.c2b2.columbia.edu/workbench/index.php/Sequence_Retriever</a>"</div>
		<div id="catlinks" class="catlinks catlinks-allhidden"></div>		<!-- end content -->
				<div class="visualClear"></div>
	</div>
</div></div>
<div class="visualClear"></div>
<div id="footer">
	<div id="f-poweredbyico"><img src="sequenceRetriever_files/poweredby_mediawiki_88x31.png" alt="Powered by MediaWiki" height="31" width="88"></div>
	<ul id="f-list">
		<li id="lastmod"> This page was last modified on 6 March 2012, at 20:22.</li>
	</ul>
</div>
</div>

