<meta http-equiv="Content-Language" content="en-us">

<meta http-equiv="Content-Type" content="text/html; charset=windows-1252">

<meta name="GENERATOR" content="Microsoft FrontPage 4.0">

<meta name="ProgId" content="FrontPage.Editor.Document">

<title>User Comments</title>

<link rel="stylesheet" type="text/css" href="ncicb.css">

</head>




<body><h2>Blast and Other Algorithms </h2> 
    <p>The application offers some popular sequence alignment algorithms and a simple Blast Parser. The Paracel Blast Machine is used as default backend blast Machine, GeneMatcher2 are used for other algorithms which are both hosted at Columbia Genome Center. The Alignment Results help section describes the Blast results. </p> 
    <p>  This help section covers: <a href="#_To_check_the_status_of_Blast_Machin">checking blast server status, running a Blast job</a>,  <a href="#_advanced">Advanced Options, and <a href="#_Other_Algorithms">using other algorithms</a>.</p> 
    <h3><a name="_To_check_the_status_of_Blast_Machin"></a>Checking the status of Blast Machine</h3> 
    <p>1 Select the <b>Sequence Alignment</b> tab in the <b>Analysis Area</b> in the lower left of the application.</p> 
    <p>2 Click on the <b>Service </b>tab.</p>  
    <p>&#12290; Default Columbia : By default, jobs will be submitted to the BLAST server at Columbia. This server runs Paracel BLAST and is optimized for large searches. </p> 
    <p>&#12290; NCBI: This has the advantage of providing access to the latest version of the NCBI BLAST program. </p> 
    <p>Note: The text field at the bottom shows that one sequence has been selected. If you have a Fasta file that has multiple sequences, you can select the ones you want in the Markers component and activate this selection, letting you search on a subset. 
    <p>You may search on all the sequences in a file by clicking the <b>All Markers </b> checkbox. </p> 
    <p>3 Click on <b>Refresh</b> to check the status of the service. </p> 
    <p><img border="0" width="692" height="377" src="BlastOverview_files/image001.gif"></p> 
    <h3><a name="_To_run_a_Blast_job"></a>Running a Blast</h3> 
    <p>1 Load a FASTA file into a project.</p> 
    <p>2 In the <b>Analysis Area, </b>select the <b>Sequence Alignment</b> tab.</p> 
    <p>3 Select the <b>Blast</b> tab.</p> 
    <p>4 Select <b>program</b> from the dropdown list. This includes:</p> 
    <p>&#12290; blastp: Compares an amino acid query sequence against a protein sequence database. </p> 
    <p>&#12290; blastn: Compares a nucleotide query sequence against a nucleotide sequence database. </p> 
    <p>&#12290; blastx: Compares a nucleotide query sequence translated in all reading frames against a protein sequence database. </p> 
    <p>&#12290; tblastn: Compares a protein query sequence against a nucleotide sequence database dynamically translated in all reading frames.</p> 
    <p>&#12290; tblastx: Compares the six-frame translations of a nucleotide query sequence against the six-frame translations of a nucleotide sequence database.</p> 
    <p>5 Modify the values of <b>subsequence</b> if needed.</p> 
    <p>6 Click on the database name to select the <b>Database</b>. </p> 
    <p>7 Click on the curved arrow submit button. It will automatically check Blast Machine status first, then if the sever is listening, the job will be submitted to the server.</p> 
    <p>8 After the job is submitted, the progress bar will indicate that BLAST is now running. If you want to stop the job, click <b>Stop </b>(red hexagon).</p> 
    <p>9 The result will be showed on the BlastParser module and based on your choice; it may show on a new web browser. You can unselect &#8220;Display your result in your browser&#8221;, which is located under <b>Advanced Options</b>. The Alignment Results help section describes the Blast results. </p> 
    <p><a name="_advanced"></a>Advanced Options include: <a name="sm"></a> 
    <ul> 
   <p>&#12290;Frame Shift Penalty:TBD</p> 
   <p>&#12290;Query Genetic Code: TBD</p> 
    <p>&#12290;Matrix: TBD</p> 
    <p>&#12290;Expect: This sets the cutoff for which BLAST hits will be displayed. </p> 
    <p>&#12290;PFP Filter: PFP filtering for repeated sequence elements (Paracel Filtering Package).</p> 
    <p>&#12290;Low Complexity:TBD</p> 
    <p>&#12290;Mask the Lookup Table: TBD</p> 
    <p>&#12290;Display results:TBD</li> </p> 
    <p><img border="0" width="934" height="620" src="BlastOverview_files/image002.jpg"></p> 
    <h3><a name="_Other_Algorithms"></a>Other Algorithms <a name="sm"></a></h3> 
    <p>To use other algorithms, select from &quot;Other Algorithms&quot; panel. All other algorithms requests will be processed on GeneMatcher2 machine.</p> 
    <p>1 Load a FASTA file into a project</p> 
    <p>2 In the <b>Analysis Area</b>, select the <b>Sequence Alignment</b> tab</p> 
    <p>3 Select the <b>Other Algorithms</b> tab</p> 
    <p>4 Select a <b>Database </b>from the dropdown list.</p> 
    <p>5 Modify the values of <b>subsequence</b> if needed</p> 
    <p>6 Select an Algorithm from the dropdown list. The includes:</p> 
    <p>&#12290; Smith-Waterman Protein: compares an amino acid query sequence against a protein sequence database </p> 
    <p>&#12290; Smith-Waterman DNA: compares a nucleotide query sequence against a nucleotide sequence database </p> 
    <p>&#12290; Frame(from DNA sequence to protein DB): compares a nucleotide query sequence against a protein sequence database </p> 
    <p>&#12290; Frame(from Protein sequence to DNA DB): compares an amino acid query sequence against a nucleotide sequence database</p> 
    <p>7 Click on <b>Search</b>. </p> 
    <p>&#12290; The results are displayed in the Alignment Results visualization component.<a name="_To_Add_Sequences_to_a_Project"></a></p>
    </body>



</html>

