<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html dir="ltr" lang="en"><head>
<title>BLAST - Workbench</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="generator" content="MediaWiki 1.16.0">
<meta name="robots" content="noindex,follow">

</head>
<body class="mediawiki ltr ns-0 ns-subject page-BLAST skin-monobook">
<div id="globalWrapper">
<div id="column-content"><div id="content">
	<a id="top"></a>
	
	<h1 id="firstHeading" class="firstHeading">BLAST</h1>
	<div id="bodyContent">
		<!-- start content -->
<p><br>
</p>
<a name="Overview"></a><h1> <span class="mw-headline" id="Overview">Overview</span></h1>
<p>"The Basic Local Alignment Search Tool (BLAST) finds regions of local
 similarity between sequences. The program compares nucleotide or 
protein sequences to sequence databases and calculates the statistical 
significance of matches. BLAST can be used to infer functional and 
evolutionary relationships between sequences as well as help identify 
members of gene families" (quoted from the <a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastHome" class="external text" rel="nofollow">NCBI BLAST homepage</a>).
</p><p>geWorkbench submits BLAST jobs to the NCBI server.  
NCBI-supported sequence databases and search algorithms can be selected 
in the user interface.  Since release 2.1.0, geWorkbench supports almost
 all BLAST setting options available through the NCBI web interface.  
</p><p>Please note that although, in geWorkbench, we have adopted the 
default settings for each BLAST algorithm as seen on the NCBI website, 
those settings are subject to change at any time by NCBI.  Before 
submitting a BLAST job from geWorkbench, the user should verify that the
 parameter settings are appropriate for their query.
</p><p><br>
The BLAST analysis is available when a protein or DNA sequence is loaded
 and selected in the Project Folders component.  The BLAST Analysis and 
BLAST Results Viewer must be loaded in the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Component_Configuration_Manager" title="Component Configuration Manager">Component Configuration Manager</a>.
</p><p><br>
</p><p><img alt="BLAST Parameters Main-blastn-full.png" src="blast_files/BLAST_Parameters_Main-blastn-full.png" height="561" width="1095">
</p><p><b>Figure legend: BLAST Main Parameters</b>. A nucleotide sequence has been loaded into the Project Folders component.
</p>
<a name="NCBI_Documentation"></a><h1> <span class="mw-headline" id="NCBI_Documentation">NCBI Documentation</span></h1>
<p>geWorkbench serves as a interface to the NCBI BLAST server and 
implements the same options as the NCBI BLAST website.  Detailed 
information about each option can be found on NCBI webpages, including:
</p>
<ul><li> <a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi" class="external text" rel="nofollow">NCBI BLAST Home Page</a>
</li><li> <a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs" class="external text" rel="nofollow">BLAST Docs - Main help page</a>
<ul><li> <a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=ProgSelectionGuide" class="external text" rel="nofollow">Blast Program Selection Guide</a>
</li></ul>
</li></ul>
<p>Older help pages...
</p>
<ul><li> <a href="http://www.ncbi.nlm.nih.gov/blast/html/blastcgihelp.html" class="external text" rel="nofollow">NCBI BLAST Help page 1</a> and
</li><li> <a href="http://www.ncbi.nlm.nih.gov/BLAST/blastcgihelp.shtml" class="external text" rel="nofollow">NCBI BLAST help page 2 </a>.
</li></ul>
<p><br>
</p>
<a name="BLAST_job_setup"></a><h1> <span class="mw-headline" id="BLAST_job_setup">BLAST job setup</span></h1>
<a name="Prerequisites"></a><h2> <span class="mw-headline" id="Prerequisites">Prerequisites</span></h2>
<ul><li> The BLAST Analysis and Viewer components must be loaded in the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Component_Configuration_Manager" title="Component Configuration Manager">Component Configuration Manager</a>.  
</li><li> A protein or nucleotide sequence must be loaded in the Project Folders component.
</li></ul>
<a name="Query_sequences"></a><h2> <span class="mw-headline" id="Query_sequences">Query sequences</span></h2>
<p>BLAST accepts nucleotide or amino-acid query sequences in the FASTA 
format. A query file can contain one or multiple sequences.  The file 
can be loaded from disk using the <b>File Open</b> command, or may have 
been placed into the Project Folders component by another component such
 as the Sequence Retriever, or add from the result of a previous BLAST 
run.
</p>
<a name="Invoking_BLAST"></a><h2> <span class="mw-headline" id="Invoking_BLAST">Invoking BLAST</span></h2>
<p>BLAST is a normal geWorkbench analysis component and can be invoked 
either by right-clicking on a sequence node in the Project Folders 
component, or through the Commands entry in the Menu Bar at the top of 
the geWorkbench GUI.
</p><p><br>
Invoking BLAST via right-clicking on a sequence node in the Project Folders component:
</p><p><br>
<img alt="BLAST Analysis invocation.png" src="blast_files/BLAST_Analysis_invocation.png" height="287" width="374">
</p><p><br>
</p><p>Invoking BLAST via the Commands menu.  The desired sequence node must be selected first in the Project Folders component:
</p><p><br>
<img alt="BLAST Analysis Command invocation.png" src="blast_files/BLAST_Analysis_Command_invocation.png" height="151" width="411">
</p>
<a name="Parameters_-_Main"></a><h2> <span class="mw-headline" id="Parameters_-_Main">Parameters - Main</span></h2>
<p>In addition to the parameters shown in the previous image, BLASTX and TBLASTX add the Genetic Code option:
</p><p><img alt="BLAST Parameters Main-blastx.png" src="blast_files/BLAST_Parameters_Main-blastx.png" height="448" width="472">
</p>
<a name="Algorithms"></a><h3> <span class="mw-headline" id="Algorithms">Algorithms</span></h3>
<p>The user must make sure that the algorithm chosen matches the type of
 query sequence (protein or nucleotide) that has been loaded.  Some of 
the algorithms translate a nucleotide query, a nucleotide database, or 
both into amino acid sequence before executing the query.  Searching in 
the amino-acid space is more sensitive for certain types of query, as it
 ignores synonymous, non-functional changes in nucleotide sequence.
</p>
<a name="For_protein_query_sequences:"></a><h4> <span class="mw-headline" id="For_protein_query_sequences:">For protein query sequences:</span></h4>
<ul><li> <b>blastp</b> - Compares an amino acid query sequence against a protein sequence database.
</li><li> <b>tblastn</b> - Compares a amino acid query sequence against a nucleotide database translated in all reading frames.
</li></ul>
<a name="For_nucleotide_query_sequences:"></a><h4> <span class="mw-headline" id="For_nucleotide_query_sequences:">For nucleotide query sequences:</span></h4>
<ul><li> <b>blastn</b> - three algorithms are available under the blastn
 choice, in order of decreasing similarity of the query to the target 
sequences,  Each compares a nucleotide query sequence against a 
nucleotide sequence database:
<ul><li> <b>megablast</b> - optimize for highly similar sequences.
</li><li> <b>discontinuous megablast</b> - optimize for more dissimilar sequences
</li><li> <b>blastn</b> - optimize for somewhat similar sequences.
</li></ul>
</li><li> <b>blastx</b> - Compares a nucleotide query sequence translated in all reading frames against a protein sequence database.
</li><li> <b>tblastx</b> - Compares the 6-frame translations of a 
nucleotide query sequence against the 6-frame translations of a 
nucleotide sequence database.
</li></ul>
<a name="Databases"></a><h3> <span class="mw-headline" id="Databases">Databases</span></h3>
<p>Standard protein and nucleic acid databases maintained at NCBI are 
supported.  The appropriate databases for the search algorithm chosen 
will be displayed.  A window to the right, "Database Details", displays a
 summary of the database currently selected in the list.
</p>
<a name="For_nucleic_acids:"></a><h4> <span class="mw-headline" id="For_nucleic_acids:">For nucleic acids:</span></h4>
<ul><li> <b>Nucleotide Collection (nr/nt)</b>
</li><li> <b>Reference mRNA seqeunces (refseq_rna)</b>
</li><li> <b>Reference genomic sequneces (refseq_genomic</b>
</li><li> <b>NCBI Genomes (chromosome)</b>
</li><li> <b>Expressed Sequence Tags (est)</b>
</li><li> <b>Human subset of EST (est_human)</b>
</li><li> <b>Mouse subset of EST (est_mouse)</b>
</li><li> <b>Non-human, non-mouse ESTs (est_others)</b>
</li><li> <b>Genomic survey sequences (gss)</b>
</li><li> <b>High throughput genomic sequences (htgs)</b>
</li><li> <b>Patent sequences (pat)</b>
</li><li> <b>Protein Data Bank (pdb)</b>
</li><li> <b>Human ALU repeat elements (alu)</b>
</li><li> <b>Sequence tagges sites (dbsts)</b>
</li><li> <b>Whole genome shotgun reads (wgs)</b>
</li><li> <b>Environmental Samples (env_nt)</b>
</li></ul>
<a name="For_proteins:"></a><h4> <span class="mw-headline" id="For_proteins:">For proteins:</span></h4>
<ul><li> <b>Non-redundant protein sequences (nr)</b> 
</li><li> <b>Reference proteins (refseq_protein)</b>
</li><li> <b>Swissprot protein sequences (swissprot)</b>
</li><li> <b>Patented protein sequences (pat)</b>
</li><li> <b>Protein Data Bank sequences (pdb)</b>
</li><li> <b>Environmental samples (env_nr)</b>
</li></ul>
<a name="Search_Choice"></a><h3> <span class="mw-headline" id="Search_Choice">Search Choice</span></h3>
<p>These settings provide the ability to restrict the search in certain ways
</p>
<ul><li> <b>Exclude</b>: exclude certain specialized database entries from the search.
<ul><li> <b>Models</b> (XM/XP)
</li><li> <b>Uncultured/environmental sequences</b>
</li></ul>
</li></ul>
<ul><li> <b>Entrez Query</b> - an Entrez format query can be entered directly to restrict a search e.g. to a particular species.
</li></ul>
<a name="Genetic_Code"></a><h3> <span class="mw-headline" id="Genetic_Code">Genetic Code</span></h3>
<p>Genetic code to be used in blastx (and tblastx) translation of the query.
</p>
<a name="Algorithm_Parameters"></a><h2> <span class="mw-headline" id="Algorithm_Parameters">Algorithm Parameters</span></h2>
<a name="General_Parameters"></a><h3> <span class="mw-headline" id="General_Parameters">General Parameters</span></h3>
<ul><li> <b>Max target sequences</b> - maximum number of hits to return. 
</li><li> <b>Automatically adjust parameters for short input sequences</b>.
</li><li> <b>Expect threshold</b> - Expected number of chance matches in a random model.
</li><li> <b>Word size</b> - The length of the seed that initiates an alignment.
</li><li> <b>Max matches in a query range</b> - Limit the number of 
matches to a query range. This option is useful if many strong matches 
to one part of a query may prevent BLAST from presenting weaker matches 
to another part of the query.
<ul><li> Note - NCBI reports that this feature may not work if there are
 a large number of full-length sequence matches in the chosen database.
</li><li> The function of this feature is described in <a href="http://www.ncbi.nlm.nih.gov/pubmed/10890403" class="external text" rel="nofollow">Berman et al., 2000</a>.
</li></ul>
</li></ul>
<a name="Scoring_Parameters"></a><h3> <span class="mw-headline" id="Scoring_Parameters">Scoring Parameters</span></h3>
<ul><li> <b>Match/mismatch scores</b> (blastn, megablast, discontinuous megablast) - scores to use for a match or mismatch.
</li><li> <b>Matrix</b> - Various scoring matrices (BLOSUM, PAM) are available for protein and translated queries.  
</li><li> <b>Gap Costs</b> - The pull-down menu shows the available choices of gap costs for the current scoring matrix. 
</li><li> <b>Compositional adjustments</b> -  "...takes into account the
 amino acid composition of the individual database sequences involved in
 reported alignments. This improves E-value accuracy, thereby reducing 
the number of false positive results."
</li></ul>
<a name="Filters_and_Masking"></a><h3> <span class="mw-headline" id="Filters_and_Masking">Filters and Masking</span></h3>
<ul><li> <b>Low Complexity</b> - filter out low compositional complexity sequence.
</li><li> <b>Species-specific repeats filter</b> - masks 
species-specific repeats (e.g. human LINE's and SINE's).  This option 
can speed searches involving long query sequences or databases 
containing sequences with many repeats.
</li><li> <b>Mask for lookup table only</b> - masks low-complexity 
sequence only while constructing the lookup table used by the initial 
hit-find phase of BLAST. The second phase, hit extension, is not not 
affected and hits can be extended through low-complexity sequence.  NCBI
 notes that this option is experimental and subject to change.
</li><li> <b>Mask lower case letter</b> - filter out sequence which is in lower case in the FASTA query sequence.
</li></ul>
<a name="Discontiguous_Word_Options"></a><h3> <span class="mw-headline" id="Discontiguous_Word_Options">Discontiguous Word Options</span></h3>
<p>Please see the <a href="http://www.ncbi.nlm.nih.gov/blast/discontiguous.shtml" class="external text" rel="nofollow">NCBI page on discontiguous megablast</a> for a detailed explanation of these options.
</p>
<ul><li> <b>Template Length</b> 
</li><li> <b>Template Type</b>
</li></ul>
<a name="Other"></a><h3> <span class="mw-headline" id="Other">Other</span></h3>
<ul><li> <b>Display result in your web browser</b> - geWorkbench will display the HTML page returned by NCBI BLAST in your web browser as well is within its own display.
</li><li> <b>Restore defaults</b> - restore all settings for the currently selected algorithm to their default values.
</li></ul>
<p><br>
</p>
<a name="Algorithm_Parameter_Setting_Defaults"></a><h3> <span class="mw-headline" id="Algorithm_Parameter_Setting_Defaults">Algorithm Parameter Setting Defaults</span></h3>
<p>The default settings for each query type were taken from the NCBI BLAST website.
</p><p>Please note that options and default settings on the NCBI BLAST website are subject to change at any time.
</p><p>The user should verify all settings are appropriate for his or her particular BLAST query.
</p>
<a name="blastn_-_megablast"></a><h4> <span class="mw-headline" id="blastn_-_megablast">blastn - megablast</span></h4>
<p><img alt="BLAST Parameters Main-blastn.png" src="blast_files/BLAST_Parameters_Main-blastn.png" height="432" width="483">
</p><p><br>
<img alt="BLAST Parameters blastn-megablast.png" src="blast_files/BLAST_Parameters_blastn-megablast.png" height="436" width="788">
</p>
<a name="blastn_-_discontiguous_megablast"></a><h4> <span class="mw-headline" id="blastn_-_discontiguous_megablast">blastn - discontiguous megablast</span></h4>
<p><img alt="BLAST Parameters blastn-discontinuous-megablast.png" src="blast_files/BLAST_Parameters_blastn-discontinuous-megablast.png" height="431" width="791">
</p>
<a name="blastn_-_blastn"></a><h4> <span class="mw-headline" id="blastn_-_blastn">blastn - blastn</span></h4>
<p><img alt="BLAST Parameters blastn.png" src="blast_files/BLAST_Parameters_blastn.png" height="435" width="793">
</p>
<a name="blastp"></a><h4> <span class="mw-headline" id="blastp">blastp</span></h4>
<p><img alt="BLAST Parameters Main-blastp.png" src="blast_files/BLAST_Parameters_Main-blastp.png" height="421" width="483">
</p><p><br>
</p><p><img alt="BLAST Parameters blastp.png" src="blast_files/BLAST_Parameters_blastp.png" height="436" width="792">
</p>
<a name="blastx"></a><h4> <span class="mw-headline" id="blastx">blastx</span></h4>
<p><img alt="BLAST Parameters Main-blastx.png" src="blast_files/BLAST_Parameters_Main-blastx.png" height="448" width="472">
</p><p><br>
<img alt="BLAST Parameters blastx.png" src="blast_files/BLAST_Parameters_blastx.png" height="423" width="693">
</p><p><br>
</p>
<a name="tblastx"></a><h4> <span class="mw-headline" id="tblastx">tblastx</span></h4>
<p><img alt="BLAST Parameters Main-tblastx.png" src="blast_files/BLAST_Parameters_Main-tblastx.png" height="422" width="483">
</p><p><br>
<img alt="BLAST Parameters tblastx.png" src="blast_files/BLAST_Parameters_tblastx.png" height="409" width="726">
</p>
<a name="tblastn"></a><h4> <span class="mw-headline" id="tblastn">tblastn</span></h4>
<p><img alt="BLAST Parameters Main-tblastn.png" src="blast_files/BLAST_Parameters_Main-tblastn.png" height="428" width="470">
</p><p><br>
<img alt="BLAST Parameters tblastn.png" src="blast_files/BLAST_Parameters_tblastn.png" height="433" width="793">
</p>
<a name="Analyze"></a><h2> <span class="mw-headline" id="Analyze">Analyze</span></h2>
<p>The BLAST analysis is launched by pushing the <b>Analyze</b> button.  A dialog with a progress bar will appear.  The analysis can be canceled by pushing the <b>Cancel</b> button on this dialog.
</p><p><br>
<img alt="BLAST Analysis Progress Dialog.png" src="blast_files/BLAST_Analysis_Progress_Dialog.png" height="134" width="326">
</p>
<a name="Analysis_Controls_-_releases_2.2.2_and_earlier"></a><h2> <span class="mw-headline" id="Analysis_Controls_-_releases_2.2.2_and_earlier">Analysis Controls - releases 2.2.2 and earlier</span></h2>
<ul><li> <b>All Markers</b> - if selected, use all sequences loaded, overriding any activated sets in the Marker Sets component.
</li><li> <b>Total Sequence Number</b> - indicates how many sequences have been selected for query.
</li><li> <b>Curling arrow</b> - start BLAST search
</li><li> <b>Stop sign</b> - stop BLAST search (if pushed, geWorkbench will not wait for or retrieve the BLAST results).
</li></ul>
<p><img alt="T SequenceAlignment BLAST general controls.png" src="blast_files/T_SequenceAlignment_BLAST_general_controls.png" height="44" width="380">
</p>
<a name="BLAST_Results_Viewer"></a><h1> <span class="mw-headline" id="BLAST_Results_Viewer">BLAST Results Viewer</span></h1>
<p>When the BLAST search results are returned they are placed in a new 
node in the Project Folders component as a child of the query sequence 
used.  Mousing over the result set will show how many sequences are in 
it.
</p><p><br>
<img alt="Blast result projects folder.png" src="blast_files/Blast_result_projects_folder.png" height="162" width="508">
</p><p>Each different hit is listed on a line in the results table, 
shown below.  Note that a query sequence can hit a database target 
sequence in more than one place, resulting in multiple alignments 
displayed per target hit.  The results viewer also shows statisics for 
each hit, including the E-value, start position and length of the hit, 
and the percent identity.
</p><p>If the "Display result in your web browser" option was chosen, then the browser will open with the HTML formated results.   
</p><p>In the pane at left in the picture below, the name of the input 
query sequence is shown, e.g. the gi number of a Genbank sequence.  If 
there had been more than one query sequence, then this pane would show a
 list of query sequence names, allowing you to select the results to be 
viewed.
</p><p><img alt="Blast Wilms NM result.png" src="blast_files/Blast_Wilms_NM_result.png" height="585" width="1051">
</p><p><br>
</p>
<a name="Controls"></a><h2> <span class="mw-headline" id="Controls">Controls</span></h2>
<a name="Within_the_list_of_returned_hits"></a><h3> <span class="mw-headline" id="Within_the_list_of_returned_hits">Within the list of returned hits</span></h3>
<ul><li> <b>Include</b> check boxes - when checked, selects these sequences for import into the Project Folders component.
</li></ul>
<a name="At_the_bottom_of_the_pane"></a><h3> <span class="mw-headline" id="At_the_bottom_of_the_pane">At the bottom of the pane</span></h3>
<ul><li> <b>Reset</b> - uncheck all "Include" boxes.
</li><li> <b>Select All</b> - mark as checked all the "Include" boxes.
</li><li> <b>Add Selected Sequences to Project</b> - for each hit whose "Include" box is checked, add its sequence to a sequence node in the Project Folders component.
</li><li> <b>Only Add Aligned Parts</b> - for each hit whose "Include" 
box is checked, add to the Project Folders component only the portion of
 its sequence which aligned with the query sequence .
</li></ul>
<a name="In_the_query_list"></a><h3> <span class="mw-headline" id="In_the_query_list">In the query list</span></h3>
<ul><li> <b>Search</b> - input a text search to find entries in the list of queries.
</li><li> <b>Find Next</b> - search for the next occurence of the entered text.
</li></ul>
<a name="Adding_selected_sequence_hits_to_the_project"></a><h2> <span class="mw-headline" id="Adding_selected_sequence_hits_to_the_project">Adding selected sequence hits to the project</span></h2>
<p>The sequences corresponding to individual hits in the BLAST search 
can be retrieved from NCBI and added to the Project Folders component.
</p><p><img alt="Blast Wilms NM include 5prime.png" src="blast_files/Blast_Wilms_NM_include_5prime.png" height="524" width="831">
</p><p>Here we select a particular hit, by checking the "include" box 
next to it, and press the button "Add Selected Sequences to Project".  
The sequence is retrieved and placed into the Project Folders component 
as shown below:
</p><p><img alt="Blast Wilms NM include 5prime added.png" src="blast_files/Blast_Wilms_NM_include_5prime_added.png" height="167" width="329">
</p>
<a name="Example:_Running_a_BLAST_search"></a><h1> <span class="mw-headline" id="Example:_Running_a_BLAST_search">Example: Running a BLAST search</span></h1>
<p>Two Genbank sequence files in FASTA format are included in the 
geWorkbench data/public_data folder: a nucleotide sequence, "NM 
_024426-Wilms.Fasta", and its protein sequence, "NP_077744-Wilms.fasta".
</p><p>For a simple search using the nucleotide query file, one can select the <b>blastn/megablast</b> program and search against the <b>nr/nt</b> database of nucleotide sequences.  
</p>
<ul><li> In the Project Folders component, 
<ul><li> right-click on a Project and select "Open File(s)", or
</li><li> from the top-level "File" menu, select "Open-&gt;File".  
</li></ul>
</li><li> Select a file type of FASTA.  
</li><li> Navigate to data/public_data within the geWorkbench distribution and select the file "NM _024426-Wilms.Fasta".
</li><li> Press "Open".
</li><li> Right-click on the new sequence data node.
</li><li> Select Analysis-&gt;BLAST Analysis.
</li></ul>
<p><br>
</p>
<ul><li> For program select <b>blastn</b>, and leave the default algorithm choice set to megablast.
</li><li> For database, select <b>nr</b> - the complete nucleotide database. 
</li></ul>
<p><b>Note:</b> - If you have a fasta file that has multiple sequences, 
you can create a set in the Markers component and use only this set for 
the search, by activating the set (check the box by its name in the 
Markers component). You can override an activated Marker Set and search 
on all sequences in a file by clicking the <b>All Markers</b> checkbox.
</p>
<ul><li> Click on the <b>Algorithm Parameters</b> Tab
</li><li> Change the <b>Expect threshold</b> to 0.01.  This sets the cutoff for which BLAST hits will be displayed.
</li><li> Leave the <b>Display result in your web browser</b> checked.
</li></ul>
<ul><li> Hit the "Analyze" button.  The job will be submitted and the results returned as shown in the sections above.
</li></ul>
<a name="Technical_Notes"></a><h1> <span class="mw-headline" id="Technical_Notes">Technical Notes</span></h1>
<ul><li> The NCBI BLAST server may occasionally return an error when 
sequences are searched from geWorkbench. The problem appears to depend 
on the load on the NCBI BLAST server. 
</li><li> When geWorkbench is asked to submit multiple sequences to the 
NCBI BLAST server, it will submit them one at a time and wait for the 
results before submitting the next sequence.  This is done to simplify 
the subsequent parsing and display of the results.
</li></ul>
<a name="References"></a><h1> <span class="mw-headline" id="References">References</span></h1>
<ul><li> Berman P, Zhang Z, Wolf YI, Koonin EV, Miller W. (2000) 
Winnowing sequences from a database search. J Comput Biol. 
7(1-2):293-302.
</li></ul>
<ul><li> The NCIB BLAST site provides a comprehensive <a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=References" class="external text" rel="nofollow">list of references</a>.
</li></ul>

<!-- 
NewPP limit report
Preprocessor node count: 162/1000000
Post-expand include size: 1907/2097152 bytes
Template argument size: 0/2097152 bytes
Expensive parser function count: 0/100
-->

<!-- Saved in parser cache with key workbench:pcache:idhash:1519-0!1!0!!en!2!printable=1 and timestamp 20120306225524 -->
<div class="printfooter">
Retrieved from "<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/BLAST">http://wiki.c2b2.columbia.edu/workbench/index.php/BLAST</a>"</div>
		<div id="catlinks" class="catlinks catlinks-allhidden"></div>		<!-- end content -->
				<div class="visualClear"></div>
	</div>
</div></div>
<div class="visualClear"></div>
<div id="footer">
	<div id="f-poweredbyico"><img src="blast_files/poweredby_mediawiki_88x31.png" alt="Powered by MediaWiki" height="31" width="88"></div>
	<ul id="f-list">
		<li id="lastmod"> This page was last modified on 6 March 2012, at 17:54.</li>
	</ul>
</div>
</div>

