   <html>



<head>

<meta http-equiv="Content-Language" content="en-us">

<meta http-equiv="Content-Type" content="text/html; charset=windows-1252">

<meta name="GENERATOR" content="Microsoft FrontPage 4.0">

<meta name="ProgId" content="FrontPage.Editor.Document">

<title>User Comments</title>

<link rel="stylesheet" type="text/css" href="ncicb.css">

</head>



<body><h2>MINDY</h2> 
    <p>The MINDY algorithm uses gene expression data in order to quantify modulation effects on the regulatory activity of a transcription factor gene (TF) of interest. The details of the algorithm are described in <i><a href="https://sharepoint.c2b2.columbia.edu/c2b2/geWorkbench%20v11/Use%20Cases/MINDY/Califano_JCB.pdf">MINDY: An Algorithm for the Genome-wide Discovery of Modulators of Transcriptional Interactions</a></i>. The general workflow includes 1) selection of modulator genes in the modulator panel, 2) defining (result) marker sets in the Table and List panel and displaying results for a single modulator and its targets in the Heat Map panel. This help topic describes executing <a href="#_Run_MINDY">MINDY analysis</a> and <a href="#_View_MINDY_Results">viewing the results.</a></p> 
    <h3><img border="0" width="553" height="147" src="MINDY_help_files/image001.jpg"></h3> 
    <h3>Run MINDY Analysis</h3> 
    <p>1. Select a microarray set node in the Project Folder. </p> 
    <p>2. In the analysis pane (lower right of the application), select <b> MINDY Analysis</b>. </p> 
    <p>3. In the Main tab, populate the <b> Modulators List</b> by (1) manually entering a comma separated list of marker names to be used as <b>Modulators</b> or (2) click on <b>Load</b> to select a file containing the list of Modulators. </p> 
    <p>4. Populate the <b>Target List</b> text box by (1) manually entering a comma separated list of marker names to be used as Targets or (2) click on <b>Load</b> to select a file containing the list of Targets. </p> 
    <p>5. Populate the <b>Hub Gene</b> textbox to designate the TF gene by (1) typing the gene name or (2) in the Selection Area (lower left) Marker Tab, click on the marker name corresponding to the TF .</p> 
    <p>6. In the Advanced Tab, accept the default values or update the parameter values by clicking the parameter text box and typing a new value.</p> 
    <p>7. Click <b>Analyze</b>. If successful, the project window is updated to reflect the MINDY result node which is a child node of the input dataset. Please note that the Dataset History captures the analysis parameters. </p> 
    <h3>View MINDY Results</h3> 
    <h3><img border="0" width="553" height="284" src="MINDY_help_files/image002.jpg"></h3> 
    <p>1. Select the MINDY result node in the Project Folder. </p> 
    <p>2. In the Modulator Tab, indicate the modulators of interest using the checkboxes or click on <b>Select All</b> to display all modulators in the Table, List,and Heat Map views. The <b>Modulators Selected</b> is updated to reflect the number of modulators selected. Only selected Modulators are displayed on the Table, List and Heat Map views. Additional actions include:</p> 
    <p>&#183; <b><i>Marker Display: </i></b>Indicate marker display preferences.</p> 
    <p>&#183; <b><i>Sort:</i></b> Click on the column headers or use sort options available in the left pane. </p> 
    <p>&#183; <b><i>Add to Set: </i></b> Adds selected modulators to a Marker Set. You can select one or more Targets and/or Modulators, using the selection check boxes</p> 
    <p>&#183; <b><i>All Markers:</i></b> This checkbox determines if all the target genes are displayed or only genes that belong in activated marker groups. </p> 
    <p>3. Select from the various tabs to view the data in alternate formats. See <a href="#_Navigating_Mindy_Visualization">Navigating Mindy</a> for additional information on these data views.</p> 
    <h3>Navigating Mindy Visualization</h3> 
    <p>MINDY includes the following data views: <a href="#mod">Modulator</a>, <a href="#table">Table</a>, <a href="#list">List</a> and <a href="#heatmap">Heat Map</a>.</p> 
    <p><a name="#mod"><b>Modulator</b></a>: This table-based view contains one row per modulator gene. Only modulators selected in this tab are included the other data views. The value of the Mode column for a modulator M is either &#8220;+&#8221;, &#8220;-&#8220; or null (0) depending on if M+ is larger, smaller or equal to M-. </p> 
    <p><a name="#table"><b>Table</b></a>: The rows of the table represent target genes and the columns represent modulators. Additional actions include:</p> 
    <p><b><i>Marker Display: </i></b>Indicate marker display preferences.</p> 
    <p><b><i>Sorting: </i></b> Displays columns (modulators) from left to right in descending order by; Aggregate ( M#), Enhancing (M+) or Negative (M-).</p>  
    <p><b><i>Modulator Limits: </i></b>Activates the checkbox to limit the columns (modulators) display to a defined value. This selectionfilters the modulator display based upon the current display order.</p> 
    <p><b><i>Display Options: </i></b></p> 
    <p>&#183; <b>Color View: </b>Enables a heat map display of each cell based on the value of the score. 1 is displayed as absolute blue; +1 is displayed as absolute red; 0:1 is mapped uniformly from white to shades of red; -1:0 is mapped uniformly from shades of blue to white.</p> 
    <p>&#183; <b>Score View:</b> Displays the discretized score values.</p> 
    <p><img border="0" width="483" height="505" src="MINDY_help_files/image003.jpg"></p> 
    <p><b>List: </b> The table has three columns: Modulator, Target and Score. Additional actions include:</p> 
    <p><b><i>Marker Display:</b></i>Indicate marker display preferences.</p> 
    <p><b><i>Marker Override: </b></i>Marker selection preferences. As markers are selected, the number of markers selected is listed next to Enable Selection field.  This does not reflect the number of rows. </p> 
    <p><img border="0" width="572" height="333" src="MINDY_help_files/image004.gif"></p> 
    <p><a name="#heatmap"><b>Heat Map:</b></a> The heat map contains two color mosaic panels. The rows correspond to target genes and arrays are ordered according to the expression of the TF gene (low to high). The left panel correspond to the L- arrays where the modulator is least expressed while the columns on the right panel to the L+ arrays where the modulator is most expressed. Additional actions include:</p> 
    <p>&#183; <b><i>Marker Display:</b></i> Indicate marker display preferences.</p> 
    <p>&#183; <b><i>Transcription Factor:</b></i> Displays the TF entered in the Analysis parameters.</p> 
    <p>&#183; <b><i>Modulator: </b></i>Select a modulator from the &#8220;Selected Modulators&#8221; list to update the Heat Map display.</p> 
    <p>&#183; <b><i>Refresh: </b></i>Resets the heat map display.</p> 
    <p>&#183; <b><i>Image Snapshot:</b></i> Captures the heat map as an image node in the Project Folder.</p> 
    <p><b><img border="0" width="863" height="319" src="MINDY_help_files/image005.jpg"></b></p>
	</body>



</html>

