<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html dir="ltr" lang="en"><head>
<title>MINDy - Workbench</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="generator" content="MediaWiki 1.16.0">
<meta name="robots" content="noindex,follow">

</head>
<body class="mediawiki ltr ns-0 ns-subject page-MINDy skin-monobook">
<div id="globalWrapper">
<div id="column-content"><div id="content">
	<a id="top"></a>
	
	<h1 id="firstHeading" class="firstHeading">MINDy</h1>
	<div id="bodyContent">
		<!-- start content -->
<p><br>
</p><p><br>
</p>
<a name="MINDy_Analysis"></a><h1> <span class="mw-headline" id="MINDy_Analysis">MINDy Analysis</span></h1>
<p>The MINDy algorithm (Modulator Inference by Network Dynamics) uses 
gene expression data to determine whether a putative modulator gene (Mj)
 influences the regulatory activity of a transcription factor gene (TF) 
over a set of target genes (Ti).  This influence is measured in terms of
 whether there is a <b>change in the correlation (measured as mutual information)</b>
 of expression between the TF and its targets Ti conditional on a change
 in the expression of Mj.  The change in correlation is calculated as 
the difference in mutual information (<b>delta (MI)</b>) for each TF-Ti 
pair between the two conditions (modulator high or low).  The mutual 
information values used in MINDy are calculated using the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Tutorial_-_ARACNE" title="Tutorial - ARACNE" class="mw-redirect">ARACNe</a> algorithm, which is also a part of geWorkbench.
</p><p><br>
</p>
<a name="Outline_of_MINDy_calculations"></a><h2> <span class="mw-headline" id="Outline_of_MINDy_calculations">Outline of MINDy calculations</span></h2>
<ol><li> A microarray gene expression dataset is selected.
</li><li> The user specifies a set of one or more candidate modulator 
genes (Mj), a hub transcription factor (TF), and a set of putative 
targets of the transcription factor (Ti).
</li><li> Parameters for the MINDy run are set.
</li><li> Using the expression value of the chosen modulator gene Mj, 
the arrays in the experiment are ordered (as columns in the data 
matrix), from lowest to highest.  
</li><li> Two subsets of arrays are then chosen from each end (tail) of 
the ordered list. One subset contains arrays in which Mj shows the 
lowest expression, and the other subset contains arrays in which Mj 
shows the highest expression.  The subsets are non-overlapping. A 
typical trial might involve assigning the lowest 35% of the arrays to 
the low group (M-), as measured by expression of Mj, and the highest 35%
 to the high group (M+).  The remaining arrays are not further 
considered.
</li><li> For each target Ti, the conditional mutual information between
 the hub TF and the target is then calculated for the array subsets M+ 
and M- separately, and the difference is taken (delta (MI)).
</li><li> The resulting delta (MI)s are displayed.  At present, a 
p-value is not calculated on the delta (MI).  Larger values of delta 
(MI) may indicate an interesting change in the mutual information 
conditional on the expression of the modulator, that is, the modulator 
has an effect on the <b>correlation of expression</b> between the hub TF and the target gene.  
</li><li> The sign of the influence of the modulator is also displayed. 
 A positive modulation effect (+) is where high expression of the 
modulator gene Mj <b>increases </b> the mutual information between the 
hub TF and the target gene.  Likewise, A negative modulation effect (-) 
is where increasing the expression of the modulator gene Mj <b>decreases</b> the mutual information between the hub TF and the target gene.
</li></ol>
<a name="Prerequisites_for_MINDy_calculations"></a><h2> <span class="mw-headline" id="Prerequisites_for_MINDy_calculations">Prerequisites for MINDy calculations</span></h2>
<ol><li> <b>Number of arrays</b> - A microarray gene expression data set
 with a sufficient number of arrays must be present.  For optimal 
results, at least 250 to 300 microarrays of a homogenous cellular system
 should be used, for example, isolable tumor cells or cell lines, with a
 range of different expression conditions (distinct cellular 
phenotypes).  (300 arrays has been found to give good results, while 250
 has been found to be an absolute minimum).
</li><li> <b>Modulator expression variation</b> - The expression of the 
modulator (Mj) must have a sufficient expression range to separate its 
two expression tails compared to the experimental noise level.  Low 
variation markers can be removed by running the deviation filter 
(Filtering component) on the dataset before starting the MINDy 
calculation.
</li><li> <b>Independence of modulator and TF hub</b> - Any modulator 
(Mj) whose expression profile  is not statistically independent of that 
of the hub transcription factor (TF) must be excluded.  This can be 
determined using a mutual information calculation (ARACNe).  This 
functionality is not currently directly implemented within MINDy in 
geWorkbench, but can be run directly using the ARACNe component.
</li><li> <b>Note</b> - The "Target List" also is used to represent all 
markers which will be used in the calculations.  As such, all hub 
markers and candidate modulator markers must be included in this list.
</li></ol>
<a name="Parameters_-_Main"></a><h2> <span class="mw-headline" id="Parameters_-_Main">Parameters - Main</span></h2>
<p><img alt="MINDy Parameters Main.png" src="mindy_files/MINDy_Parameters_Main.png" height="351" width="1210">
</p><p><br>
The Modulators List, Target List, and Hub Marker fields are populated 
using marker IDs as represented in the Markers component.  Note that 
these are not gene names, but the identifiers of the particular markers 
(e.g. Affymetrix probesets) from the expression platform used to collect
 the data.
</p><p><br>
</p>
<a name="Modulators_List"></a><h3> <span class="mw-headline" id="Modulators_List">Modulators List</span></h3>
<p>The list of candidate modulators can either be loaded from a file as a
 comma separated list, or a set of markers can be selected from the 
Markers component.  The gene expression profiles of the modulators 
should be independent of that of the hub TF gene as measured by mutual 
information.  This could be determined using a preliminary run of ARACNE
 including just the modulators and the transcription factor.
</p><p><br>
Modulators List pulldown menu options are:
</p>
<ul><li> <b>From File</b> - Load a list of candidate modulators from a file (containing a comma separated list).
</li><li> <b>From Set</b> - Select a set of candidate modulators defined in the Markers component.  When <b>From Set</b> is selected, entries can also be typed directly into the text box.
</li></ul>
<p><b>Note</b> - any markers in the modulator list must also appear in the target set (see Target List).
</p><p><br>
</p>
<a name="Target_List"></a><h3> <span class="mw-headline" id="Target_List">Target List</span></h3>
<p>The target list can include all markers or can be restricted to some 
subset of candidates e.g. thought to be regulated by the Hub Marker 
transcription factor. 
</p><p>Target List pulldown menu options are:
</p>
<ul><li> <b>All Markers</b> - Run MINDy on all markers in the data set.
</li><li> <b>From File</b> - Load a list of target markers from a file (containing a comma separated list).
</li><li> <b>From Set</b> - Select a target marker set defined in the Markers component.
</li></ul>
<p><br>
</p>
<ul><li> <b>Important - Target list must also include the Hub Marker and all Modulator markers</b> 
<ul><li> The MINDy main parameters tab requires the selection of 
Modulators, Targets, and a hub marker.  The Target List must also 
contain the Hub Marker and all the Modulator markers, because a single 
expression profile dataset is transferred to the algorithm for 
calculations.   
</li><li> If "All Markers" is chosen, then no further attention to this point is required.
</li></ul>
</li></ul>
<ul><li> Note - the "<b>All Markers</b>" <i>checkbox</i> at the bottom of the Analysis component should <b>not</b> be used in the MINDy component.
</li></ul>
<p><br>
</p>
<a name="Hub_Marker"></a><h3> <span class="mw-headline" id="Hub_Marker">Hub Marker</span></h3>
<p>Enter the marker ID for a known or putative transcription factor gene. 
</p>
<ul><li> The Hub marker can be entered directly in the text field, or 
the most recently selected marker in the Markers component will be used,
 selected either in the list or in the default Marker set "Selection". 
</li><li> <b>Note</b> - Even if one directly types in a marker name, it will be replaced if any selection is made in the Markers component.
</li><li> <b>Note</b> - The hub marker must also appear in the target set (see Target List).
</li></ul>
<p><br>
</p>
<a name="Parameters_-_Advanced"></a><h2> <span class="mw-headline" id="Parameters_-_Advanced">Parameters - Advanced</span></h2>
<p><img alt="MINDy Parameters Advanced .png" src="mindy_files/MINDy_Parameters_Advanced_.png" height="351" width="1210">
</p><p><br>
</p>
<a name="Sample_per_Condition_.28.25.29"></a><h3> <span class="mw-headline" id="Sample_per_Condition_.28.25.29">Sample per Condition (%)</span></h3>
<p>MINDy calculates the difference in mutual information for the 
TF-Target interaction between the set where the modulator gene is most 
expressed (+) and the set where the modulator gene is least expressed 
(-). This parameter specifies the percentage of the available samples to
 include in each group. E.g. 35% means that the top and bottom 35% of a 
list of samples ranked by expression would be used.
</p><p><br>
</p>
<a name="Conditional_.28threshold_settings.29"></a><h3> <span class="mw-headline" id="Conditional_.28threshold_settings.29">Conditional (threshold settings)</span></h3>
<p>The underlying ARACNe calculation of the conditional mutual 
information allows a threshold to be set.  The threshold for the 
conditional calculations can be specified as a raw mutual information 
score or as a P-value.  An above-threshold MI value must be obtained in 
at least one of the two conditional ARACNe runs in order for the target 
to be included in the output data.
</p><p>Options:
</p>
<ul><li> <b>Mutual Info </b>- If selected, the user specifies a 
threshold for the mutual information (MI) estimates in terms of the raw 
MI score. For example, a value of 0.1 filters out target genes with a MI
 score of less than 0.1 in both the high and low modulator expression 
sets.  By default, a MI threshold of 0.1 is set.
<ul><li> Note - if the MI score is above threshold in one condition but 
not the other, the lower score will be set to zero when calculating 
delta (MI).  
</li></ul>
</li><li> <b>P-value</b> - If selected, the user specifies a threshold 
for the conditional mutual information estimate in terms of a p-value.  
This is a value between 0 and 1, with 1 indicating no threshold. By 
default, the value is 0.01.  The specified p-value is converted to a MI 
threshold.
</li></ul>
<ul><li> <b>Correction</b> -  correct for multiple testing if a p-value is specified.  The choices are
<ul><li> None - no correction of the p-value
</li><li> Bonferroni -  apply the Bonferroni correction to the p-value before its is converted to a threshold MI score.
</li></ul>
</li></ul>
<ul><li> Note on p-value calculation in MINDy in geWorkbench - The 
p-value calculation for the conditional runs of ARACNe is calculated 
using an approximation described in Margolin et al., 2006.
</li></ul>
<a name="Unconditional_.28threshold_settings.29_.28Not_used_in_MINDy.29"></a><h3> <span class="mw-headline" id="Unconditional_.28threshold_settings.29_.28Not_used_in_MINDy.29">Unconditional (threshold settings) (Not used in MINDy)</span></h3>
<p>The unconditional MI is intended for use in the calculation of 
statistical significance of the final delta (MI) score and is not 
currently used.  This control is disabled.
</p><p><br>
</p><p><br>
</p>
<a name="ARACNe_parameter_files_not_supported_in_MINDy_in_geWorkbench"></a><h3> <span class="mw-headline" id="ARACNe_parameter_files_not_supported_in_MINDy_in_geWorkbench">ARACNe parameter files not supported in MINDy in geWorkbench</span></h3>
<p>ARACNe allows files config_threshold.txt and config_kernel.txt to be 
read in from disk if present.  However, the version of ARACNe used in 
MINDy does not support this feature.  It uses default parameters to 
derive the threshold and kernel width values.
</p>
<a name="Important_notes_on_the_calculation"></a><h2> <span class="mw-headline" id="Important_notes_on_the_calculation">Important notes on the calculation</span></h2>
<a name="delta_.28MI.29"></a><h3> <span class="mw-headline" id="delta_.28MI.29">delta (MI)</span></h3>
<p>As implemented in geWorkbench, the significance of the delta (MI) values is not calculated.
</p><p><br>
</p>
<a name="Marker_and_Array_Selection"></a><h3> <span class="mw-headline" id="Marker_and_Array_Selection">Marker and Array Selection</span></h3>
<ul><li> <b>Marker Sets</b> - All marker selection is done within the 
MINDy component interface.  If the option "From Sets" is chosen, one 
marker set from the Markers component can be selected.  MINDy <b>does not</b>
 respect activated marker subsets in the Markers component - that is, 
checking the box next to a marker subset in the Markers component has no
 effect on the markers used for the Mindy calculation or display.
</li></ul>
<ul><li> <b>Array Sets</b> - MINDy <b>does</b> respect array subsets 
activated in the Arrays component.  That is, the arrays used can be 
limited to particular subsets by activating those subsets in the Arrays 
component (by checking the boxes next to them).
</li></ul>
<ul><li> <b>Important - Target list must also include the Hub Marker and all Modulator markers</b> 
<ul><li> The MINDy main parameters tab requires the selection of 
Modulators, Targets, and a hub marker.  The Target List must also 
contain the Hub Marker and all the Modulator markers, because a single 
expression profile dataset is transferred to the algorithm for 
calculations.   
</li><li> If "All Markers" is chosen, then no further attention to this point is required.
</li></ul>
</li></ul>
<ul><li> <b>Testing of multiple modulators</b> - When testing multiple 
modulators, consider the false-positive implications of multiple tests, 
even though no significance value is being calculated.
</li></ul>
<a name="ARACNe_configuration_files"></a><h3> <span class="mw-headline" id="ARACNe_configuration_files">ARACNe configuration files</span></h3>
<p>The following discussion of configuration files applies only to the 
local version of MINDy, not the grid version.  On the grid version, only
 the default parameters for kernel width and threshold will be used.
</p><p>MINDy in geWorkbench uses the original, fixed-bandwidth version 
of ARACNe.  This version of ARACNe uses two configuration files, 
config_kernel.txt and config_threshold.txt.  If these two files are not 
supplied, default parameters will be used, which should be sufficient 
for most cases.  The parameter files can also be generated using ARACNe2
 in geWorkbench.  However, the files will be named after the dataset 
from which they are generated, and must be renamed to config_kernel.txt 
and config_threshold.txt to be seen by ARACNe.  Files with those names, 
if present in the geWorkbench installation root folder, will override 
any other dataset-specific configuration files for ARACNe2, and so 
should not be left on the system after MINDy has been run.
</p>
<a name="Services_.28Grid.29"></a><h1> <span class="mw-headline" id="Services_.28Grid.29">Services (Grid)</span></h1>
<p>MINDy can be run either locally within geWorkbench, or remotely as a grid job on caGrid.  See the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Tutorial_-_Grid_Services" title="Tutorial - Grid Services" class="mw-redirect"> Grid Services</a> section for further details on setting up a grid job.  A Columbia grid login must be obtained to use the Columbia grid service.
</p><p><br>
</p>
<a name="Running_an_example_MINDy_Analysis"></a><h1> <span class="mw-headline" id="Running_an_example_MINDy_Analysis">Running an example MINDy Analysis</span></h1>
<a name="Analysis_Framework"></a><h2> <span class="mw-headline" id="Analysis_Framework">Analysis Framework</span></h2>
<p>For general details on saving and storing parameter settings, and launching the analysis, see the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Tutorial_-_Analysis" title="Tutorial - Analysis" class="mw-redirect">Analysis</a> tutorial page.
</p><p><br>
</p>
<a name="Setup"></a><h2> <span class="mw-headline" id="Setup">Setup</span></h2>
<ul><li> For this example, we use a list of four candidate MAPK markers,
 contained in a CSV format file. Right-click on the following link and 
save the file <a href="http://wiki.c2b2.columbia.edu/workbench/images/4/4f/Mapk_list.csv" class="internal" title="Mapk list.csv">Mapk_list.csv</a> to disk.
</li><li> In the Component Configuration Manager, check whether the MINDy component has been loaded, and if not, load it.
</li></ul>
<p><br>
</p>
<a name="Run"></a><h2> <span class="mw-headline" id="Run">Run</span></h2>
<p>The figure illustrates the MINDy main parameter tab set up to run the example below.
</p><p><br>
Modulator list loaded from file:
</p><p><br>
<img alt="MINDy parameters mapk run.png" src="mindy_files/MINDy_parameters_mapk_run.png" height="314" width="1195">
</p><p><br>
Modulator list loaded from Marker set:
</p><p><br>
<img alt="MINDy parameters mapk run fromset.png" src="mindy_files/MINDy_parameters_mapk_run_fromset.png" height="294" width="1191">
</p><p><br>
</p>
<ol><li> Load the Bcell-100.exp microarray dataset, which is available 
in the geWorkbench data directory under "public_data".  If you wish to 
see gene names in the results, you must also load the associated 
annotation file.  See e.g. the tutorial <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Local_Data_Files" title="Local Data Files">Local Data Files</a> for further details.
</li><li> In the analysis tab (at lower right in the application), select<b> MINDy Analysis</b>.
</li><li> In the MINDy Parameters Main tab, populate the <b> Modulators List</b> by loading the file <a href="http://wiki.c2b2.columbia.edu/workbench/images/4/4f/Mapk_list.csv" class="internal" title="Mapk list.csv">Mapk_list.csv</a>.
  Or, the file can first be loaded into the Markers component with the 
"Load Set" button, then selected as "From Sets" in the Mindy parameters.
</li><li> Populate the <b>Target List</b> textbox by selecting the choice "All Markers".   
</li><li> Set the hub gene to be marker (probeset) name 37724_at (MYC). 
 Type in the marker name directly, or search for and select it in the 
Markers component. 
</li><li> Parameter values for the conditional mutual information 
calculation can be set in the Advanced Tab.  The values will depend on 
the specifics of the data set being used, in terms of number of arrays 
and number of markers.   Here we use the default parameters:
<ol><li> Sample per Condition: 35%
</li><li> Conditional: MI 0.1
</li><li> Unconditional: not used, control disabled.
</li><li> DPI target list: not used, control disabled.
</li><li> DPI tolerance: not used, control disabled.
</li></ol>
</li><li>Click <b>Analyze</b>. If successful, the Project Folders 
component is updated to add the MINDy result node.  The result node is 
shown as a child node of the input dataset Bcell-100.exp. Please note 
that the Dataset History tab captures the analysis parameters.
</li></ol>
<a name="Viewing_MINDy_Results"></a><h1> <span class="mw-headline" id="Viewing_MINDy_Results">Viewing MINDy Results</span></h1>
<a name="General"></a><h2> <span class="mw-headline" id="General">General</span></h2>
<p>1. The MINDy result node should be automatically selected in the 
Project Folder once the result is available. If not, select it.  This 
will display the MINDy result viewer.
</p><p>2. In the Modulator Tab, indicate the modulators of interest using the checkboxes or click on <b>Select All</b> to display all modulators in the Table, List, and Heat Map views. The <b>Modulators Selected</b>
 is updated to reflect the number of modulators selected. Only selected 
Modulators are displayed on the Table, List and Heat Map views. 
</p><p><br>
</p>
<a name="Common_Features"></a><h2> <span class="mw-headline" id="Common_Features">Common Features</span></h2>
<a name="Net_modulatory_effect_values"></a><h3> <span class="mw-headline" id="Net_modulatory_effect_values">Net modulatory effect values</span></h3>
<p>The first step of the MINDy algorithm is to sort the input expression
 arrays by the expression value for the candidate modulator.  It then 
forms two groups of arrays, those where the candidate modulator is most 
highly expressed, and those where it is least expressed.  Here we will 
refer to these as the "conditional high set" and the "conditional low 
set" - that is, they are sets of arrays conditioned on the expression of
 the candidate modulator.
</p><p>The following symbols are used to break out the total, positive, and negative effects:
</p>
<ul><li> <b>M#</b> - For reach modulator, the total number of above-threshold transcription-factor-target (TF-Ti) MI scores found.
</li><li> <b>M+</b> - The number of targets for which the TF-Ti pairs 
showed higher MI in the conditional high set compared with the MI in the
 conditional low set.
</li><li> <b>M-</b> - The number of targets for which the TF-Ti pairs 
showed lower MI in the conditional high set compared with the MI in the 
conditional low set.
</li></ul>
<p><br>
</p>
<a name="Controls"></a><h3> <span class="mw-headline" id="Controls">Controls</span></h3>
<a name="Marker_Display"></a><h4> <span class="mw-headline" id="Marker_Display">Marker Display</span></h4>
<p>Controls how the marker name is displayed.  Options are:
</p>
<ul><li> <b>Symbol</b> - If an annotation file has been loaded, use the Gene Symbol associated with each marker. 
</li><li> <b>Probe Name</b> - Use the marker probe name as given in the dataset.
</li></ul>
<a name="Add_to_Set"></a><h4> <span class="mw-headline" id="Add_to_Set">Add to Set</span></h4>
<p>(Except Heat Map) -  Adds selected markers to a Marker Set. You can 
select one or more Targets and/or Modulators, using the selection 
check-boxes.
</p><p><br>
</p>
<a name="Export"></a><h4> <span class="mw-headline" id="Export">Export</span></h4>
<p>The results shown in the Modulator, Table, or List tabs can be 
exported to a CSV format file on disk using the "Export" button. Only 
the table in the  currently displayed tab is exported.
</p>
<a name="Displayed_targets_filter"></a><h3> <span class="mw-headline" id="Displayed_targets_filter">Displayed targets filter</span></h3>
<p>This menu is located at the bottom of the MINDy viewer component and 
controls the target markers displayed in the various view tabs just 
described. It contains a list of all markers sets available in the 
Markers component.  Any one set can be chosen, and only MINDy targets 
which are also in this selected subset will be displayed (the 
intersection of the MINDy result set and the Marker set). 
</p><p><b>Note</b> - Marker sets do not need to be activated to be used for result filtering here.
</p>
<a name="Values"></a><h4> <span class="mw-headline" id="Values">Values</span></h4>
<ul><li> <b>All non-zero markers</b> - all markers with delta (MI) &gt; 0 are displayed.
</li><li> <b>Selection</b> - This refers to the default "Selection" set in the Markers component. 
</li><li> any other marker set name - all available marker sets will be listed in the menu and any one can be chosen.
</li></ul>
<p><br>
</p>
<a name="Modulator_tab"></a><h2> <span class="mw-headline" id="Modulator_tab">Modulator tab</span></h2>
<p><img alt="MINDy mapk initial result modulator tab.png" src="mindy_files/MINDy_mapk_initial_result_modulator_tab.png" height="464" width="936">	
</p><p>This table-based view contains one row per modulator gene.  It 
summarizes the results, and is used to control the targets displayed in 
the other view tabs.
</p>
<a name="Controls_2"></a><h3> <span class="mw-headline" id="Controls_2">Controls</span></h3>
<ul><li> <b>List Selections</b>
<ul><li> <b>Select All</b> checkbox - When checked, all modulators will 
be selected.  If not checked, the individual markers can be selected 
using the individual check boxes in the table.
</li></ul>
</li><li> <b>Modulators selected</b> - Shows a count of the number of individual modulators that have been selected in the table.
</li></ul>
<p><br>
</p>
<a name="Columns"></a><h3> <span class="mw-headline" id="Columns">Columns</span></h3>
<ul><li> M#, M+ and M- have already been described above under <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Tutorial_-_MINDy#Net_modulatory_effect_values" title="Tutorial - MINDy" class="mw-redirect">"Net modulatory effect values"</a>. 
</li></ul>
<ul><li> <b>Check-boxes</b> - Use these to select which modulators to include in generating the data views on the other tabs (Table, List, and Heat Map). 
</li><li> <b>Modulator</b> - The gene symbol or probe name for the putative modulators tested.
</li></ul>
<ul><li> <b>Mode</b> - Shows whether the net sum effect of the modulator over all its targets was enhancing or negative. 
<ul><li>  If M+ - M- &gt; 0, the result is "+", that is the candidate had a net positive modulatory effect (increased MI).
</li><li>  If M+ - M- &lt; 0, the result is "-", that is the candidate had a net negative modulatory effect (decreased MI).
</li><li>  If M+ - M- = 0, the result is "0", that is the candidate modulator had a balanced effect.
</li></ul>
</li></ul>
<p><br>
Here all four modulators in the example have been selected, activating the other tabs.
</p><p><img alt="MINDy mapk initial result modulator tab select all.png" src="mindy_files/MINDy_mapk_initial_result_modulator_tab_select_all.png" height="381" width="449">
</p><p><br>
</p>
<a name="Table"></a><h2> <span class="mw-headline" id="Table">Table</span></h2>
<p><img alt="MINDy mapk Table.png" src="mindy_files/MINDy_mapk_Table.png" height="864" width="935">
</p><p><br>
The column "Target" represents the target genes and the remaining columns represent the modulators tested. 
</p>
<ul><li> <b>Discretization of scores</b> - By default, the MI scores are
 discretized to +1 and -1 for positive and negative scores, 
respectively. Discretized scores are used to quantify the number of 
positive and negative modulation effects, as shown e.g. in the numbers 
in the column headers. If the "Score View" option is chosen, the actual 
scores will be shown.  
</li></ul>
<p><br>
</p>
<a name="Controls_3"></a><h3> <span class="mw-headline" id="Controls_3">Controls</span></h3>
<p>If many modulators were tested, it may be desirable to sort the display by their results.
</p>
<ul><li> <b>Display Options:</b>
<ul><li> <b>Color View:</b> Enables a heat map display of each cell 
based on the value of the score.  Positive values are displayed in 
shades of red, while negative values are displayed in shades of blue. 
The saturation of the color increases (starting from white for 0) with 
increasing absolute value of the score.
</li><li> <b>Score View:</b> Displays the actual score values rather than the default discretized values.
</li></ul>
</li></ul>
<p><br>
Here, both the "Color View" and "Score View" options have been checked.
</p><p><br>
<img alt="MINDy mapk Table display options1.png" src="mindy_files/MINDy_mapk_Table_display_options1.png" height="863" width="932">
</p><p><br>
The modulators in the above figure are by default sorted by the 
aggregate count of targets for which a modulatory effect was seen (M#).
</p><p>The table can be sorted on the values in any column by clicking 
on its header.  In addition, the column headers (the modulators) can be 
sorted as described next:
</p>
<ul><li> <b>Modulator Sorting:</b> - Displays columns (modulators) from 
left to right in descending order by the counts of: Aggregate ( M#), 
Enhancing (M+) or Negative (M-).
<ul><li> <b>Aggregate (M#)</b>: The column header displays "M#" and the count of all targets for which a positive or negative modulatory effect was seen. 
</li><li> <b>Enhancing (M+)</b>: The column header displays "M+" and the count of all targets for which a positive modulatory effect was seen. 
</li><li> <b>Negative (M-)</b>: The column header displays "M-" and the count of all targets for which a negative modulatory effect was seen. 
</li></ul>
</li></ul>
<p>Example of sorting by "Enhancing":  The first modulator column header is "MAP4K4 (M+ 103).
</p><p><br>
<img alt="MINDy mapk Table display options enhancing.png" src="mindy_files/MINDy_mapk_Table_display_options_enhancing.png" height="298" width="1114">
</p><p><br>
Example of sorting by "Negative":  The first modulator column header is "MAP4K2 (M- 54).
</p><p><br>
<img alt="MINDy mapk Table display options negative.png" src="mindy_files/MINDy_mapk_Table_display_options_negative.png" height="296" width="1111">
</p><p><br>
</p>
<ul><li> <b>Modulator Limits:</b> When the checkbox is selected, the number of columns (modulators) is limited to the value set in the selector box.
</li></ul>
<ul><li> <b>Marker Selection</b> 
<ul><li> <b>Enable Selection</b> - When checked, a column of checkboxes 
appears in the table to allow individual selection of targets.  Shows a 
count of all selected modulators and targets.
</li><li> <b>All Modulators</b> - Select or Clear buttons - Selects or clear all modulator check boxes (table columns).
</li><li> <b>All Targets</b> - Select or clear buttons - Select or clear all displayed targets (table rows).
</li><li> <b>Add to Set</b> (button) - All selected markers (modulators and/or targets) will be added to a new set in the Markers component.
</li></ul>
</li></ul>
<p><br>
</p>
<ul><li> <b>Displayed targets filter</b>  - The displayed target markers
 can be filtered using marker set defined in the Markers component.  
After the filter is selected, only those markers contained in the 
selected set will appear.
</li></ul>
<a name="List"></a><h2> <span class="mw-headline" id="List">List</span></h2>
<p>In the list view, all modulators are listed in the first column, 
their targets in the second column, while the third column contains the 
delta (MI) scores.  That is, each modulator/target pair is listed 
individually.  
</p><p>This view has the advantage of displaying only actual data 
values.  This contrasts with the Table view, where the results are 
displayed in a spreadsheet format.  Because each modulator will have its
 own set of targets, not each modulator/target cell in the table will 
have a value.  Results in the Table view are padded with zeros as 
necessary.
</p><p><br>
<img alt="MINDy mapk List tab.png" src="mindy_files/MINDy_mapk_List_tab.png" height="472" width="695">
</p><p><br>
</p>
<ul><li> <b>Marker Selection</b> (checkbox) - Controls which markers are used by the "Add to Set" button.
<ul><li> <b>Enable Selection</b> - When checked, a column of checkboxes 
appears beside each target and beside each marker to allow individual 
selection of each.  Shows a count of all selected modulators and 
targets.
</li><li> <b>Select all Modulators</b> - Selects all modulators.
</li><li> <b>Select all Targets</b> - Selects all targets.
</li><li> <b>Add to Set</b> (button) - All selected markers will be added to a new set in the Markers component.
</li></ul>
</li></ul>
<p><br>
</p>
<a name="Heat_Map"></a><h2> <span class="mw-headline" id="Heat_Map">Heat Map</span></h2>
<p>The Heat Map represents the expression values for individual markers 
(target genes).  It contains two color mosaic panels. The rows 
correspond to target genes and are ordered according to their Pearson's 
correlation to the expression of the TF.   The columns (arrays) are 
ordered according to the expression of the TF gene, low (left) to high 
(right). The mosaic at left corresponds to the arrays where modulator 
was least expressed.  The mosaic at right corresponds to the arrays 
where the modulator expression was highest. 
</p>
<a name="Controls_4"></a><h3> <span class="mw-headline" id="Controls_4">Controls</span></h3>
<ul><li> <b>Transcription Factor:</b> Displays the TF hub gene entered in the MINDy Analysis parameters.
</li></ul>
<ul><li> <b>Modulators: </b> - The heat map is generated for the targets
 of only one modulator at a time.  The list shows the available 
modulators, and the text box above it shows the selected modulator.
</li></ul>
<p>Here the first modulator on the list is selected:
</p><p><br>
<img alt="MINDy mapk Heat Map tab.png" src="mindy_files/MINDy_mapk_Heat_Map_tab.png" height="497" width="1127">
</p><p><br>
As shown below, scrolling to the bottom of the Heat Map image shows how 
the effect of modulation can differ for different genes.  The genes at 
top are directly correlated with MYC when MAP4K4 is low, whereas the 
genes at bottom are anti-correlated.
</p><p><br>
<img alt="MINDy mapk Heat Map tab MAP4K4 lower.png" src="mindy_files/MINDy_mapk_Heat_Map_tab_MAP4K4_lower.png" height="498" width="1127">
</p><p><br>
</p>
<ul><li> <b>Displayed targets filter</b> - The targets displayed in the Heat Map view can be limited to those defined in a marker set.  
</li></ul>
<ul><li> <b>Image Snapshot:</b> - Captures the Heat Map as an image node in the Project Folder.
</li></ul>
<p><br>
<img alt="MINDy Heat Map node.png" src="mindy_files/MINDy_Heat_Map_node.png" height="155" width="231">
</p>
<a name="References"></a><h1> <span class="mw-headline" id="References">References</span></h1>
<ol><li> Margolin, A., Wang, K., Lim, W.K., Kustagi, M., Nemenman, I., 
and Califano, A. (2006) Reverse Engineering Cellular Networks. Nature 
Protocols 1(2):662-671. <a href="http://www.ncbi.nlm.nih.gov/pubmed/17406294" class="external text" rel="nofollow">link to pub.</a>.
</li><li> Wang K, Saito M, Bisikirska BC, Alvarez MJ, Lim WK, 
Rajbhandari P, Shen Q, Nemenman I, Basso K, Margolin AA, Klein U, 
Dalla-Favera R, Califano A. (2009)  Genome-wide identification of 
post-translational modulators of transcription factor activity in human B
 cells.  Nat Biotechnol. 27(9):829-39. <a href="http://www.ncbi.nlm.nih.gov/pubmed/19741643" class="external text" rel="nofollow">link to pub.</a>.
</li></ol>

<!-- 
NewPP limit report
Preprocessor node count: 162/1000000
Post-expand include size: 1907/2097152 bytes
Template argument size: 0/2097152 bytes
Expensive parser function count: 0/100
-->

<!-- Saved in parser cache with key workbench:pcache:idhash:1842-0!1!0!!en!2!printable=1 and timestamp 20120309163439 -->
<div class="printfooter">
Retrieved from "<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/MINDy">http://wiki.c2b2.columbia.edu/workbench/index.php/MINDy</a>"</div>
		<div id="catlinks" class="catlinks catlinks-allhidden"></div>		<!-- end content -->
				<div class="visualClear"></div>
	</div>
</div></div>
<div class="visualClear"></div>
<div id="footer">
	<div id="f-poweredbyico"><img src="mindy_files/poweredby_mediawiki_88x31.png" alt="Powered by MediaWiki" height="31" width="88"></div>
	<ul id="f-list">
		<li id="lastmod"> This page was last modified on 23 February 2012, at 18:53.</li>
	</ul>
</div>
</div>

