<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html dir="ltr" lang="en"><head>
<title>Differential Expression - Workbench</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="generator" content="MediaWiki 1.16.0">

</head>
<body class="mediawiki ltr ns-0 ns-subject page-Differential_Expression skin-monobook">
<div id="globalWrapper">
<div id="column-content"><div id="content">
	<a id="top"></a>
	
	<h1 id="firstHeading" class="firstHeading">Differential Expression</h1>
	
		<!-- start content -->
<p><br> 
</p><p><br>
</p>
<a name="Overview"></a>
<h1> <span class="mw-headline" id="Overview">Overview</span></h1>
<p>A t-Test analysis can be used to identify markers with statistically 
significant differential expression between two sets of microarrays.  In
 geWorkbench, these groups are specified as the "Case" and "Control" 
sets.  
</p><p>There are several steps to setting up a t-test analysis in geWorkbench.
</p>
<ol><li> At least two sets of arrays must be available in the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Data_Subsets_-_Arrays" title="Data Subsets - Arrays"> Arrays</a> component.
</li><li> The array sets to be used in the analysis must be "activated" by checking the box adjacent to their names in the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Data_Subsets_-_Arrays" title="Data Subsets - Arrays"> Arrays</a> component.
</li><li> One or more activated array sets must be designated "Case", and the others "Control" (which is the default classification).
</li><li> The t-test parameters must be set.
</li></ol>
<p><br>
After the t-test is run, the results will be displayed graphically, and 
all markers meeting the significance threshold are placed into a new 
Marker Set called "Significant Genes".
</p><p>Please see the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Differential_Expression#Example" title="Differential Expression"> Example</a> section below for instructions on preparing array sets for the t-test analysis.
</p>
<a name="t-Test_Parameters"></a>
<h1> <span class="mw-headline" id="t-Test_Parameters">t-Test Parameters</span></h1>
<h2> <span class="mw-headline" id="P-value_Parameters">P-value Parameters</span></h2>
<h3> <span class="mw-headline" id="p-values_based_on">p-values based on</span></h3>
<p>The p-values can be calculated by transforming the t-statistic 
directly, or by carrying out a permutation analysis.  The permutation 
analysis measures how often a t-statistic at least as large as that 
observed occurs by chance after array labels of case and control are 
permuted.
</p>
<ul><li> <b>t-distribution</b> (the default)
</li><li> <b>Permutation</b> -  If chosen, the number of permutations to carry out must also be specified.
</li></ul>
<h3> <span class="mw-headline" id="Overall_alpha_.28Critical_p-value.29">Overall alpha (Critical p-value)</span></h3>
<p>The threshold for a difference in expression between Case and Control
 sets being called significant.  A value of 0.05 is often used for a 
single test.  Multiple-testing corrections can be specified in the Alpha
 Corrections tab.
</p>
<h3> <span class="mw-headline" id="Data_is_Log2-transformed">Data is Log2-transformed</span></h3>
<p>If the dataset has been Log2 transformed, check this box.  Having 
this information allows the fold-change displayed in the Volcano Plot to
 be calculated in a consistent fashion.
</p><p>The system will examine the current dataset and make a guess as 
to whether the data has been log2 transformed.  The user can override 
this guess using the check box. 
</p><p><br>
  <img alt="T-test Pvalue params.png" src="Ttest_files/T-test_Pvalue_params.png" height="399" width="1001">
</p>
<p><br>
</p>
<h2> <span class="mw-headline" id="Alpha_corrections">Alpha corrections</span></h2>
<p>For multiple testing (alpha) correction, the following options are offered:
</p>
<ul><li> <b>no correction</b>
</li><li> <b>Standard Bonferonni Correction</b> - the value of alpha is divided by the number of markers included in the analysis.
</li><li> <b>Adjusted (step down) Bonferonni Correction</b>
</li><li> Two variants of the Westfall and Young method are available if the p-value is estimated by permuation:
<ul><li> <b>minP</b>
</li><li> <b>maxT</b>
</li></ul>
</li></ul>
<p><br>
  <img alt="T-test alpha corrections.png" src="Ttest_files/T-test_alpha_corrections.png" height="216" width="1005">
</p>
<h2> <span class="mw-headline" id="Degrees_of_Freedom">Degrees of Freedom</span></h2>
<p>Group variances can be declared as:
</p>
<ol><li> unequal (Welch approximation) (default)
</li><li> Equal.
</li></ol>
<p><br>
  <img alt="T-test degrees of freedom.png" src="Ttest_files/T-test_degrees_of_freedom.png" height="223" width="1004">
</p>
<a name="Example"></a>
<h1> <span class="mw-headline" id="Example">Example</span></h1>
<h2> <span class="mw-headline" id="Preparation">Preparation</span></h2>
<p>Obtain the file "BCell-100.exp", which is contained in the 
data/public_data directory of the geWorkbench distribution, or can be 
directly downloaded from the tutorial data <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Download_and_Installation#Tutorial_data" title="Download and Installation"> download</a> area.
</p><p>You may also wish to load the Affymetrix HG-U95Av2 annotation file, although it is not required for this example.  See the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/FAQ" title="FAQ"> FAQ</a> section for information on downloading this file from Affymetrix.
</p><p><br>
For tips on loading data files, see <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Local_Data_Files" title="Local Data Files"> Local Data Files</a> and <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Projects" title="Projects"> Projects</a>.
</p><p>In this example, we apply two normalization steps to the data set.
</p>
<ol><li> <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Normalization#Threshold_Normalizer" title="Normalization"> Threshold Normalizer</a> - set a minimum value of 1.  Any value less than 1 will be set to 1.
</li><li> <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Normalization#Log2_Transformation" title="Normalization"> Log2 Transformation Normalizer</a> - Log2 transform the data.
</li></ol>
<p>For an actual data analysis, you should apply data normalization steps appropriate to your own data and analysis design.
</p>
<h2> <span class="mw-headline" id="Array_Classification">Array Classification</span></h2>
<p>The t-test in geWorkbench requires that at least two sets of arrays 
be "activated".  Only such "activated" sets are considered.  In 
addition, at least one such set must be designated as "Case", and at 
least one other as "Control" (which is the default classification).   
Note that more than one set of arrays can be marked as "Case" or 
control.  
</p><p>Array set classification is covered in the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Data_Subsets_-_Arrays" title="Data Subsets - Arrays">Arrays/Phenotypes</a> chapter.  However, for convenience, the steps are illustrated here.
</p><p>The desired sets of arrays should be activated in the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Data_Subsets_-_Arrays" title="Data Subsets - Arrays">Arrays/Phenotypes</a> component.  This is done by checking the boxes by the desired Sets.
</p><p><img alt="T-test Set activation BCell.png" src="Ttest_files/T-test_Set_activation_BCell.png" height="410" width="228">
</p>
<p><br>
</p><p>The classification can be made directly by left-clicking on the "thumb-tack" icon adjacent to an array set name.
</p><p><img alt="T-test Set classification left click Bcell.png" src="Ttest_files/T-test_Set_classification_left_click_Bcell.png" height="239" width="228">
</p>
<p><br>
The array classification can also be set by right-clicking on the desired array set and selecting "Classification":
</p><p><br>
  <img alt="T-test Set classification right click Bcell.png" src="Ttest_files/T-test_Set_classification_right_click_Bcell.png" height="347" width="335">
</p>
<p><br>
Using either method, the desired array set can be classified as "Case":
</p><p><br>
  <img alt="T-test Set selection BCell.png" src="Ttest_files/T-test_Set_selection_BCell.png" height="411" width="230">
</p>
<p><br>
The thumbtack image next to activated Array Sets is colored red.
</p>
<h2> <span class="mw-headline" id="Seting_the_Analysis_Parameters">Seting the Analysis Parameters</span></h2>
<ol><li> The t-test component should be loaded by default in the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Component_Configuration_Manager" title="Component Configuration Manager">Component Configuration Manager</a>.
</li><li> From the Analysis Panel, select <b>T-Test Analysis</b>. 
</li><li> P-value Parameters tab:
<ol><li> P-values based on t-distribution.
</li><li> Note that here the default alpha (critical p-value) is set to 0.01.
</li><li> Check-mark the box "Data is log2 Transformed".
</li></ol>
</li><li> Alpha-corrections tab
<ol><li>  Standard Bonferonni
</li></ol>
</li><li> Degree of Freedom tab
<ol><li> Welch approximation - unequal group variances.
</li></ol>
</li></ol>
<p><br>
The P-value Parameters tab set for the example analysis:
</p><p><br>
  <img alt="T-test Example setup.png" src="Ttest_files/T-test_Example_setup.png" height="280" width="1003">
</p>
<h2> <span class="mw-headline" id="Running_the_t-test_analysis">Running the t-test analysis</span></h2>
<ol><li> Click <b>Analyze</b>.  The results will be returned in three locations: The Project Folder, the Markers component, and the Visualization area.
</li></ol>
<a name="t-Test_Results"></a>
<h1> <span class="mw-headline" id="t-Test_Results">t-Test Results</span></h1>
<h2> <span class="mw-headline" id="Result_Sets">Result Sets</span></h2>
<p>A t-test result node is placed into the Projects Folder as a child of
 the microarray dataset that was analyzed (upper red arrow).
</p><p>The list of significant markers is placed into a new set in the 
Markers component (lower red arrow).  This set is labeled "Significant 
Genes".  The number in square brackets indicates the number of markers 
in the set.
</p><p><br>
  <img alt="T-test Example Result Full.png" src="Ttest_files/Volcano_plot.png">
</p>
<p><br>
</p>
<a name="Volcano_Plot_Visualizer"></a>
<h2> <span class="mw-headline" id="Volcano_Plot_Visualizer">Volcano Plot Visualizer</span></h2>
<p>The results are displayed in Volcano Plot visualizer (seen at right in the figure above).
</p><p><b>Axes</b> - The log2 fold change for each marker is plotted against the -log10 of the P-value.
</p><p><b>Fold Change</b> - The fold change is calculated, for each 
marker, as the average expression in the Case set divided by the average
 expression in the control set. If the data is known to already be log2 
transformed, the antilog is first taken before averaging. The log2 of 
the ratio is then taken for use in the display.
</p><p><b>Color scheme</b> - the lower 2/3 of the absolute values of the
 (fold change) * (significance) are colored in shades from light blue 
(lowest values) to dark blue (highest values).  The highest 1/3 of such 
values are colored from dark blue (lowest values) to red (highest 
values).
</p><p><br>
<b>Export Data</b> - The data depicted in the volcano plot is exported 
to a CSV format file.  The column headers are: Probe Set Name, Gene 
Name, p-Value, Fold Change.
</p>
<a name="Color_Mosaic_Visualizer"></a>
<h2> <span class="mw-headline" id="Color_Mosaic_Visualizer">Color Mosaic Visualizer</span></h2>
<p>The Color Mosaic tab shows all of the arrays and the p-value 
calculated for each marker.  It also can display annotation for each 
marker.
</p><p><br>
</p><p><img alt="T-test Example Result Color Mosaic.png" src="Ttest_files/T-test_Example_Result_Color_Mosaic.png" height="472" width="1128">
</p>
<p><br>
</p>
<h2> <span class="mw-headline" id="Color_Mosaic_Controls">Color Mosaic Controls</span></h2>
<p><img alt="T-Test Color Mosaic Controls.png" src="Ttest_files/T-Test_Color_Mosaic_Controls.png" height="36" width="1109">
</p>
<ul><li> <b>Printer Icon</b> - print
</li><li> <b>Display</b> - Turn the display on or off
</li><li> <b>Array names</b> - When active, display the array names along the top of the color mosaic display.
</li><li> <b>Accession</b> - When active, display the accession ids (probeset ids) for each marker at the right edge of the color mosaic.
</li><li> <b>Label</b> - When active, display the gene name for the marker is present in the annotation file.
</li><li> <b>Sort</b> - reverse the sort order based on the p-values.
</li><li> <b>Lightbulb</b> <img alt="Color Mosaic Lightbulb.png" src="Ttest_files/Color_Mosaic_Lightbulb.png" height="24" width="28">
 - Activate the tooltip function over the color mosaic display.  The 
tooltip displays the array name, the marker id, and the signal value 
from the array for the cell below the cursor.
</li>
<li> <b>Search fields</b> - These three search fields feature 
progressive search.  As each character is typed in, the first match to 
the entered string will be automatically highlighted. The next match can
 be found by hitting the "Enter" key.  Typing additional characters 
further narrows the search.  Each search field is active only when the 
label it searches being displayed (corresponding button is activated).
<ul><li> <b>Search Array</b> - Searches Array Names when the <b>Array names</b> button is on.
</li><li> <b>Search Accession</b> - Searches Marker IDs when the <b>Accession</b> button is on.
</li><li> <b>Search Label</b> - Searches Gene Names when the <b>Label</b> button is on.
</li></ul>
</li></ul>
<p>The figure below shows the effects of activating several of the display controls.
</p><p>The controls for display of Array Name, Accession ID (probeset id) and the lightbulb tooltip control have been activated. 
</p><p>Search fields for each activated label type are enabled.
</p><p>The tooltip displays the Array Name (chip), the marker id, and 
the signal value for the selected cell (array and marker) in the Color 
Mosaic display.
</p><p><br>
  <img alt="T-Test Color Mosaic Control Descriptions.png" src="Ttest_files/T-Test_Color_Mosaic_Control_Descriptions.png" height="471" width="1127">
</p>
<p><br>
The figure below shows the results, with labels highlighted in purple, of searches using all three search fields.
</p><p><br>
  <img alt="T-Test Color Mosaic Search.png" src="Ttest_files/T-Test_Color_Mosaic_Search.png" height="281" width="1006">
</p>


</body></html>