<!DOCTYPE HTML PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html dir="ltr" xml:lang="en" xmlns="http://www.w3.org/1999/xhtml" lang="en"><head>
<head>

  
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <meta name="KEYWORDS" content="Tutorial - Master Regulator Analysis">

    <title>Master Regulator Analysis</title>

	<link rel="stylesheet" type="text/css" href="ncicb.css">

</head>

<body>
<a name="top" id="top"></a>
<h1>Master Regulator Analysis</h1>


<a name="Overview"></a><h2>Overview</h2>
<p>The goal of Master Regulator Analysis (MRA)is to identify
transcription factors (TFs) which control the regulation of a set of
target genes (TGs) that demonstrate significant differential expression
across two cellular phenotypes, e.g. &quot;Case&quot; and &quot;Control&quot; in a
microarray dataset. Differential expression is measured using a simple
t-test. Sets of genes putatively controlled by each TF (each TF's
regulon) are obtained from an adjacency matrix (interaction network)
calculated by <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Tutorial_-_ARACNE" title="Tutorial - ARACNE"> ARACNe</a> or other source prior to MRA.<br>
</p>
<p>The dataset from which the adjacency matrix is derived would not
necessarily be the same one used for the t-test. An ARACNe run requires
a dataset which explores many different expression phenotypes of a
particular cell type, whereas a differential expression experiment
compares only two classes.<br>
</p>
<p>For each TF, MRA then calculates, using Fisher's Exact test,
whether there is greater overlap between the set of the TF's target
genes and the set of differentially expressed genes than would be
expected by chance. </p><p>The types of data which will be used in the MRA then are:
</p>
<ol><li> A microarray dataset appropriate for examining differential gene expression using a t-test. 
</li><li> A list of putative transcription factors which are to be tested against the differentially expressed genes.
</li><li> An interaction network in the form of an ARACNe adjacency
matrix. It should contain the results of an ARACNe run including, as
hub markers, at least all of the transcription factors that will be
tested in MRA.
</li></ol>
<a name="MRA Parameters_and_Settings"></a>
<h2>MRA  settings</h2>
<a name="Load_Network"></a><h3>Load Network</h3>
<p>The network consists of an adjacency matrix generated by ARACNe.  
</p>
<ul><li> <b>From File</b> - load an adjacency matrix generated by an external run of ARACNe. 
</li><li> <b>From Project</b> - load an ARACNe adjacency matrix from a result node in the Project Folders component.
</li></ul>
<a name="Transcription_Factors"></a><h3>Transcription Factors</h3>
<ul><li> <b>From File</b> - Load a comma-separate list of transcription factors from a file.
</li><li> <b>From Sets</b> - Use a set defined in the Markers component as the list of transcription factors.
</li></ul>
<a name="Significance_Threshold"></a><h3>Significance Threshold</h3>
<ul><li> <b>T-test p-value (alpha)</b> - The cutoff p-value by which to
establish whether a particular marker shows a significant difference in
expression between the two groups. (Note that multiple testing
corrections are offered on the t-test parameters tab).
</li></ul>
<p><br>
</p><p> <img src="MasterRegulatorAnalysis_files/T_MRA_Setup.png"  width="1116" height="383">
</p><p><br>
</p>
<a name="t-test"></a>
<h2>t-test Parameters and Settings </h2>
<p>The parameter settings available for the MRA t-test are shown in the
figure below. These parameters are the same as those described in the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Tutorial_-_Differential_Expression" title="Tutorial - Differential Expression"> t-test component tutorial</a>.
</p>
<a name="P-values_based_on"></a><h3>P-values based on</h3>
<ul>
  <li> <strong>t-distribution</strong> - directly calculate the p-value
</li><li> <strong>Permutation</strong> - determine the p-value empirically through repeated trials against permuted data sets.
<ul><li> Randomly group experiments  - #-times - how many permuations to carry out 
</li><li> All permutations
</li></ul>
</li></ul>
<a name="Correction_method"></a><h3>Correction method</h3>
<ul><li> <strong>Just alpha</strong> (no correction)
</li><li> <strong>Standard Bonferroni</strong> - divide given p-value threshold by number of markers tested.
</li><li> <strong>Adjusted Bonferroni</strong> - same as Standard Bonferroni, except the divisor for each successive marker tested is decreased by one.
</li></ul>
<p>&nbsp;</p>
<a name="Step-down_Westfall_and_Young_methods"></a><h3>Step-down Westfall and Young methods</h3>
<p>(only if permuation is selected for p-value calculation)
</p>
<ul><li><strong> minP
</strong></li><li><strong> maxT
</strong></li></ul>
<a name="Group_Variances"></a><h3>Group Variances</h3>
<p>Choose whether the variances in the two groups being compared are expected to be equal or not.
</p>
<ul><li> <strong>Unequal</strong> (Welch approximation)
</li>
<li><strong> Equal
</strong></li>
</ul>

<img src="MasterRegulatorAnalysis_files/T_MRA_t-test.png" />

<a name="Multiple_testing_considerations"></a><h2>Multiple testing considerations</h2>
<ul><li> <b>t-test</b> - The t-test for differential expression is run
on each marker in turn, so that potentially thousands of tests may be
performed. The t-test tab within MRA offers simple multiple testing
corrections such as the Bonferroni correction. </li></ul>
<ul><li> <b>Fisher's Exact Test</b> - Note that Fisher's Exact test is
run for each transcription factor and a p-value reported. No correction
is supplied for this occurrence of multiple testing.
</li></ul>
<a name="Running_MRA"></a><h2>Running MRA</h2>
<ol><li> Select or load an adjacency matrix from an ARACNe run or other source.
</li><li> Select or load a list of transcription factors.
</li><li> Define two classes of arrays, e.g. case and control in the Arrays/Phenotypes component.
</li><li> Set the significance threshold and t-test parameters as desired.
</li><li> Press the <b>Analyze</b> button.  The t-test followed by the Fisher's Exact tests will be carried out.
</li><li> A table and graphic showing transcription factors for whose
interactions significant overlap with the set of differentially
expressed genes was found will be displayed.
</li>
</ol>

 <p>&nbsp;</p>
</body></html>
