<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html dir="ltr" lang="en"><head>
<title>Normalization - Workbench</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="generator" content="MediaWiki 1.16.0">
<meta name="robots" content="noindex,follow">

</head>
<body class="mediawiki ltr ns-0 ns-subject page-Normalization skin-monobook">
<div id="globalWrapper">
<div id="column-content"><div id="content">
	<a id="top"></a>
	
	<h1 id="firstHeading" class="firstHeading">Normalization</h1>
	<div id="bodyContent">
		<!-- start content -->
<p><br>
</p>
<a name="Overview"></a><h1> <span class="mw-headline" id="Overview">Overview</span></h1>
<p>Normalization is used to transform data in preparation for analysis. 
 Many normalizers are oriented towards decreasing the effects of 
systematic differences across a set of microarrays, aiding in 
cross-microarray comparisons.  A log transformation of the data can be 
used to improve its statistical distribution.  geWorkbench offers a 
selection of pluggable normalization components (see list below).  The 
Normalizer panel is located in the <b>Commands and Analysis</b> area in the lower right side of the application.
</p><p>All normalization operations directly alter the loaded data set 
and are not reversible. A separate copy of the data is not generated. 
Normalization operations do not respect any marker or array sets that 
may be activated; normalization always acts on the entire data set. 
</p><p>The <b>Missing Value Computation</b> normalizer can be used to 
replace missing values with imputed values. Note that some analysis 
routines also will optionally compute replacements for missing values if
 needed.
</p><p>Data preparation methods such as RMA and GCRMA are not directly 
available in geWorkbench.  Affymetrix CEL files can be processed 
externally to geWorkbench using a program such as RMAExpress or in 
R/Bioconductor and then the processed data imported into geWorkbench.
</p><p>A short introduction to various methods of Affymetrix data preparation is available at <a href="http://discover.nci.nih.gov/microarrayAnalysis/Affymetrix.Preprocessing.jsp" class="external text" rel="nofollow">Affymetrix Preprocessing</a>.
</p><p>As of geWorkbench 2.3.0, all normalizers are activated in one of two ways:
1. By right-clicking on a microarray data node in the Project Folders component and following "Normalization".
2. By use of the top level Commands Menu, and again selecting and following Normalization.
</p><p>This is depicted for the Project Folders component 
</p><p><br>
<img alt="Project Folders Normalization Menu.png" src="normalization_files/Project_Folders_Normalization_Menu.png" height="405" width="438">
</p><p><br>
and for the Commands menu:
</p><p><br>
<img alt="Menu Bar Commands Normalization.png" src="normalization_files/Menu_Bar_Commands_Normalization.png" height="230" width="451">
</p>
<a name="Normalizer_Configuration"></a><h1> <span class="mw-headline" id="Normalizer_Configuration">Normalizer Configuration</span></h1>
<p>Some normalizers may not be loaded by default in geWorkbench.  To configure which normalizers to load, use the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Component_Configuration_Manager" title="Component Configuration Manager"> Component Configuration Manager</a>.  It is available in the top menu-bar under  <b>Tools-&gt;Component Configuration</b>.
</p><p>Only normalizers that have been loaded in the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Component_Configuration_Manager" title="Component Configuration Manager"> Component Configuration Manager (CCM)</a> will appear in the menus.
</p><p><br>
<img alt="Normalizers in CCM.png" src="normalization_files/Normalizers_in_CCM.png" height="351" width="794">
</p>
<a name="Available_Normalizers"></a><h1> <span class="mw-headline" id="Available_Normalizers">Available Normalizers</span></h1>
<p>geWorkbench comes with the following normalization routines installed: 
</p><p><br>
</p>
<a name="Basic_Controls"></a><h1> <span class="mw-headline" id="Basic_Controls">Basic Controls</span></h1>
<a name="Overview_2"></a><h2> <span class="mw-headline" id="Overview_2">Overview</span></h2>
<p>See any of the below normalizers for a depiction of these controls.
</p>
<ul><li> <b>Saved Parameters</b> (menu) - Allows selection of stored parameter settings.
</li><li> <b>Normalize</b> (button)- Run the selected transformation.
</li><li> <b>Save Settings</b> - Save the current settings (see following section).
</li><li> <b>Delete Settings</b>- Delete the currently selected parameter set.
</li></ul>
<a name="Saving_Parameters"></a><h2> <span class="mw-headline" id="Saving_Parameters">Saving Parameters</span></h2>
<p>The current parameter settings can be saved to a named parameter set.
  The saved set will be displayed in the pull-down menu at upper right 
in the component. Any number of parameter sets can be saved.  Selecting a
 saved parameter set will load its settings.  If the currently set 
parameters match a saved set, that set's entry will be shown in the 
menu.
</p>
<ul><li> <b>Save Settings</b> - Save the current settings to a new parameter set.
</li><li> <b>Delete Settings</b> - Delete the currently selected parameter set from the menu.
</li></ul>
<a name="Specific_Controls_for_each_Normalizer"></a><h1> <span class="mw-headline" id="Specific_Controls_for_each_Normalizer">Specific Controls for each Normalizer</span></h1>
<a name="Housekeeping_Genes_Normalizer"></a><h2> <span class="mw-headline" id="Housekeeping_Genes_Normalizer">Housekeeping Genes Normalizer</span></h2>
<p><img alt="Normalization-HouseKeeping.png" src="normalization_files/Normalization-HouseKeeping.png" height="323" width="888">
</p><p><br>
The Housekeeping Genes normalization method normalizes data using the 
expression values of selected housekeeping genes. In particular, for 
each microarray, the average expression of all markers corresponding to 
housekeeping genes is calculated and then all expression levels within 
the microarray are divided by that average. To specify the housekeeping 
gene markers, the user can load a CSV (comma separated values) file 
containing these markers. The loaded markers are displayed in the <b>Current Selected Genes</b> list box.  Markers can be moved back and forth between this list and the list at left, <b>Excluded Genes</b> through use of the right and left arrow buttons (<b>&lt; &gt;</b>), or simply by double-clicking on a marker in either list.  The current contents of the <b>Current Selected Genes</b> marker set can be saved as a CSV file for later re-use.  
</p><p><br>
</p>
<a name="Controls"></a><h3> <span class="mw-headline" id="Controls">Controls</span></h3>
<ul><li> <b>&lt;, &gt;</b> = Shift selected markers between lists.  Double-clicking on a list item will also shift it to the opposite list.
</li><li> <b>Clear All</b> - Clear both lists.
</li><li> <b>Load</b> - Load a list of markers from a comma-separated 
value (CSV) format file.  One marker per line is also an acceptable 
format (.csv suffix).
</li><li> <b>Save</b> - Save the list shown in the <b>Current Selected Genes</b> list to a file.  The output file format is one marker per line (suffix .csv).
</li></ul>
<ul><li> <b>Right-click menu</b> - If one right-clicks on a marker in the <b>Current Selected Genes</b> list, a pop-up menu appears, offering two choices:
</li></ul>
<ol><li> <b>Rename</b> - Rename the marker in the list.  Note that the marker name must match a marker in the currently loaded dataset.
</li><li> <b>Delete</b> - Delete a marker from the list.
</li></ol>
<a name="Missing_Value_Options_for_Housekeeping_Genes"></a><h3> <span class="mw-headline" id="Missing_Value_Options_for_Housekeeping_Genes">Missing Value Options for Housekeeping Genes</span></h3>
<ul><li> <b>Ignore</b> - Do not replace missing values.
</li><li> <b>Microarray Average</b> - Missing values are replaced for 
the housekeeping gene normalization with the average of all values on 
the microarray on which the missing value was found.
</li></ul>
<p><br>
</p>
<a name="Selected_Genes_Not_Available"></a><h3> <span class="mw-headline" id="Selected_Genes_Not_Available">Selected Genes Not Available</span></h3>
<p>When you have attempted to normalize the data with genes that are not
 included in the data, the system generates a warning message and 
highlights the missing genes in yellow in the <b>Current Selected Genes</b> list box.  A pop-up dialog box will offer the following choices:
</p>
<ul><li>  <b>OK</b> – Proceeds to normalize the data using only the 
markers that are present in the data set, disregarding the 
yellow-highlighted markers.
</li><li> <b>Cancel</b> – Cancels the normalization operation. The markers in the list which are not in the dataset will remain highlighted in yellow.
</li><li> <b>Right-click menu options</b> - Right-clicking on a highlighted marker will pop-up a menu with the following choices:
<ul><li> <b>Exclude Highlighted</b> - Move the highlighted markers back to the <b>Excluded Genes</b> list on left.
</li><li> <b>Delete Highlighted</b> - Delete the highlighted markers from the list.
</li><li> <b>Clear Highlights</b> - Clear all highlights, leaving the list unchanged.
</li></ul>
</li></ul>
<p><img alt="Normalization-Housekeeping invalid marker options.png" src="normalization_files/Normalization-Housekeeping_invalid_marker_options.png" height="204" width="367">
</p>
<a name="Log2_Transformation"></a><h2> <span class="mw-headline" id="Log2_Transformation">Log2 Transformation</span></h2>
<p><img alt="Normalization-Log2 Transformation.png" src="normalization_files/Normalization-Log2_Transformation.png" height="222" width="326">
</p><p>Transforms the expression values in all microarrays in the 
current data set of taking the base 2 log of the data.  The 
transformation is applied to all expression measurements in all 
microarrays in the currently loaded data set. The operation is applied 
only if all values in all microarrays are positive.  If this is not the 
case, an error message appears indicating that the normalization could 
not be performed.
</p>
<a name="Marker_Centering_Normalizer"></a><h2> <span class="mw-headline" id="Marker_Centering_Normalizer">Marker Centering Normalizer</span></h2>
<p><img alt="Normalization-Marker Centering.png" src="normalization_files/Normalization-Marker_Centering.png" height="222" width="491">
</p><p>This component allows normalizing using a choice of the mean or 
median value for each marker profile across all arrays in the currently 
loaded data set.  The calculated value for a given marker is subtracted 
from each measurement for that marker across all arrays.
</p>
<a name="Centering_Parameters"></a><h3> <span class="mw-headline" id="Centering_Parameters">Centering Parameters</span></h3>
<ul><li> <b>Averaging method</b> - Designates if the mean or the median 
should be used for centering. Mean and median values are computed from 
the non-missing values within the profile.
<ul><li> <b>Mean</b> - Use the mean value of each marker profile.
</li><li> <b>Median</b> - Use the median value of each marker profile.
</li></ul>
</li><li> <b>Missing values</b> - Designates how to handle missing values in a marker profile. Available options are:
<ul><li> <b>Min profile</b>: Missing values are replaced with the minimum (post-centering) profile value.
</li><li> <b>Max profile</b>: Missing values are replaced with the maximum (post-centering) profile value.
</li><li> <b>Zero</b>: Missing values are set to 0.
</li><li> <b>Ignore</b>: Missing values are left untouched.
</li></ul>
</li></ul>
<a name="Mean-Variance_Normalizer"></a><h2> <span class="mw-headline" id="Mean-Variance_Normalizer">Mean-Variance Normalizer</span></h2>
<p><img alt="Normalization-Mean Variance.png" src="normalization_files/Normalization-Mean_Variance.png" height="222" width="489">
</p><p><br>
For each marker profile, the mean measurement of the entire profile is 
subtracted from each measurement in the profile and the resulting value 
is divided by the profile's standard deviation.
</p>
<a name="Mean-Variance_Parameters"></a><h3> <span class="mw-headline" id="Mean-Variance_Parameters">Mean-Variance Parameters</span></h3>
<ul><li> <b>Missing values</b> - Designates how to handle missing values in a marker profile. Available options are:
<ul><li> <b>Min profile</b>: Missing values are replaced with the minimum (post-centering) profile value.
</li><li> <b>Max profile</b>: Missing values are replaced with the maximum (post-centering) profile value.
</li><li> <b>Zero</b>: Missing values are set to 0.
</li><li> <b>Ignore</b>: Missing values are left untouched.
</li></ul>
</li></ul>
<a name="Microarray_Centering_Normalizer"></a><h2> <span class="mw-headline" id="Microarray_Centering_Normalizer">Microarray Centering Normalizer</span></h2>
<p><img alt="Normalization-Microarray Centering.png" src="normalization_files/Normalization-Microarray_Centering.png" height="207" width="506">
</p>
<a name="Centering_Parameters_2"></a><h3> <span class="mw-headline" id="Centering_Parameters_2">Centering Parameters</span></h3>
<ul><li> <b>Averaging method</b> - Designates if the mean or the median 
should be used for centering. Mean and median values are computed from 
the non-missing values within the profile.
<ul><li> <b>Mean</b> - Use the mean value of each marker profile.
</li><li> <b>Median</b> - Use the median value of each marker profile.
</li></ul>
</li><li> <b>Missing values</b> - Designates how to handle missing values in a marker profile. Available options are:
<ul><li> <b>Min profile</b>: Missing values are replaced with the minimum (post-centering) profile value.
</li><li> <b>Max profile</b>: Missing values are replaced with the maximum (post-centering) profile value.
</li><li> <b>Zero</b>: Missing values are set to 0.
</li><li> <b>Ignore</b>: Missing values are left untouched.
</li></ul>
</li></ul>
<a name="Missing_Values_Normalizer"></a><h2> <span class="mw-headline" id="Missing_Values_Normalizer">Missing Values Normalizer</span></h2>
<p><img alt="Normalization-Missing Values.png" src="normalization_files/Normalization-Missing_Values.png" height="207" width="542">
</p><p>This normalizer replaces a missing measurement with a consensus 
value. This value is computed based upon the selection at the drop-down 
box titled "<b>Averaging method</b>". 
</p>
<a name="Averaging_method"></a><h3> <span class="mw-headline" id="Averaging_method">Averaging method</span></h3>
<ul><li> <b>Mean profile marker</b> - Replaces the missing measurement 
with the average value (across all microarrays) of its corresponding 
marker. The average is computed by taking into account only non-missing 
values.
</li><li> <b>Mean microarray value</b> - Replaces the missing 
measurement with the average value of its corresponding microarray. The 
average is computed by taking into account only non-missing values.
</li></ul>
<a name="Quantile_Normalizer"></a><h2> <span class="mw-headline" id="Quantile_Normalizer">Quantile Normalizer</span></h2>
<p><img alt="Normalization-Quantile.png" src="normalization_files/Normalization-Quantile.png" height="207" width="627">
</p><p><br>
Quantile normalization adjusts the expression measurements in each 
microarray so that the expression value distribution is the same for all
 microarrays in the microarray set being normalized. This is a 
non-linear normalization method which is not dependent upon the choice 
of a reference or baseline array. A description of the underlying 
computations can be found in: Bolstad, B<i> et al.</i> (2003). In geWorkbench, the normalization is applied to the probeset level values (not the individual probe values).
</p>
<a name="Missing_Values_Averaging_method"></a><h3> <span class="mw-headline" id="Missing_Values_Averaging_method">Missing Values Averaging method</span></h3>
<ul><li> <b>Mean profile marker</b> - Replaces the missing measurement 
with the average value (across all microarrays) of its corresponding 
marker. The average is computed by taking into account only non-missing 
values.
</li><li> <b>Mean microarray value</b> - Replaces the missing 
measurement with the average value of its corresponding microarray. The 
average is computed by taking into account only non-missing values.
</li></ul>
<p><br>
</p>
<a name="References:"></a><h3> <span class="mw-headline" id="References:">References:</span></h3>
<p>Bolstad BM, Irizarry RA, Astrand M, Speed TP (2003) A comparison of 
normalization methods for high density oligonucleotide array data based 
on variance and bias. Bioinformatics 19(2):185-93. (<a href="http://www.ncbi.nlm.nih.gov/pubmed/12538238" class="external free" rel="nofollow">http://www.ncbi.nlm.nih.gov/pubmed/12538238</a>)
</p><p><br>
Supplementary information: <a href="http://bmbolstad.com/misc/normalize/normalize.html" class="external free" rel="nofollow">http://bmbolstad.com/misc/normalize/normalize.html</a>.
</p>
<a name="Threshold_Normalizer"></a><h2> <span class="mw-headline" id="Threshold_Normalizer">Threshold Normalizer</span></h2>
<p><img alt="Normalization-Threshold.png" src="normalization_files/Normalization-Threshold.png" height="228" width="474">
</p><p>Allows a minimum or maximum value for the data to be set.  If a 
minimum value is specified, any value less than that will be set to the 
given minimum value.  If a maximum value is specified, any value greater
 than that will be set to the given maxium value.
</p><p><br>
</p>
<a name="Threshold_Parameters"></a><h3> <span class="mw-headline" id="Threshold_Parameters">Threshold Parameters</span></h3>
<ul><li> <b>Cut-off value</b> - Enter a real-number value for the desired cut-off.
</li><li> <b>Cut-off type</b>
<ul><li> <b>Minimum</b> - Any expression value less than the specified cut-off value is set equal to the cut-off value (floor).
</li><li> <b>Maximum</b> - Any expression value greater than the specified cut-off value is set equal to the cut-off value (ceiling).
</li></ul>
</li><li> <b>Missing values</b>
<ul><li> <b>Ignore</b> - Do not change missing values.
</li><li> <b>Replace</b> - Replace missing values with the specified cut-off value.
</li></ul>
</li></ul>

<!-- 
NewPP limit report
Preprocessor node count: 102/1000000
Post-expand include size: 1907/2097152 bytes
Template argument size: 0/2097152 bytes
Expensive parser function count: 0/100
-->

<!-- Saved in parser cache with key workbench:pcache:idhash:2109-0!1!0!!en!2!printable=1 and timestamp 20120309212950 -->
<div class="printfooter">
Retrieved from "<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Normalization">http://wiki.c2b2.columbia.edu/workbench/index.php/Normalization</a>"</div>
		<div id="catlinks" class="catlinks catlinks-allhidden"></div>		<!-- end content -->
				<div class="visualClear"></div>
	</div>
</div></div>
<div class="visualClear"></div>
<div id="footer">
	<div id="f-poweredbyico"><img src="normalization_files/poweredby_mediawiki_88x31.png" alt="Powered by MediaWiki" height="31" width="88"></div>
	<ul id="f-list">
		<li id="lastmod"> This page was last modified on 1 March 2012, at 12:20.</li>
	</ul>
</div>
</div>

