<html>



<head>

<meta http-equiv="Content-Language" content="en-us">

<meta http-equiv="Content-Type" content="text/html; charset=windows-1252">

<meta name="GENERATOR" content="Microsoft FrontPage 4.0">

<meta name="ProgId" content="FrontPage.Editor.Document">

<title>User Comments</title>

<link rel="stylesheet" type="text/css" href="ncicb.css">

</head>



<body>
    <h3>Synteny</h3> 
    <p>Synteny is a utility for comparison of extended genomic regions located in the <b>Analysis Area </b>in the lower-right of the application. To perform the analyses, complete the following steps:</p> 
    <p>- Gather the input data from<a href="#_Add_genomic_region_from the Marker "> <b>Microarray markers from an activated panel </b></a><b>or <a href="#_Add_genomic_region_to selected regi">specified genomic region.</a> </b></p> 
    <p>- Then <a href="#_Perform_the_following_steps_to_crea_1">run the regions comparison</a> and the result is displayed in the <a href="#_Basic_Usage_of_the_Dot_Matrix_Panel">Dot Matrix</a> visualization.</p> 
    <b><p>Basic Usage </p> </b>
    <p>The steps of basic usage are indicated in the below screenshot.</p> 
    <p><img border="0" width="553" height="279" src="syteny_files/image001.jpg"></p> 
    <p>A. The<b> Markers </b>subtab supports selection of marker for further analysis. </p> 
    <p>B. The <b>Program</b> subtab lists analysis programs.</p> 
    <p>C. The <b>Annotation </b>subtab changes the view of annotation tracks in the <b>Dot Matrix</b> view. Select or deselect the annotation track and than click<b> Redraw </b>. Changes are made to the<b> DotMatrix</b> view.</p> 
    <p>D. The <b>Genome s</b>ubtab supports Selection of genomic regions to<b> Add to Selected Regions</b> panel.</p> 
    <p>E. Displays markers from the active Gene Panel(s).</p> 
    <p>F. Genomic locations associated with selected markers are added to the tree in the right panel.</p> 
    <p>G. Defines the genomic region for the marker in the Selected Region From tree. The region can be modified prior to perform <b>Add to Selected Regions</b>.</p> 
    <h3><a name="_Perform_the_following_steps_to_add_"></a><a name="_Add_genomic_region_from the Marker "></a>Add genomic region from the Marker subtab:</h3> 
    <p>1 Navigate to the Synteny tab located in the <b>Analysis Area</b> in the lower-right of the application. </p> 
    <p>2 In the Markers Tab, select a marker and right click. The menu with single item <b>Add marker</b> will appear.</p> 
    <p> - If the server database has genomic locations associated with the marker, it's added to the tree in the right panel.</p> 
    <p> - The message <b>&#8220;</b><i>No sequence for this marker</i><b>&#8221;</b> appears in the status line when genomic locations can't be found. In this case the genomic location can be entered manually in <b>Genome</b> subtab for the marker. </p> 
    <p>3 Open all internal nodes in the tree and select a marker genomic location. Note that in some cases marker can have more than one genomic location. The genomic location string includes marker name, genome ID, chromosome number, chromosome start and end positions, and associated gene ID. </p> 
    <p>4 Right click on selected genomic region and menu will appears. Click<b> Add to Selected Regions </b>to add specified genomic location to the list of selected genomics regions.<b> </b></p> 
    <p>The <b>Genome</b> subtab supports selection of genomic regions<b> </b>to <b>Add to selected</b> to ass to the selected regions panel. In this panel you can select genomic positions by selecting the genome, chromosomes and chromosome start and end positions. </p> 
    <p><img border="0" width="323" height="197" src="syteny_files/image002.gif"></p> 
    <h3><a name="_Perform_the_following_steps_to_crea"></a><a name="_Add_genomic_region_to selected regi"></a>Add genomic region to selected regions panel from the Genome subtab:</h3> 
    <p>1 Navigate to the <b>Genome</b> subtab in the Synteny tab.</p> 
    <p>2 Select the genome, chromosome and start and end positions.</p> 
    <p>3 Click <b>Add to selected</b> button.</p> 
    <h3><img border="0" width="501" height="283" src="syteny_files/image003.gif"></h3> 
    <h3><a name="_Perform_the_following_steps_to_crea_1"></a>Create genomic regions comparison: </h3> 
    <p>1 Navigate to the selected region list in the lower right. </p> 
    <p>2 Populate values for the sequence along the X Axis and Sequence along the Y Axis by right clicking on the genomic regions in selected regions list and selecting <b>Add to X</b> or <b>Add to Y</b> from popup menu.</p> 
    <p>3 Select program for performing analysis in the <b>Program tab</b>. </p> 
    <p>4 Click<b> Run </b>to start the comparison. The results will be displayed in the <b>Dot Matrix Pane</b>l in visual area. <a name="_Basic_Usage_of_the_Dot_Matrix_Panel"></a><a name="_Dot_Matrix_Panel"></a></p>
</ul>



</body>



</html>

