<html>

<head>
<meta http-equiv=Content-Type content="text/html; charset=us-ascii">
<meta name=Generator content="Microsoft Word 11 (filtered)">
<title>Title </title>
<style>
<!--
 /* Font Definitions */
 @font-face
	{font-family:Wingdings;
	panose-1:5 0 0 0 0 0 0 0 0 0;}
@font-face
	{font-family:PMingLiU;
	panose-1:2 2 3 0 0 0 0 0 0 0;}
@font-face
	{font-family:"Arial Unicode MS";
	panose-1:2 11 6 4 2 2 2 2 2 4;}
@font-face
	{font-family:Verdana;
	panose-1:2 11 6 4 3 5 4 4 2 4;}
@font-face
	{font-family:Tahoma;
	panose-1:2 11 6 4 3 5 4 4 2 4;}
@font-face
	{font-family:"\@Arial Unicode MS";
	panose-1:2 11 6 4 2 2 2 2 2 4;}
@font-face
	{font-family:"\@PMingLiU";
	panose-1:2 2 3 0 0 0 0 0 0 0;}
 /* Style Definitions */
 p.MsoNormal, li.MsoNormal, div.MsoNormal
	{margin:0cm;
	margin-bottom:.0001pt;
	font-size:12.0pt;
	font-family:"Times New Roman";}
h1
	{margin-top:9.0pt;
	margin-right:0cm;
	margin-bottom:9.0pt;
	margin-left:21.6pt;
	text-indent:-21.6pt;
	line-height:300%;
	page-break-after:avoid;
	font-size:26.0pt;
	font-family:Arial;
	font-weight:bold;}
h2
	{margin:0cm;
	margin-bottom:.0001pt;
	line-height:125%;
	page-break-after:avoid;
	font-size:13.5pt;
	font-family:Arial;
	color:#312C2C;
	font-weight:bold;}
h3
	{margin-right:0cm;
	margin-left:0cm;
	line-height:10.5pt;
	font-size:10.0pt;
	font-family:Verdana;
	color:#312C2C;
	font-weight:bold;}
p.MsoCommentText, li.MsoCommentText, div.MsoCommentText
	{margin:0cm;
	margin-bottom:.0001pt;
	font-size:10.0pt;
	font-family:"Times New Roman";}
p.MsoBodyTextIndent, li.MsoBodyTextIndent, div.MsoBodyTextIndent
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:24.0pt;
	font-size:12.0pt;
	font-family:"Times New Roman";}
a:link, span.MsoHyperlink
	{color:blue;
	text-decoration:underline;}
a:visited, span.MsoHyperlinkFollowed
	{color:purple;
	text-decoration:underline;}
p
	{margin-right:0cm;
	margin-left:0cm;
	font-size:12.0pt;
	font-family:"Times New Roman";}
p.MsoCommentSubject, li.MsoCommentSubject, div.MsoCommentSubject
	{margin:0cm;
	margin-bottom:.0001pt;
	font-size:10.0pt;
	font-family:"Times New Roman";
	font-weight:bold;}
p.MsoAcetate, li.MsoAcetate, div.MsoAcetate
	{margin:0cm;
	margin-bottom:.0001pt;
	font-size:8.0pt;
	font-family:Tahoma;}
span.Heading2Char
	{font-family:Arial;
	color:#312C2C;
	font-weight:bold;}
p.TableText, li.TableText, div.TableText
	{margin-top:1.0pt;
	margin-right:0cm;
	margin-bottom:1.0pt;
	margin-left:36.0pt;
	text-align:justify;
	text-justify:inter-ideograph;
	word-break:break-all;
	font-size:10.0pt;
	font-family:Arial;}
p.bulletABC, li.bulletABC, div.bulletABC
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:10.0pt;
	margin-left:96.0pt;
	text-align:justify;
	text-justify:inter-ideograph;
	text-indent:-24.0pt;
	font-size:12.0pt;
	font-family:Arial;}
p.Imageannotation, li.Imageannotation, div.Imageannotation
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:36.0pt;
	margin-bottom:.0001pt;
	font-size:12.0pt;
	font-family:Arial;
	color:#003300;
	font-weight:bold;
	font-style:italic;}
p.StyleHeading1h1, li.StyleHeading1h1, div.StyleHeading1h1
	{margin-top:30.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:21.6pt;
	margin-bottom:.0001pt;
	text-indent:-21.6pt;
	page-break-after:avoid;
	border:none;
	padding:0cm;
	font-size:16.0pt;
	font-family:Arial;
	font-weight:bold;}
p.proclabel, li.proclabel, div.proclabel
	{margin-right:0cm;
	margin-left:0cm;
	font-size:12.0pt;
	font-family:"Times New Roman";}
p.note, li.note, div.note
	{margin-right:0cm;
	margin-left:0cm;
	font-size:12.0pt;
	font-family:"Times New Roman";}
span.StyleHeading312ptChar
	{font-family:Verdana;
	color:#312C2C;
	font-weight:bold;}
p.StyleHeading312pt, li.StyleHeading312pt, div.StyleHeading312pt
	{margin-right:0cm;
	margin-left:0cm;
	font-size:12.0pt;
	font-family:Verdana;
	color:#312C2C;
	font-weight:bold;}
p.HEADING41, li.HEADING41, div.HEADING41
	{margin-right:0cm;
	margin-left:0cm;
	line-height:10.5pt;
	font-size:8.5pt;
	font-family:Verdana;
	color:#312C2C;
	font-weight:bold;}
p.BUTTON, li.BUTTON, div.BUTTON
	{margin-right:12.0pt;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	line-height:10.5pt;
	font-size:9.0pt;
	font-family:Verdana;
	color:#312C2C;
	font-weight:bold;}
p.helpnormaltex, li.helpnormaltex, div.helpnormaltex
	{margin-right:0cm;
	margin-left:0cm;
	font-size:9.0pt;
	font-family:Verdana;}
ins
	{text-decoration:none;}
span.msoIns
	{text-decoration:underline;}
span.msoDel
	{text-decoration:line-through;
	color:red;}
 /* Page Definitions */
 @page Section1
	{size:595.3pt 841.9pt;
	margin:72.0pt 90.0pt 72.0pt 90.0pt;
	layout-grid:18.0pt;}
div.Section1
	{page:Section1;}
 /* List Definitions */
 ol
	{margin-bottom:0cm;}
ul
	{margin-bottom:0cm;}
-->
</style>

</head>

<body lang=ZH-TW link=blue vlink=purple style='text-justify-trim:punctuation'>

<div class=Section1 style='layout-grid:18.0pt'>

<p class=StyleHeading312pt><span lang=EN-US>Synteny</span></p>

<p class=helpnormaltex><span lang=EN-US>Synteny is a utility for comparison of
extended genomic regions located in the <b>Analysis Area </b>in the lower-right
of the application.&nbsp; To perform the analyses, complete the following steps:</span></p>

<p class=helpnormaltex style='margin-left:84.0pt;text-indent:-24.0pt'><span
lang=EN-US style='font-family:Wingdings'>n<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>Gather the input data from<a
href="#_Add_genomic_region_from the Marker "> <b>Microarray markers from an
activated panel </b></a><b>or <a href="#_Add_genomic_region_to selected regi">specified
genomic region.</a> </b></span></p>

<p class=BUTTON style='margin-left:84.0pt;text-indent:-24.0pt'><span
lang=EN-US style='font-family:Wingdings;font-weight:normal'>n<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US style='color:windowtext;font-weight:normal'>Then</span><span
lang=EN-US> <a href="#_Perform_the_following_steps_to_crea_1">run the regions
comparison</a> </span><span lang=EN-US style='font-weight:normal'>and the result
is displayed in the </span><span lang=EN-US style='font-weight:normal'><a
href="#_Basic_Usage_of_the_Dot_Matrix_Panel"><span style='color:windowtext;
text-decoration:none'>Dot Matrix</span></a> </span><span lang=EN-US
style='font-weight:normal'>visualization.</span></p>

<p class=StyleHeading312pt><span lang=EN-US>Basic Usage </span></p>

<p class=MsoNormal style='margin-left:28.5pt;text-autospace:none'><span
lang=EN-US style='font-size:9.0pt;font-family:Verdana'>The steps of basic usage
are indicated in the below screenshot.</span></p>

<p class=helpnormaltex><span lang=EN-US><img border=0 width=553 height=279
src="syteny_files/image001.jpg"></span></p>

<p class=helpnormaltex style='margin-left:18.0pt;text-indent:-18.0pt'><span
lang=EN-US>A.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>The<b> Markers </b>subtab supports selection of
marker for further analysis. </span></p>

<p class=helpnormaltex style='margin-left:18.0pt;text-indent:-18.0pt'><span
lang=EN-US>B.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>The <b>Program</b> subtab lists analysis
programs.</span></p>

<p class=helpnormaltex style='margin-left:18.0pt;text-indent:-18.0pt'><span
lang=EN-US>C.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>The <b>Annotation </b>subtab changes the view of
annotation tracks in the <b>Dot Matrix</b> view. Select or deselect the
annotation track and than click<b> Redraw </b>button. Changes are made to the<b>
DotMatrix</b> view.</span></p>

<p class=helpnormaltex style='margin-left:18.0pt;text-indent:-18.0pt'><span
lang=EN-US>D.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>The <b>Genome s</b>ubtab supports Selection of
genomic regions to<b> Add to Selected Regions</b> panel.</span></p>

<p class=helpnormaltex style='margin-left:18.0pt;text-indent:-18.0pt'><span
lang=EN-US>E.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>Displays markers from the active Gene Panel(s).</span></p>

<p class=helpnormaltex style='margin-left:18.0pt;text-indent:-18.0pt'><span
lang=EN-US>F.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>Genomic locations associated with selected
markers are added to the tree in the right panel.</span></p>

<p class=helpnormaltex style='margin-left:18.0pt;text-indent:-18.0pt'><span
lang=EN-US>G.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>Defines the genomic region for the marker in the
Selected Region From tree.&nbsp; The region can be modified prior to perform <b>Add
to Selected Regions</b>.</span></p>

<h3><a name="_Perform_the_following_steps_to_add_"></a><a
name="_Add_genomic_region_from the Marker "></a><span lang=EN-US>Add genomic
region from the Marker subtab:</span></h3>

<p class=helpnormaltex style='margin-left:21.25pt;text-indent:-21.25pt'><span
lang=EN-US>1<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>Navigate to the Synteny tab located in the <b>Analysis
Area</b> in the lower-right of the application. </span></p>

<p class=helpnormaltex style='margin-left:21.25pt;text-indent:-21.25pt'><span
lang=EN-US>2<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>In the Markers Tab, select a marker and right
click. The menu with single item <b>Add marker</b> will appear.</span></p>

<p class=helpnormaltex style='margin-left:21.25pt;text-indent:-21.25pt'><span
lang=EN-US>3<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>&nbsp;If in server database has genomic
locations associated with this marker they will be added to the tree in the
right panel.</span></p>

<p class=helpnormaltex style='margin-left:21.25pt;text-indent:-21.25pt'><span
lang=EN-US>4<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>&nbsp;If there now genomic location for this
marker in database the message <b>&#8220;</b><i>No sequence for this marker</i><b>&#8221;</b>
will appears in the status line, in this case the genomic location for this
marker can be entered manually in <b>Genome</b> subtab. </span></p>

<p class=helpnormaltex style='margin-left:21.25pt;text-indent:-21.25pt'><span
lang=EN-US>5<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>In the tree open all internal nodes and select
one marker genomic location, note that in some cases marker can have more than
one genomic location. The genomic location string includes marker name, genome
ID, chromosome number, chromosome start and end positions, and associated gene
ID. </span></p>

<p class=helpnormaltex style='margin-left:21.25pt;text-indent:-21.25pt'><span
lang=EN-US>6<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>Right click on selected genomic region and menu
will appears. Click<b> Add to Selected Regions </b>to add specified genomic
location to the list of selected genomics regions.</span><b><span lang=EN-US> </span></b></p>

<p class=helpnormaltex><span lang=EN-US>The <b>Genome</b> subtab supports
selection of genomic regions<b> </b>to <b><span style='color:#312C2C'>Add to
selected</span></b>&nbsp; to ass to the selected regions panel. In this panel
you can select genomic positions by selecting the genome, chromosomes and
chromosome start and end positions. </span></p>

<p class=helpnormaltex><span lang=EN-US><img border=0 width=323 height=197
src="syteny_files/image002.gif"></span></p>

<h3><a name="_Perform_the_following_steps_to_crea"></a><a
name="_Add_genomic_region_to selected regi"></a><span lang=EN-US>Add genomic
region to selected regions panel from the Genome subtab:</span></h3>

<p class=helpnormaltex style='margin-left:21.25pt;text-indent:-21.25pt'><span
lang=EN-US>1<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>Navigate to the <b>Genome</b> subtab in the
Synteny tab.</span></p>

<p class=helpnormaltex style='margin-left:21.25pt;text-indent:-21.25pt'><span
lang=EN-US>2<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>Select the genome, chromosome and start and end
positions.</span></p>

<p class=helpnormaltex style='margin-left:21.25pt;text-indent:-21.25pt'><span
lang=EN-US>3<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>Click <b>Add to selected</b> button.</span></p>

<h3><span lang=EN-US style='font-weight:normal'><img border=0 width=501
height=283 src="syteny_files/image003.gif"></span></h3>

<h3><a name="_Perform_the_following_steps_to_crea_1"></a><span lang=EN-US>Create
genomic regions comparison: </span></h3>

<p class=helpnormaltex style='margin-left:21.25pt;text-indent:-21.25pt'><span
lang=EN-US>1<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>Navigate to the selected region list in the
lower right. </span></p>

<p class=helpnormaltex style='margin-left:21.25pt;text-indent:-21.25pt'><span
lang=EN-US>2<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>Populate values for the sequence along the X
Axis and Sequence along the Y Axis by right clicking on the genomic regions in
selected regions list and selecting <b>Add to X</b> or <b>Add to Y</b> from
popup menu.</span></p>

<p class=helpnormaltex style='margin-left:21.25pt;text-indent:-21.25pt'><span
lang=EN-US>3<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>Select program for performing analysis in the <b>Program
tab</b>. </span></p>

<p class=helpnormaltex style='margin-left:21.25pt;text-indent:-21.25pt'><span
lang=EN-US>4<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><span lang=EN-US>Click<b> Run </b>to start the comparison. The
results will be displayed in the <b>Dot Matrix Pane</b>l in visual area. <a
name="_Basic_Usage_of_the_Dot_Matrix_Panel"></a><a name="_Dot_Matrix_Panel"></a></span></p>

</div>

</body>

</html>
