<!DOCTYPE HTML PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html dir="ltr" xml:lang="en" xmlns="http://www.w3.org/1999/xhtml" lang="en"><head>

  
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <title>Tutorial - Cellular Networks KnowledgeBase - Workbench</title>
    <style type="text/css" media="screen,projection">/*<![CDATA[*/ @import "/workbench/skins/monobook/main.css"; /*]]>*/</style>
    <link rel="stylesheet" type="text/css" href="cnkb_files/commonPrint.css">
<p><br>
</p>
<a name="Overview"></a><h1>Overview</h1>
<a name="General"></a><h2>General</h2>
<p>The Cellular Network Knowledge Base (CNKB) is a repository of
interactions between protein-protein and protein-DNA interactions
(these interactions can be either computationally or experimentally
derived). Both direct, physical interactions can be captured as well as
indirect transcriptional relationships (where an interaction is between
a transcription factor and its gene target). Each pairwise interaction
has an associated confidence indicator (a value between 0 and 1)
reflecting the strength of the underlying data, whether experimental or
computational. Details on the methodology used to construct the CNKB
are available in <a href="http://www.ncbi.nlm.nih.gov/pubmed/18277385?ordinalpos=1&amp;itool=EntrezSystem2.PEntrez.Pubmed.Pubmed_ResultsPanel.Pubmed_RVBrief" class="external text" title="http://www.ncbi.nlm.nih.gov/pubmed/18277385?ordinalpos=1&amp;itool=EntrezSystem2.PEntrez.Pubmed.Pubmed ResultsPanel.Pubmed RVBrief" rel="nofollow">Mani et al. 2008</a>. 
</p><p>Gene interaction information from the CNKB can be used, for
example, in order to assess the plausibility of a hypothesis of
concerted molecular action represented by a gene set that has been
discovered using computational approaches (e.g., by running a
clustering analysis on a microarray set to identify tandems of
co-expressed genes). If the genes in such a set are reported in the
CNKB to have several direct interactions (or several common targets)
then this may be evidence that the gene set indeed reflects at some
level a real biological process.
</p><p>The CNKB component allows the user to select a group of markers
and specify for each the interaction type of interest (Protein-Protein
and/or Protein-DNA). All interactions (of the designated types)
involving the selected markers are retrieved from the CNKB and
displayed (along with associated information such as GO annotation,
interaction attributes, etc) both in the CNKB component and also in
Cytoscape.
</p>
<a name="The_CNKB_graphical_interface"></a><h2>The CNKB graphical interface</h2>
<p><a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_ProtDNA_confidence75.png" class="image" title="Image:T_CNKB_ProtDNA_confidence75.png"><img src="cnkb_files/T_CNKB_ProtDNA_confidence75.png" alt="Image:T_CNKB_ProtDNA_confidence75.png" longdesc="/workbench/index.php/Image:T_CNKB_ProtDNA_confidence75.png" width="898" height="723"></a>
</p><p>The CNKB component displays a list of markers that have been
activated in the Markers component. The user can select from these
markers by double-clicking them; those selected will be added to the
Selected Marker List just below.
</p><p>The graphical user interface (GUI) of the CNKB component has
three areas of distinct functionality: The "Activated Marker List", the
"Selected Marker List", and the "Throttle Graph". To use the component
a microarray set node must first be selected in the Project Folders
area of geWorkbench and then one or more marker sets from the Markers
component must be activated. The markers in those activated marker sets
will appear in the "Activated Marker List" area of the the CNKB GUI.
From there, one or more markers can be selected and moved into the
"Selected Marker List" (markers can be selected one at a time by
double-clicking; or they can be selected as a group and moved by
right-clicking on the selection). Communication with the CNKB database
is initiated by clicking on the "Refresh" button and all pairwise
interactions that involve genes represented by a marker in the "Select
Marker List" are retrieved (an interaction is retrieved if at least one
of its 2 members is in the "Selected Marker List"). There are 2 types
of interactions stored in the CNKB: Protein-Protein and Protein-Dna
(the latter indicating transcriptional relationships between
transcription factors and their target). The aggregate number (across
all genes) of retrieved interactions is displayed on the Throttle
Graph. Three distinct graphs are drawn, showing the aggregate for each
of the two interaction categories as well as their combined total. </p><p>The interactions can be assembled into a network and
visualized in the Cytoscape component. This is achieved by clicking on
the "Create Network" button. The x-axis slider in the Throttle graph
can be used to threshold which interactions will be included in the
Cytoscape network; by moving the slider to any given threshold value
from 0-1, only interactions whose confidence level is above the
threshold will be retained. Further, it is possible to include/exclude
interactions that belong to one or the other type. To that end, the
checkboxes in the "Selected Marker List" can be used. </p>
<a name="Working_with_the_CNKB_graphical_interface"></a><h1>Working with the CNKB graphical interface</h1> 
<p><br>
</p>
<a name="Activated_Marker_List"></a><h2>Activated Marker List</h2>
<p>The "Activated Marker List" contains the markers that belong to
activated marker sets from the Markers component. It contains 3
columns:
</p>
<a name="Marker"></a><h3>Marker</h3>
<p>The marker name (comes from the microarray set node selected in the Project Folders area).
</p>
<a name="Gene"></a><h3>Gene</h3> 
<p>The gene name corresponding to the marker, if loaded from an annotation file.
</p>
<a name="Type"></a><h3>Type</h3> 
<p>A gene type designation, derived from the gene's GO annotation: 
</p>
<ul><li> <b>TF</b> - Transcription Factor,
</li><li> <b>K</b> - Kinase,
</li><li> <b>P</b> - Phosphatase, and
</li><li> (no entry) - type is unknown.
</li></ul>
<p>Individual markers in the Activated Marker List can be moved to the Selected Markers List by double-clicking on them.
</p><p>Alternatively, a right-click menu allow a group of highlighted
markers, or all markers, to be moved to the "Selected Markers List".
</p><p><br>
<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_Activated_List_rightclick_menu_small.png" class="image" title="Image:T_CNKB_Activated_List_rightclick_menu_small.png"><img src="cnkb_files/T_CNKB_Activated_List_rightclick_menu_small.png" alt="Image:T_CNKB_Activated_List_rightclick_menu_small.png" longdesc="/workbench/index.php/Image:T_CNKB_Activated_List_rightclick_menu_small.png" width="465" height="368"></a>
</p>
<a name="Selected_Marker_List"></a><h2>Selected Marker List</h2>
<p>The "Selected Markers List" table contains the subset of markers
from the "Activated Marker List" about which information will be
retrieved from the CNKB.
</p><p>Items that have been added to the Selected Markers list are
initially shown in red using an italic font, to indicate the CKNB has
not yet been queried to retrieve their interaction information. </p><p><a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_SelectedMarkers_before_query.png" class="image" title="Image:T_CNKB_SelectedMarkers_before_query.png"><img src="cnkb_files/T_CNKB_SelectedMarkers_before_query.png" alt="Image:T_CNKB_SelectedMarkers_before_query.png" longdesc="/workbench/index.php/Image:T_CNKB_SelectedMarkers_before_query.png" width="857" height="278"></a>
</p><p><br>
After the query has been run against the CNKB database, the marker entries are shown in regular font, blue letters.
</p><p><br>
<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_SelectedMarkers_after_query.png" class="image" title="Image:T_CNKB_SelectedMarkers_after_query.png"><img src="cnkb_files/T_CNKB_SelectedMarkers_after_query.png" alt="Image:T_CNKB_SelectedMarkers_after_query.png" longdesc="/workbench/index.php/Image:T_CNKB_SelectedMarkers_after_query.png" width="893" height="257"></a>
</p><p><br>
Items that have been added to the Selected Markers list can be removed
and sent back to the Activated Markers list by double clicking on their
entries or through a right-click menu. The right-click menu gives the
choice of moving only highlighted, or all markers back to the Activated
Markers list.
</p><p><br>
<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_SelectedMarkers_rightclick_menu.png" class="image" title="Image:T_CNKB_SelectedMarkers_rightclick_menu.png"><img src="cnkb_files/T_CNKB_SelectedMarkers_rightclick_menu.png" alt="Image:T_CNKB_SelectedMarkers_rightclick_menu.png" longdesc="/workbench/index.php/Image:T_CNKB_SelectedMarkers_rightclick_menu.png" width="457" height="185"></a>
</p><p><br>
</p><p>The Selected Marker List table contains the following columns:
</p>
<a name="Include_Protein-DNA.2C_Include_Protein-Protein"></a><h3>Include Protein-DNA, Include Protein-Protein</h3>
<p>Check boxes used to specify which type(s) of interactions to use
when assembling the Cytoscape network. These check boxes do not affect
retrieval of interactions from the CNKB database, only network
creation.
</p>
<ul><li> <b>All Protein-DNA</b> - checking this box will select all Include Protein-DNA boxes.
</li><li> <b>All Protein-Protein</b> - checking this box will select all Include Protein-Protein boxes.
</li></ul>
<a name="Marker_2"></a><h3>Marker</h3>
<p>The marker name (same as in the "Activated Marker List"). Appears
italicized to indicate that interaction information has not yet been
retrieved. Bold face font indicates that interaction information has
been retrieved (and is displayed).
</p>
<a name="Gene_2"></a><h3>Gene</h3> 
<p>The gene name corresponding to the marker, if loaded from an annotation file (same as in the "Activated Marker List").
</p>
<a name="Gene_Type"></a><h3>Gene Type</h3> 
<p>A gene type designation, derived from the gene's GO annotation The list of type codes is the same as that under <b>Type</b> under "Activated Marker List".
</p>
<a name="GO_Description"></a><h3>GO Description</h3>
<p>The GO annotation of the gene. Right clicking on the column brings
up the the gene GO classification of the gene across the 3 top-level GO
categories: Component, Function and Process.
</p><p>Hovering the mouse cursor over an entry in the GO Description
column will display a short summary of the Gene Ontology terms
associated with that entry.
</p><p><br>
<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_GO_hover_text.png" class="image" title="Image:T_CNKB_GO_hover_text.png"><img src="cnkb_files/T_CNKB_GO_hover_text.png" alt="Image:T_CNKB_GO_hover_text.png" longdesc="/workbench/index.php/Image:T_CNKB_GO_hover_text.png" width="603" height="131"></a>
</p><p><br>More extensive GO annotations can be viewed for desired
genes in the Selected Marker list by right-clicking on its entry. A
pop-up menu will offer a choice of the three categories of GO
annotation: Component, Function and Process. Expanding one of these
terms will show the available annotations for that gene.
</p><p><br>
<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_GO_rightclick_menu.png" class="image" title="Image:T_CNKB_GO_rightclick_menu.png"><img src="cnkb_files/T_CNKB_GO_rightclick_menu.png" alt="Image:T_CNKB_GO_rightclick_menu.png" longdesc="/workbench/index.php/Image:T_CNKB_GO_rightclick_menu.png" width="513" height="152"></a>
</p><p><br>
</p><p><br>
</p>
<a name=".23_of_Protein-DNA.2C_Protein-Protein_Interactions"></a><h3># of Protein-DNA, Protein-Protein Interactions</h3>
<p>These entries display the number of interactions (of each type) that
pass the confidence threshold specified by the x-axis slider in the
Throttle Graph. The values in these columns will change when the slider
moves. A value will be displayed grayed out if its corresponding
checkbox (from the first 2 columns in the table) is not enabled. This
acts as a visual reminder that those interaction will not be included
in the interaction network constructed when clicking on the "Create
Network" button.
</p>
<a name="Display_Preferences"></a><h3>Display Preferences</h3>
<p>The “"Display Preferences" button can be used to control how many of
the columns listed above are visible within the Selected Marker List.
</p>
<a name="Throttle_Graph"></a><h2>Throttle Graph</h2>
<p>This interactive graph allows users to "throttle" (for the genes in
the Selected Markers table) which interactions to work with, using as a
criterion the interactions’ confidence indicator. As the required
threshold of likelihood of the interactions is increased, the sum of
interactions meeting this criterion decreases.
</p><p><br>
<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_Throttle_Graph.png" class="image" title="Image:T_CNKB_Throttle_Graph.png"><img src="cnkb_files/T_CNKB_Throttle_Graph.png" alt="Image:T_CNKB_Throttle_Graph.png" longdesc="/workbench/index.php/Image:T_CNKB_Throttle_Graph.png" width="653" height="438"></a>
</p><p><br>
</p>
<ul><li> <b>Right-click menu</b> - Right-clicking on the Throttle Graph will bring up a menu which allows the graph to be customized.
</li></ul>
<p><br>
<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_ThrottleGraph_rightclick_menu.png" class="image" title="Image:T_CNKB_ThrottleGraph_rightclick_menu.png"><img src="cnkb_files/T_CNKB_ThrottleGraph_rightclick_menu.png" alt="Image:T_CNKB_ThrottleGraph_rightclick_menu.png" longdesc="/workbench/index.php/Image:T_CNKB_ThrottleGraph_rightclick_menu.png" width="282" height="277"></a>
</p>
<a name="Creating_an_interaction_network_and_viewing_it_in_Cytoscape"></a><h1>Creating an interaction network and viewing it in Cytoscape</h1>
<p>Once the desired set of markers is present and its interation data
has been retrieved from the database, an adjacency matrix can be
computed by clicking the "<b>Create Network</b>" button. The resulting
matrix is placed in the Project Folders component under its parent
microarray dataset. The adjacency matrix is visualized in the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Tutorial_-_Cytoscape_Network_Viewer" title="Tutorial - Cytoscape Network Viewer"> Cytoscape Viewer</a>.
</p><p><b>Please note that if the targets retrieved from the Knowledge
Base include genes not present in the active microarray dataset, then
those targets will not be displayed in the Cytoscape viewer.</b> 
</p><p><br>
The Cytoscape Viewer maintains a list of networks which it has
currently loaded. It allows individual loaded networks to be deleted.
However, the network can be reloaded by clicking on its entry in the
Project Folders component. Cytoscape controls are more fully described
in the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Tutorial_-_Cytoscape_Network_Viewer" title="Tutorial - Cytoscape Network Viewer"> Cytoscape</a> component tutorial.
</p>
<a name="Example_of_using_the_CNKB_component"></a><h1>Example of using the CNKB component</h1>
<p>This example uses the Bcell-100.exp dataset available in the
data/public_data directory of geWorkbench, and further described on the
<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Download" title="Download">Download</a>
page. Briefly, this dataset is composed of 100 Affymetrix HG-U95Av2
arrays on which various B-cell lines, both normal and cancerous, were
analyzed. Thus it explores a potentially wide variety of expression
phenotypes. </p>
<a name="Prerequisites"></a><h2>Prerequisites</h2>
<ul><li> Obtain the annotation file for the HG-U95Av2 array type from the Affymetrix NetAffx website (<a href="http://www.affymetrix.com/analysis/index.affx" class="external free" title="http://www.affymetrix.com/analysis/index.affx" rel="nofollow">http://www.affymetrix.com/analysis/index.affx</a>).
The name will be similar to "HG_U95Av2.na29.annot.csv", where na29 is
the version number. Loading the annotation file associates gene names
and Gene Ontology information with the Affymetrix probeset IDs (see the
geWorkbench FAQ for details on obtaining these files).
</li></ul>
<a name="Loading_the_example_data"></a><h2>Loading the example data</h2>
<ol><li> Load the Bcell-100.exp dataset into geWorkbench as type "Affymetrix File Matrix".  (See <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Tutorial_-_Local_Data_Files" title="Tutorial - Local Data Files"> Local Data Files</a>).
</li><li> When prompted, load the annotation file.
</li><li> Create and activate a set of markers in the Markers
component. For this example, load the file
"70_TFs_from_HG-U95Av2.na28.csv" directly into the Markers component by
pushing its "<b>Load Set</b>" button and browsing to the file located in the geWorkbench directory data/public_data.
</li></ol>
<p><br>
<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_70TFs_loaded_in_Markers.png" class="image" title="Image:T_CNKB_70TFs_loaded_in_Markers.png"><img src="cnkb_files/T_CNKB_70TFs_loaded_in_Markers.png" alt="Image:T_CNKB_70TFs_loaded_in_Markers.png" longdesc="/workbench/index.php/Image:T_CNKB_70TFs_loaded_in_Markers.png" width="230" height="704"></a>
</p><p><br>
When the set is activated by checking the box to the left of its name,
the 70 transcription factor markers will appear in the "Activate
Markers" list in the CNKB component.
</p><p><br>
</p>
<a name="Setting_up_the_query_in_the_CNKB_component"></a><h2>Setting up the query in the CNKB component</h2>
<p><br>1. We will now show how to select just a subset of the activated
markers for use in the query. Highlight the first 10 markers on the
list (Shift-right-click). Right-clicking on the "Activated Markers"
list will then bring up a menu which allows one to move all activated
markers or just the highlighted markers to the "Selected Markers List"
for querying. </p><p>Select "<b>Add Selected Markers to Selected Markers List</b>".
</p><p><br>
<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_Activated_List_rightclick_menu.png" class="image" title="Image:T_CNKB_Activated_List_rightclick_menu.png"><img src="cnkb_files/T_CNKB_Activated_List_rightclick_menu.png" alt="Image:T_CNKB_Activated_List_rightclick_menu.png" longdesc="/workbench/index.php/Image:T_CNKB_Activated_List_rightclick_menu.png" width="865" height="591"></a>
</p><p><br>
The 10 selected markers now appear in the the "Selected Markers List".
They appear in red and in italics, indicating the query has not yet
been performed.
</p><p><br>
<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_10_Selected_Markers.png" class="image" title="Image:T_CNKB_10_Selected_Markers.png"><img src="cnkb_files/T_CNKB_10_Selected_Markers.png" alt="Image:T_CNKB_10_Selected_Markers.png" longdesc="/workbench/index.php/Image:T_CNKB_10_Selected_Markers.png" width="855" height="528"></a>
</p><p><br>
2.  Now hit the "<b>Refresh</b>" button to perform the
query against the Cellular Networks Knowledge Base database. The
results are now displayed in regular font, blue letters to show that
the query has been completed.
</p><p><br>
<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_ProtDNA_confidence75.png" class="image" title="Image:T_CNKB_ProtDNA_confidence75.png"><img src="cnkb_files/T_CNKB_ProtDNA_confidence75.png" alt="Image:T_CNKB_ProtDNA_confidence75.png" longdesc="/workbench/index.php/Image:T_CNKB_ProtDNA_confidence75.png" width="898" height="723"></a>
</p><p><br>3. The Throttle Graph allows one to set a minimum confidence
requirement on interactions that will be used to create a network. In
the figure above we have adjusted the cutoff to 75% confidence.
</p><p><br>
4. Check the "<b>All Protein-DNA</b>" box to limit network creation to only protein-DNA interactions.
</p><p>5. Hit the "<b>Create Network</b>" button.
</p><p>8. The resulting adjacency matrix is displayed in the <b>Cytoscape</b> component.
</p><p><br>
<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_10at75_Cytoscape.png" class="image" title="Image:T_CNKB_10at75_Cytoscape.png"><img src="cnkb_files/T_CNKB_10at75_Cytoscape.png" alt="Image:T_CNKB_10at75_Cytoscape.png" longdesc="/workbench/index.php/Image:T_CNKB_10at75_Cytoscape.png" width="903" height="756"></a>
</p><p><br>
</p><p>9. Within Cytoscape, changing the layout to yFiles-&gt;Hierachic will redraw the network in a hierachical fashion.
</p><p><br>
<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Image:T_CNKB_10at75_Cytoscape_hierarchical.png" class="image" title="Image:T_CNKB_10at75_Cytoscape_hierarchical.png"><img src="cnkb_files/T_CNKB_10at75_Cytoscape_hierarchical.png" alt="Image:T_CNKB_10at75_Cytoscape_hierarchical.png" longdesc="/workbench/index.php/Image:T_CNKB_10at75_Cytoscape_hierarchical.png" width="908" height="641"></a>
</p>
<a name="Appendix_-_Data_Sources"></a><h1>Appendix - Data Sources</h1>
<p>Cellular Network Knowledge Base (CNKB) Source Descriptions &amp; Interaction Statistics
</p>
<a name="CNKB_for_geWorkbench_queries"></a><h2>CNKB for geWorkbench queries</h2>
<p>The CNKB itself is comprised of data from sources shown in the
following sections. However, geWorkbench only queries for two types of
interactions, protein-protein and protein-dna, coded in the database as
"ppi" and "pdi". </p><p>As of 7/1/2009, the following datasources will match these geWorkbench queries:
</p>
<ul><li> PPI: BIND, HPRD, INTACT, INTERACTOME, MIPS.
</li><li> PDI: BIND, INTERACTOME.
</li></ul>
<a name="B-cell_lymphoma_Interactome__.28INTERACTOME.29_.28Della_Favera.2FCalifano_labs.2C_collected_by_Celine_Lefebvre.29"></a><h3>B-cell lymphoma Interactome  (INTERACTOME) (Della Favera/Califano labs, collected by Celine Lefebvre)</h3>
<ul><li> 12,902 protein-dna interactions
</li><li> 22,734 protein-protein interactions
</li></ul>
<a name="Munich_Information_Center_for_Protein_Sequences__.28MIPS.29"></a><h3>Munich Information Center for Protein Sequences  (MIPS)</h3>
<ul><li> <a href="http://mips.gsf.de/proj/ppi/" class="external free" title="http://mips.gsf.de/proj/ppi/" rel="nofollow">http://mips.gsf.de/proj/ppi/</a>
</li><li> 322 protein-protein interactions
</li></ul>
<a name="Protein-protein_interaction_database_at_EBI__.28INTACT.29._No_species_restriction."></a><h3>Protein-protein interaction database at EBI  (INTACT). No species restriction.</h3>
<ul><li> <a href="http://www.ebi.ac.uk/intact/site/index.jsf" class="external free" title="http://www.ebi.ac.uk/intact/site/index.jsf" rel="nofollow">http://www.ebi.ac.uk/intact/site/index.jsf</a>
</li><li> 7,701 interactions
</li></ul>
<a name="The_Molecular_INTeraction_database__.28MINT.29"></a><h3>The Molecular INTeraction database  (MINT)</h3>
<ul><li> <a href="http://mint.bio.uniroma2.it/mint/Welcome.do" class="external free" title="http://mint.bio.uniroma2.it/mint/Welcome.do" rel="nofollow">http://mint.bio.uniroma2.it/mint/Welcome.do</a>
</li><li> 3,196 human interactions
</li><li> MINT focuses on experimentally verified protein interactions mined from the scientific literature by expert curators.
</li></ul>
<a name="The_Biomolecular_Interaction_Network_Database__.28BIND.29"></a><h3>The Biomolecular Interaction Network Database  (BIND) </h3>
<ul><li> <a href="http://bond.unleashedinformatics.com/Action?pg=23299#BIND" class="external free" title="http://bond.unleashedinformatics.com/Action?pg=23299#BIND" rel="nofollow">http://bond.unleashedinformatics.com/Action?pg=23299#BIND</a>
</li><li> 48,573 protein-protein interactions (various species)
</li></ul>
<a name="Database_of_Interacting_Proteins_.28DIP.29"></a><h3>Database of Interacting Proteins (DIP)</h3>
<ul><li> <a href="http://dip.doe-mbi.ucla.edu/dip/Guide.cgi" class="external free" title="http://dip.doe-mbi.ucla.edu/dip/Guide.cgi" rel="nofollow">http://dip.doe-mbi.ucla.edu/dip/Guide.cgi</a>
</li><li> 820 human protein-protein interaction pairs
</li></ul>
<a name="Reactome_knowledgebase_.28REACTOME.29"></a><h3>Reactome knowledgebase (REACTOME)</h3>
<ul><li> <a href="http://www.reactome.org/" class="external free" title="http://www.reactome.org/" rel="nofollow">http://www.reactome.org/</a>
</li><li> Reactome knowledgebase
</li><li> 27,534 human protein-protein interaction pairs
</li></ul>
<a name="Human_Protein_Reference_Database__.28HPRD.29"></a><h3>Human Protein Reference Database  (HPRD)</h3>
<ul><li> <a href="http://www.hprd.org/" class="external free" title="http://www.hprd.org/" rel="nofollow">http://www.hprd.org/</a>
</li><li> 36,501 protein-protein interactions
</li></ul>
<a name="Geneways_.28GENEWAYS.29"></a><h3>Geneways (GENEWAYS)</h3>
<ul><li> <a href="http://geneways.genomecenter.columbia.edu/" class="external free" title="http://geneways.genomecenter.columbia.edu/" rel="nofollow">http://geneways.genomecenter.columbia.edu/</a>
</li><li> 27,938 interactions 
</li><li> GeneWays is a system for automatically extracting, analzying,
visualizing and integrating molecular pathway data from the research
literature.
</li></ul>
<a name="References"></a><h1>References</h1>
<ul><li> Mani KM, Lefebvre C, Wang K, Lim WK, Basso K, Dalla-Favera R,
Califano A., " A systems biology approach to prediction of oncogenes
and molecular perturbation targets in B-cell lymphomas", Molecular
Systems Biology 4:169, 2008 <a href="http://www.ncbi.nlm.nih.gov/pubmed/18277385?ordinalpos=1&amp;itool=EntrezSystem2.PEntrez.Pubmed.Pubmed_ResultsPanel.Pubmed_RVBrief" class="external text" title="http://www.ncbi.nlm.nih.gov/pubmed/18277385?ordinalpos=1&amp;itool=EntrezSystem2.PEntrez.Pubmed.Pubmed ResultsPanel.Pubmed RVBrief" rel="nofollow">link to paper</a>
</li></ul>

</body></html>